
---
title: "Ts_Tv_Analysis_Figures"
author: "Daniel Lyons"
date: "October 4, 2016"
output: 
  html_document:
    toc: yes
    toc_depth: 4
    toc_float:
      collapsed: FALSE
      smooth_scroll: TRUE
    code_folding: show
    highlight: haddock
    number_sections: FALSE
---
Note: "#C" comments are some commented out checks of the code.

#Analysis
##Setup & Initial Parsing of Data
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = T, warnings = FALSE, message = F)
library(knitr)
library(plyr)
library(reshape2)
library(magrittr)
library(ggplot2)
library(ggrepel)
library(cowplot)
library(Biostrings)
# library(Hmisc) #masks translate in biostring...needed for rcorr function later when doing spearman for bloom and lauring.

data_path = ("../Data")
analysis.path = ("../Analysis")

```

First, load all data.
```{r LOAD_DATA, include = F}
setwd(data_path)

#other studies----
##fitness data
np = read.csv("NP_bloom_MBE_2014_Supplementary_file_1 (1).csv", stringsAsFactors = F)

np2 = read.table("NP_bloom_pr8(Supp_file_1_Text_mean_pr1934_prefs).txt", stringsAsFactors = F, header = T)

ha = read.table("Supplemental_File_2_HApreferences.txt",stringsAsFactors = F, header = T)
ha_rescaled = read.table("Supplemental_File_3_HApreferences_rescaled.txt",stringsAsFactors = F, header = T)

env = read.table("S1_File_Bloom_env.txt",stringsAsFactors = F,header = T)

#C check if env published recently is same as env used from biorxiv, replaced 170416. They are.
# env.old = read.table("S1_File_Bloom_env.biorvix.txt",stringsAsFactors = F,header = T)
# identical(env,env.old)

hiv_in = read.csv("HIV_IN.csv", stringsAsFactors = F)
hiv_ca = read.csv("HIV_CA.csv", stringsAsFactors = F)
phix174 = read.table("phiX174_Domingo_Calap_09.txt", sep = " ", stringsAsFactors = F, fill = TRUE, row.names = NULL)
qb = read.table("qb_Domingo_Calap_09.txt", sep = " ", stringsAsFactors = F, fill = TRUE, row.names = NULL)
vsv = read.delim("vsv_Sanjuan_04.txt", stringsAsFactors = F) 
vsv_full = read.csv("VSV_single_mutants_IDs.csv", stringsAsFactors = F)
tev = read.delim("TEV_Carrasco_07.txt", stringsAsFactors = F, comment.char = "#")
tev_full = read.delim("TEV_Carrasco_07_full.txt", stringsAsFactors = F, comment.char = "#")
f1_1 = read.delim("f1_pg1_Peris_10.txt", stringsAsFactors = F, comment.char = "#")
f1_2 = read.delim("f1_pg2_Peris_10.txt", stringsAsFactors = F, comment.char = "#", header=F)
f1_3 = read.delim("f1_pg3_Peris_10.txt", stringsAsFactors = F, comment.char = "#", header=F)

##bloom sequences & check-----
np_cds = readDNAStringSet("bloom_plasmids_gong_elife-00631-supp1-v1-download.txt") %>% .$`reverse complement of vRNA for A/Aichi/2/1968 (H3N2) NP as taken from pHWAichi68-NP` %>% subseq(.,46,1542)


np2_cds = readDNAStringSet("np_pr8.txt") %>% .$`EF467822.1 Influenza A virus (A/Puerto Rico/8/34(H1N1)) segment 5, complete sequence` %>% subseq(.,46,1542)

ha_cds = readDNAStringSet("bloom_plasmids_gong_elife-00631-supp1-v1-download.txt") %>% .$`reverse complement of vRNA for A/WSN/1933 (H1N1) HA as taken from pHW184-HA` %>% subseq(33,1730)

env_cds = readDNAStringSet("bloom_gp120gp41_plasmid.txt") %>% .$gp120_gp41

ha_prot = translate(ha_cds) #C check if position is correct by translating first several aas to see if its start of protein, it's correct.

np_prot = translate(np_cds) #C checked first and last several aas, its's correct.

np2_prot = translate(np2_cds)

env_prot = translate(env_cds) %>% subseq(31,707) #bloom position/site numbering is slightly different than just straight 1:length(protein), but site 142 has 142,142a-e. So number of last site is 5 less than actual length of protein, plus they didn't mutate full length of protein, instead, 677 sites.


#C check if need to take out stop codons:done. yes in some.
#C check if np is truly starting with met in mutations as well: it does.

##annot bloom WT codon----
np %<>% mutate(WT_codon = codons(np_cds %>% subseq(.,1,length(np_cds)-3)) %>% as.character %>% tolower(),
               WT_site_pref = apply(np, 1,function(row){
                 row[paste0("PI_",row["WT_AA"])]
               })) #get wt codons from cds, not stop codon

np2 %<>% rename(c(WT = "WT_AA", POSITION  = "X.SITE")) 
np2 %<>% mutate(WT_codon = codons(np2_cds %>% subseq(.,1,length(np2_cds)-3)) %>% as.character %>% tolower(),
               WT_site_pref = apply(np2, 1,function(row){
                 row[paste0("PI_",row["WT_AA"])]
               })) #get wt codons from cds, not stop codon



ha %<>% rename(c(WT = "WT_AA", POSITION  = "X.SITE")) 

ha %<>% mutate(WT_codon = codons(ha_cds %>% subseq(.,4,length(ha_cds)-3)) %>% as.character %>% tolower(),
               WT_site_pref = apply(ha, 1,function(row){
                 row[paste0("PI_",row["WT_AA"])]
               })) #get wt codons from cds, not start or stop codons.


AA2nuc_pos = function(aa_site){
  c(start = 3*aa_site-2, end = 3*aa_site)
} #to useful fnxs: get the start and end nucleotide positions for an aa position, if both are starting from 0


env %<>% subset(select = -c(PI_.)) %>%rename(c(WT = "WT_AA", POSITION  = "X.SITE")) #exclude the stop mutations.
                
env %<>% mutate(WT_codon = codons(env_cds %>% subseq(.,AA2nuc_pos(31)["start"],AA2nuc_pos(707)["end"])) %>% as.character %>% tolower(),
               WT_site_pref = apply(env, 1,function(row){
                 row[paste0("PI_",row["WT_AA"])]
               })) #get wt codons from cds, not start or stop codons.

#C 17_01_17: count(env,.(WT_AA,WT_codon)) #all AAs are annotated with correct codons, checking against an independent codon table

#lauring data----
##fitness
lethal_muts = read.csv("Table_1rev2.csv", stringsAsFactors = F) #lauring
viable_muts = read.csv("Table_3rev2.csv", stringsAsFactors = F) #lauring

#end----
```

Note: 
Stoltzfus et. al. didn't use any available mutational scanning data for various reasons.  Two studes they excluded because they had not reported low fitness mutants; we don't use these anyway. Three others they excluded because they didn't report nucleotide mutations (Bloom-HA, NP_H3N2, Melamed et al.-S. cerevisae pol-a binding protein), but that's obviously fine because we can work backwards from the AA mutations.  5 other studies they excluded to hypermutation and oxidative damage concerns, but we don't use these studies anyway. Details about the hypermutation/oxidative damage concerns can be found in their supplemental.

For Bloom deep mutational scanning data (NP, HA, env) my steps for converting the amino acid sites to nucleotide sites and figuring out TS/TV for any amino acid change:
1) Get plasmid sequence upon which Bloom made mutations:
  a)
    i) NP is Aichi/1968 NP, an H3N2 (described in Gong et. al., eLife 2013)
    ii)NP2 is A/Puerto Rico/8/1934 (H1N1, PR8) (no plasmid was given in original bloom paper nor an appropriate reference, so looked it up on genbank.  GenBank: EF467822.1)
    ii) HA is A/WSN/1933 (H1N1) (same strain as our mutants) (also described in Gong et. al., eLife 2013)
  b) HIV plasmid is provided in the paper (Experimental estimation of the effects of all amino-acid mutations to HIV Env, preprint)
    i) It was a genbank file (S4\_File\_Bloom\_env.gb), used benchling to find the gp120-gp41 annotated sections (whole env gene), then copied from benchling to a txt file

2) Find ORF using NCBI ORF Finder (https://www.ncbi.nlm.nih.gov/orffinder/)
  a) For env, the whole annotated gp120-gp41 sections was an orf, which I loaded here.

3) Blast the ORF that seems the right length to confirm it is the right ORF

4) Write down the start and end positions of the correct ORF:
  a) NP: 46,1542
  b) NP2: 46, 1542
  b) HA: 33,1730
  c) Env: whole section

5) Translate the orf, check first couple amino acid positions against the site preference table, column showing WT amino acids at those positions, just to make sure.
  i) The NP one is exactly like translation, they didn't mutate stop codon though.
  ii) The HA one-they didn't mutate start codon or stop codon.
  iii) Env: bloom position/site numbering is slightly different than just simply 1:length(protein), but site 142 has 142,142a-e. So number of last site is 5 less than actual length of protein, plus they didn't mutate full length of protein, instead, 677 sites.  Furthermore, they measured site preferences of stop mutations as well, which I excluded here bc such sites are not informative of the amino acids accessible by ts and tv.

6) Then annotate the WT codon for each position based on the ORF.

Note: steps 1-6 did in load_data chunk above.

7) Now further parse bloom data--need to annotate whether each AA_Sub from each WT codon are accessible by TS, TV, BOTH, or NA (cannot result from a single nucleotide mutation), and then annotate all other relevant info, eg a radical or conservative amino acid class change or not.

First need a codon table with all the possible mutations from it and the types of amino acid changes they are (class changes).

First code in amino acids and their types.
```{r  AA_CLASSES}

aa = read.csv("AA_list.csv", stringsAsFactors = F)
colnames(aa) = c("AA","code3","code1")

aa %<>% subset(AA != "asparagine or aspartic acid" & AA != "glutamine or glutamic acid")

#C 17_01_17: all aas have correct code1 and code3

#charge----
pos = c("R","H","K")
neg = c("D","E")
neut =  c('A', 'N', 'C', 'Q', 'G', 'I', 'L', 'M', 'F', 'P', 'S', 'T', 'W', 'Y', 'V')

#polarity----
polar = c('R', 'N',"D", 'C', 'Q', "E", 'G',"H","K",'S', 'T', 'Y')
nonpolar = c("A","I","L","M","F","P","W","V")

#polarity & volume----
special = "C"
neut_small = c("A","G","P","S","T")
polar_small = c("N","D","Q","E")
polar_large = c("R","H","K") 
nonpolar_small = c("I","L","M","V")
nonpolar_large = c("F","W","Y")

#class lists----
aa %<>% mutate(charge = ifelse(code1 %in% pos, "pos", ifelse(code1 %in% neg,"neg",ifelse(code1 %in% neut,"neut",NA))),
               polarity = ifelse(code1 %in% polar, "polar", ifelse(code1 %in% nonpolar,"nonpolar",NA)),
               pol_vol = ifelse(code1 %in% special, "special", ifelse(code1 %in% neut_small,"neut_small",ifelse(code1 %in% polar_small,"polar_small",ifelse(code1 %in% polar_large,"polar_large",ifelse(code1 %in% nonpolar_small, "nonpolar_small",ifelse(code1 %in% nonpolar_large, "nonpolar_large",NA))))))) #C 17_01_18

firstup <- function(x) {
   substr(x, 1, 1) <- toupper(substr(x, 1, 1))
x
}

aa$code3 = firstup(aa$code3)

charge = list(pos = c("R","H","K"), neg = c("D","E"), neut =  c('A', 'N', 'C', 'Q', 'G', 'I', 'L', 'M', 'F', 'P', 'S', 'T', 'W', 'Y', 'V'))

polarity = list(polar = c('R', 'N',"D", 'C', 'Q', "E", 'G',"H","K",'S', 'T', 'Y'), nonpolar = c("A","I","L","M","F","P","W","V"))

pv = list(special = "C", neut_small = c("A","G","P","S","T"), polar_small = c("N","D","Q","E"), polar_large = c("R","H","K"), nonpolar_small = c("I","L","M","V"), nonpolar_large = c("F","W","Y"))

#end----
```


Now make all possible 1 nucleotide mutations from codon table, annotating the properties of the mutation (what class the mutation is.
```{r CODON_TABLE_MK_MUTS}
#mk codon_table----
library(seqinr)
codon_freqs2 = uco("dummy",as.data.frame = T) %>% arrange(AA)
detach("package:seqinr", unload = T) 
library(Biostrings) #always detach seqinr and reload biostrings after using seqinr because it masks some things in biostrings, and biostrings is more generally useful
codon_table = codon_freqs2 %>% subset(AA!= "Stp",select = c(AA,codon)) #just get a codon table easily

codon_table = rename(codon_table,"AA" = "code3")

id_t2aa = sapply(codon_table$code3,grep,aa$code3)

codon_table %<>% mutate(AA = aa$AA[id_t2aa],code1 = aa$code1[id_t2aa],charge = aa$charge[id_t2aa],polarity = aa$polarity[id_t2aa],pv = aa$pol_vol[id_t2aa]) #C 17_01_18: all correct notations as compared to aa df

#add in stop codons to codon table
stops = c('taa','tag','tga')

stops = lapply(stops,function(s){
  c("STOP",s,rep("STOP",ncol(codon_table)-2))
}) %>% do.call(rbind,.)

colnames(stops) = colnames(codon_table)

codon_table %<>% rbind(.,stops)

#string fnxs----
str_rev = function(str_vector){
  strsplit(str_vector, "") %>% lapply(.,rev) %>% sapply(.,paste, collapse = "")
} #reverses each string in a character vector

each_v1_by_v2 = function(v1,v2,sep = ""){
  v1 = paste0(v1,sep)
  as.vector(outer(v1,v2,paste0))
} #Combines each of v1 with first of v2, then each of v1 with second of v2, etc., separated by sep.

each_v2_by_v1 = function(v1,v2,sep = ""){
  sapply(v1,function(n){
    sapply(v2,function(n2){
      paste0(n,sep,n2)
    })
  }) %>% as.vector
} #Combines first of v1 with each of v2, then second of v1 with each of v2, etc., separated by sep.

#mk muts----
nucleotides = c("a","g","c","t")
positions = c(1,2,3)

later_colnames = c(colnames(codon_table), each_v1_by_v2(nucleotides,positions))

mkmut1 = function(codon,nuc,position){
  codon_original = codon
  substr(codon,position,position) = nuc
  if( codon == codon_original){
    NA
  } else {
    substr(codon,position,position) = nuc
    codon
  }
} #mk one mutation for one codon at one position. No mutation is annotated as NA.


mkmut2 = function(nucs = nucleotides,codon, position){
  sapply(nucs,function(n){
    mkmut1(codon,n,position)
  })
} #mk all mutations for one codon at one position

mkmut3 = function(nucs = nucleotides,codon,pos = positions){
  sapply(pos,function(p){
    mkmut2(nucs = nucs,codon = codon, pos = p)
  }) %>% as.vector
} #mk all mutations for one codon at all positions


mkmut_all = function(nucs = nucleotides, codons, pos = positions){
  sapply(codons,function(c){
    mkmut3(nucs,codon = c,pos)
  }) %>% t()
}

codon_table = mkmut_all(codons = codon_table$codon %>% as.character) %>% cbind(codon_table,.)

colnames(codon_table) = later_colnames #C 17_01_18: Checked mutations and positions if codon created was correct for one codon. (Had checked this previously when I first made this for a couple other codons as well). Also checked for all codons with a at 1, g at 1 c at 2, whether the corresponding changes at those positions that would not cause a mutation (a1->a) were annotated correctly as <NA>.

#annot new AA, class----
id_index_value = function(id,index,value){
  sapply(id,function(x){
    if(is.na(x)){
      NA
    } else{
      value[grep(x,index)]
    }
  })
} #Have any vector for which I want to find the positions of those values in a second vector and then use those positions to extract values from a third vector.


ids_index_value = function(ids,index,value){
  apply(ids,2,function(i){
    id_index_value(id = i,index,value)
  })
} #Same as above, but for multiple ID vectors, where "ids" is a data.frame or matrix and ID vectors are in columns

ids_index_values = function(ids,index,values,prefix = paste0("n",colnames(values)), suffix = colnames(ids), sep = "_"){
  l = lapply(values,function(v){
    ids_index_value(ids = ids,index,value = v)
  }) %>% do.call(cbind,.)
  colnames(l) = each_v2_by_v1(prefix,suffix,sep)
  l
} #Same as above, but for multiple id vectors and multiple value vectors


codon_table = ids_index_values(codon_table %>% subset(select = -c(code3,codon,AA,code1,charge,polarity,pv)),codon_table$codon %>% as.vector(),codon_table %>% subset(select = c(code1,charge,polarity,pv))) %>% cbind(codon_table,.) #C 17_01_18: logic is correct.  Also systematically checked that all new ATG were annotated as M, that all new unique codons in a1 were only annotated with a unique amino acid code in ncode1_a1, and unique new charge,polarity,pv classes.  Also each new codon in a1 should appear exactly 3 times, as there will be 3 other unique codons from which it can mutate-checked this for a1-3, and all nucleotides actg at position 1.  All confirmed.  Then systematically checked in a batched way:
# 
# ##if all new codons and their code1 annotations are correct:
# new.comp = function(check.against = "code1",ncomp = "ncode1_"){
#   x = data.frame(ncodon= codon_table[,each_v1_by_v2(nucleotides,positions)] %>% as.matrix %>% as.vector,
#                  compare = codon_table[,paste0(ncomp,each_v1_by_v2(nucleotides,positions))] %>% as.matrix %>% as.vector,
#                  row.names = NULL)
#   x
# 
#   y = cbind(codon_table %>% subset(select = c("codon",check.against)) %>% arrange(codon),subset(x,!duplicated(ncodon) & !is.na(compare)) %>% arrange(ncodon))
#   y
#   # 
#   mutate(y,rcodon = codon == ncodon,rcompare = y[,check.against]==compare)
# }#checks the new codons and new "ncomp" column (eg any annotation of that new codon), against the true codon and annotation (obtained in a different way, from the first "original" columns of the codon_table codons 1:64).
# 
# new.comp() #yes
# 
# ##if all new codons and their different class annotations are correct
# new.comp(check.against = "charge",ncomp = "ncharge_") #yes
# new.comp(check.against = "polarity",ncomp = "npolarity_") #yes
# new.comp(check.against = "pv",ncomp = "npv_") #yes
# 
# 

#end----
```

Now annotate in "mutation" codon table whether a mutation is a TS/TV S/NS and R/C class change.
```{r ANNOT_MUT}
#annotate whether mutation was a ts,s, or class change
##ts
ts_muts = c("AG","GA","CU","UC","TC","CT") 

cods = strsplit(codon_table$codon %>% as.vector(),"")

mts = sapply(codon_table[,each_v1_by_v2(nucleotides,positions)],function(mut){
  mut = strsplit(mut %>% as.vector(),"")
  sapply(1:length(cods), function(c){
    cods = cods[c] %>% unlist
    mut = mut[c] %>% unlist
    if(NA %in% mut){
      NA
    } else{
      cods[!(cods ==  mut)] %>% paste0(mut[!(cods ==  mut)]) %in% tolower(ts_muts)
    }
  })
})

colnames(mts) = paste0("TS_",each_v1_by_v2(nucleotides,positions))

###s, class change
get_mut_colnames = function(base_colname){
  colnames(codon_table)[grep(paste0(base_colname,"_"),colnames(codon_table))]
}

compare_columns = function(data_frame,base, against, prefix = paste0(base,"_")){
  compare = data_frame[,base] %>% as.vector() == data_frame[,against] %>% as.matrix()
  colnames(compare) = paste0(prefix,against)
  compare
} #compare one base column against multiple other columns, where base and against are column names and prefix is what I want to call the comparison.

codon_table = cbind(codon_table,
                    mts,
                    compare_columns(codon_table,base = "code1",against = get_mut_colnames("code1") , prefix = "S_"),
                    compare_columns(codon_table,base = "charge",against = get_mut_colnames("charge"),prefix = "C_"),
                    compare_columns(codon_table,base = "polarity",against = get_mut_colnames("polarity"),prefix = "C_"),
                    compare_columns(codon_table,base = "pv",against = get_mut_colnames("pv"),prefix = "C_")
                    )
#end----
```

Okay now check whether TS, S_NS, and class changes have been annotated correctly. 
```{r CHECK.CODON.TABLE}
#C check whether TS, S_NS, and class changes have been annotated correctly. 
# #check if TS_ columns are annotated correctly----
# #if non mutations are annotated as NA
# ch.na = function(pos,nuc){
#   count(codon_table %>% subset(substring(codon,pos,pos) %in% nuc),c("codon",paste0(nuc,pos),paste0("TS_",nuc,pos)))
# }
# 
# 
# fn.p.n = function(fn){
#   lapply(positions,function(p){
#     lapply(nucleotides,function(n){
#       fn(p,n)
#     })
#   })
# }
# 
# fn.p.n(ch.na) #yes.
# 
# #if ts mutations are annotated as TRUE
# 
# ch.ts = function(pos,nuc){
#   if(nuc == "a"){ts = "g"}
#   else if(nuc == "g"){ts = "a"}
#   else if(nuc == "c"){ts = "t"}
#   else if(nuc == "t"){ts = "c"}
#   count(codon_table %>% subset(substring(codon,pos,pos) == ts),c("codon",paste0(nuc,pos),paste0("TS_",nuc,pos)))
# }
# 
# fn.p.n(ch.ts) #yes
# 
# #if tv mutations are annotated as F
# ch.tv = function(pos,nuc){
#   if(nuc %in% c("a","g")){tv = c("t","c")}
#   else if(nuc %in% c("t","c")){tv = c("a","g")}
#   count(codon_table %>% subset(substring(codon,pos,pos) %in% tv),c("codon",paste0(nuc,pos),paste0("TS_",nuc,pos)))
# }
# 
# fn.p.n(ch.tv) #yes
# 
# #check if comparison columns are correct----
# 
# 
# comp.check = function(original = "code1",mut.annot = "ncode1_",comparison = "S_ncode1_",t.f = T){
#   annot.col = paste0(mut.annot,each_v1_by_v2(nucleotides,positions))
#   
#   comparison.col = paste0(comparison,each_v1_by_v2(nucleotides,positions))
#   if(t.f){
#     lapply(1:length(annot.col),function(n){
#       count(codon_table[codon_table[,original] == codon_table[,annot.col[n]],],c(original, annot.col[n],comparison.col[n]))
#     })
#   }
#   else if(!t.f){
#     lapply(1:length(annot.col),function(n){
#       count(codon_table[codon_table[,original] != codon_table[,annot.col[n]],],c(original, annot.col[n],comparison.col[n]))
#     })
#   }
# } #for any annotation, get only the mutant annotations (MUT.ANNOT) that are equal to (t.f = T) or different from (t.f = F) the ORIGINAL annotation (codon mutated from), and see if the COMPARISON column corresponding to the annotation correctly identifies the mutation-so if t.f = T, should only see T (and NA for ones that are NA for whatever reason-not a mutation)
# 
# comp.check(t.f = T) #S annotated correctly
# comp.check(original = "charge",mut.annot = "ncharge_",comparison = "C_ncharge_")
# comp.check(original = "polarity",mut.annot = "npolarity_",comparison = "C_npolarity_")
# comp.check(original = "pv",mut.annot = "npv_",comparison = "C_npv_") #all C changes annotated correctly for three categories.
# 
# comp.check(t.f = F) #NS annotated correctly
# comp.check(original = "charge",mut.annot = "ncharge_",comparison = "C_ncharge_", t.f = F)
# comp.check(original = "polarity",mut.annot = "npolarity_",comparison = "C_npolarity_", t.f = F)
# comp.check(original = "pv",mut.annot = "npv_",comparison = "C_npv_", t.f = F) #all R changes annotated correctly for the three categories.

```

All annotations in codon table are correct. I have checked every single mutant annotation, codon, amino acid short code (1 letter), TS, S, AA-classes & class-changes, in a systematic yet easy way.


Now parse bloom data, working backwards from amino acid substitution to whether that substitution is caused by a TS, TV, both, or NA (NA = neither, can't get to substitution by 1 point mutation),  by using the codon table and each codon's all possible changes I made above.
```{r PARSE_BLOOM}
#naming fnxs----
remove_elements = function(vector,elements, whole = T){
  if(whole){
    vector[!vector %in% elements]
  } else{
    vector[!sapply(vector,function(name){
      grepl(elements,name)
    })]
  }
} #Remove elements from a vector by element values (elements), can use only part of element value (whole = F) to remove any element with that part in it. Where vector = vector to remove elements from, elements = vector of element values to remove, if using option whole = F, it must be a vector of length 1, otherwise can remove multiple whole element values.

retain_elements = function(vector,elements, whole = T){
  if(whole){
    vector[vector %in% elements]
  } else{
    vector[sapply(vector,function(name){
      grepl(elements,name)
    })]
  }
} #Same as remove.elements, but retains the desired elements instead.


#annotate fnxs & prep codon_table----
codon_table_m1 = melt(codon_table,.(code1,codon,AA,charge,polarity,pv),c(colnames(codon_table) %>% remove_elements(later_colnames,whole=T)))

codon_table_m2 = melt(codon_table,.(code1,codon,AA,charge,polarity,pv),get_mut_colnames("ncode1") %>% remove_elements("S_ncode1",whole=F))


cols1 = c("TS_","S_ncode1","C_ncharge","C_npolarity","C_npv")

cols2 = c("ncharge","npolarity","npv")

measure1 = sapply(cols1,function(c){
  subset(codon_table_m1,grepl(c,variable),select = value) %>% 
    rename(c(value = c))
}) 

measure2 = sapply(cols2,function(c){
  subset(codon_table_m1,grepl(c,variable) & !grepl("C_",variable),select = value) %>%
  rename(c(value = c))
})

codon_table_m1 = data.frame(codon_table_m2,do.call(cbind,measure1),do.call(cbind,measure2))
colnames(codon_table_m1) = gsub("\\..+","",colnames(codon_table_m1))
codon_table_m1 = rename(codon_table_m1,c(variable = "codon_position",value = "mut_aa",TS_ = "TS_TV",S_ncode1 = "S_NS")) %>%
  mutate(codon_position = gsub("ncode1_","",codon_position),
         mut_aa = ifelse(is.na(mut_aa),code1,mut_aa),
         TS_TV = as.vector(TS_TV)) %>% 
  mutate(TS_TV = ifelse(is.na(TS_TV),"not.mut",TS_TV)) #here label NAs in codon table, which are not mutations as not.mut

tstv.fnx = function(tstv,aamut,wtcodon){
  if(tstv == "other"){ #TS_TV is annotated as other as "TBD" place holder in annot.bloom function if it is not S_NS is not TRUE, in other words, if mutation is NS.
    ct = codon_table_m1 %>% subset(codon == wtcodon & TS_TV != "not.mut")
    if(!(aamut %in% ct$mut_aa)){
      tstv = "MUT23" #here if one of bloom's 20 amino acid mutations is not found in the aa mutations i can get from the corresponding natural codon from a single mutation in the as in codon table, it is obviously then a 2-3 codon mutation.
    } 
    else{
      ct %<>% subset(mut_aa == aamut)
      tstv = ifelse((ct$TS_TV %>% unique %>% length)>1,
                    "BOTH", 
                    ct$TS_TV %>% unique) #annotates as TS or TV if only one of those
    }
  }
  
  else{
    tstv = tstv
  }
  
  tstv
}

#C I used codon_table_m1 to check individually for atg codon whether all the mutations were correctly noted as TS-TV, S-NS. mut23 I checked for a couple.


annot.bloom = function(df){
  df = df %>% 
    melt(.,.(X.SITE,WT_AA,WT_codon,SITE_ENTROPY,WT_site_pref),retain_elements(colnames(df),"PI",whole=F)) %>% 
    rename(.,c("variable" = "AA_Mut", "value" = "Mut_site_pref")) %>% 
    mutate(AA_Mut = gsub("PI_","",AA_Mut))
  
  df %<>% mutate(S_NS = WT_AA == AA_Mut,TS_TV = ifelse(S_NS,"S","other")) #other means to be decided whether it is is ts, tv, mut23 or both.
  
  

  df = ddply(df,.(X.SITE,TS_TV,AA_Mut,WT_codon),mutate,TS_TV = tstv.fnx(TS_TV,AA_Mut,WT_codon))

  df %<>% mutate(Rel_site_pref = (Mut_site_pref %>% as.numeric) / (WT_site_pref %>% as.numeric),
                 wt.charge = ifelse(WT_AA %in% charge$pos,
                                     "pos",
                                     ifelse(WT_AA %in% charge$neg,
                                            "neg",
                                            ifelse(WT_AA %in% charge$neut,
                                                   "neut",
                                                   NA))),
                mut.charge = ifelse(AA_Mut %in% charge$pos,
                                     "pos",
                                     ifelse(AA_Mut %in% charge$neg,
                                            "neg",
                                            ifelse(AA_Mut %in% charge$neut,
                                                   "neut",
                                                   NA))),
                wt.polarity = ifelse(WT_AA %in% polarity$polar,
                                       "polar",
                                       ifelse(WT_AA %in% polarity$nonpolar,
                                              "nonpolar",
                                              NA)),
                mut.polarity = ifelse(AA_Mut %in% polarity$polar,
                                       "polar",
                                       ifelse(AA_Mut %in% polarity$nonpolar,
                                              "nonpolar",
                                              NA)),
                wt.pv = ifelse(WT_AA %in% pv$special,
                                 "special",
                                 ifelse(WT_AA %in% pv$neut_small,
                                        "neut_small",
                                        ifelse(WT_AA %in% pv$polar_small,
                                               "polar_small",
                                               ifelse(WT_AA %in% pv$polar_large,
                                                      "polar_large",
                                                      ifelse(WT_AA %in% pv$nonpolar_small,
                                                             "nonpolar_small",
                                                             ifelse(WT_AA %in% pv$nonpolar_large,
                                                                    "nonpolar_large",
                                                                    NA)))))),
                mut.pv = ifelse(AA_Mut %in% pv$special,
                                 "special",
                                 ifelse(AA_Mut %in% pv$neut_small,
                                        "neut_small",
                                        ifelse(AA_Mut %in% pv$polar_small,
                                               "polar_small",
                                               ifelse(AA_Mut %in% pv$polar_large,
                                                      "polar_large",
                                                      ifelse(AA_Mut %in% pv$nonpolar_small,
                                                             "nonpolar_small",
                                                             ifelse(AA_Mut %in% pv$nonpolar_large,
                                                                    "nonpolar_large",
                                                                    NA))))))
  )

  df %<>% mutate(C_ncharge = ifelse(TS_TV == "S",
                                   "S",
                                   ifelse(wt.charge == mut.charge,
                                          T,
                                          F)),
                 C_npolarity = ifelse(TS_TV == "S",
                                   "S",
                                   ifelse(wt.polarity == mut.polarity,
                                          T,
                                          F)),
                 C_npv = ifelse(TS_TV == "S",
                                   "S",
                                   ifelse(wt.pv == mut.pv,
                                          T,
                                          F))
                 )
  df
}

#annot bloom----
np.annot = np %>% subset(WT_AA != "N,N,H,H") %>% annot.bloom() 

np2.annot = np2 %>% annot.bloom()

ha.annot = ha %>% annot.bloom()

env.annot = env %>% annot.bloom()


#C check if rel site pref calc correctly by checking if all S are 1.
# count(np.annot %>% subset(S_NS == TRUE),.(Rel_site_pref))
# #check if it's annotating s-ns and ts-tv correctly (should never see  S_NS be TRUE and TS_TV be TRUE, BOTH,FALSE,or MUT23)
# count(np.annot,.(S_NS,TS_TV))
# #check if it's annotating wt aa's correctly. Checked by hand, each aa are correct.
# count(np.annot,.(WT_AA,wt.charge))
# count(np.annot,.(WT_AA,wt.polarity))
# count(np.annot,.(WT_AA,wt.pv))
# #check if it's annotating mut aa's correctly.  since wt-aa are correct, can compare count dfs without freq if identical.
# count(np.annot,.(WT_AA,wt.charge)) %>% subset(select = -c(freq)) %>% rename(c("WT_AA"="AA_Mut","wt.charge" ="mut.charge")) %>% identical(count(np.annot,.(AA_Mut,mut.charge)) %>% subset(select = -c(freq)))
# 
# count(np.annot,.(WT_AA,wt.polarity)) %>% subset(select = -c(freq)) %>% rename(c("WT_AA"="AA_Mut","wt.polarity" ="mut.polarity")) %>% identical(count(np.annot,.(AA_Mut,mut.polarity)) %>% subset(select = -c(freq)))
# 
# count(np.annot,.(WT_AA,wt.pv)) %>% subset(select = -c(freq)) %>% rename(c("WT_AA"="AA_Mut","wt.pv" ="mut.pv")) %>% identical(count(np.annot,.(AA_Mut,mut.pv)) %>% subset(select = -c(freq)))
# ##they're all correct.
# 
#C checked np as above a couple days before 17_01_18. #C 17_01_18: check the other data frames too:
# ch.bl.annot = function(df){
#   list (
#     #check if rel site pref calc correctly by checking if all S are 1.
#   rsp = count(df %>% subset(S_NS == TRUE),.(Rel_site_pref)),
#   #check if it's annotating s-ns and ts-tv correctly (should never see  S_NS be TRUE and TS_TV be TRUE, BOTH,FALSE,or MUT23)
#   sns.tstv = count(df,.(S_NS,TS_TV)),
#   #check if it's annotating wt aa's correctly.
#   charge = count(df,.(WT_AA,wt.charge)),
#   polarity = count(df,.(WT_AA,wt.polarity)),
#   pv = count(df,.(WT_AA,wt.pv)),
#   #check if it's annotating mut aa's correctly.  since wt-aa are correct, can compare count dfs without freq if identical.
#   mut.charge = count(df,.(WT_AA,wt.charge)) %>% subset(select = -c(freq)) %>% rename(c("WT_AA"="AA_Mut","wt.charge" ="mut.charge")) %>% identical(count(df,.(AA_Mut,mut.charge)) %>% subset(select = -c(freq))),
#   
#   mut.pol = count(df,.(WT_AA,wt.polarity)) %>% subset(select = -c(freq)) %>% rename(c("WT_AA"="AA_Mut","wt.polarity" ="mut.polarity")) %>% identical(count(df,.(AA_Mut,mut.polarity)) %>% subset(select = -c(freq))),
#   
#   mut.pv = count(df,.(WT_AA,wt.pv)) %>% subset(select = -c(freq)) %>% rename(c("WT_AA"="AA_Mut","wt.pv" ="mut.pv")) %>% identical(count(df,.(AA_Mut,mut.pv)) %>% subset(select = -c(freq)))
#   ##they're all correct.
#   )
# }
# 
# np.ch.annot = ch.bl.annot(np.annot)
# 
# ch.bl.annot(np2.annot) #everything checks out fine, except I didn't check by hand whether WT-AA categories were annotated correctly, although category annotations for the same mutant AAs as the wt AAs are the same. (mut.charge etc. ones). However, since I checked np.annot by hand, can compare if other ones are idenetical to it.
# 
# comp.np = function(df){
#   list(
#     charge = np.ch.annot$charge %>% subset(select = -c(freq)) %>%
#     identical(ch.bl.annot(df)$charge %>% subset(select = -c(freq))),
#     
#     polarity = np.ch.annot$polarity %>% subset(select = -c(freq)) %>%
#     identical(ch.bl.annot(df)$polarity %>% subset(select = -c(freq))),
#     
#     np.ch.annot$pv %>% subset(select = -c(freq)) %>%
#     identical(ch.bl.annot(df)$pv %>% subset(select = -c(freq)))
#   )
# }
# 
# comp.np(np2.annot)
# 
# ch.bl.annot(env.annot)
# comp.np(env.annot)
# 
# ch.bl.annot(ha.annot)
# comp.np(ha.annot)
#all the other dataframes check out as well.


#end----
```


Now parse lauring data.
```{r PARSE_LAURING}

rm_completely = function(data_frame,row_or_column){
  empty_cells = data.frame(lapply(data_frame,function(cell){
    grepl("^[[:space:]]*$",cell) | is.na(cell) | is.null(cell)
    }))
  if(row_or_column == "rows"){
    data_frame[!apply(empty_cells, 1, all),] #remove rows where all cells are empty, spaces, NA, or NULL
  } else {
    data_frame[,!apply(empty_cells, 2, all)] #remove columns where all cells are empty, spaces, NA, or NULL
  }
}

viable_muts = subset(viable_muts,select=-Segment) #take out segment column which has empty values, Clone.ID has segement info anyway
viable_muts = rm_completely(viable_muts,"cols")
viable_muts = rm_completely(viable_muts,"rows")

lethal_muts = subset(lethal_muts,select=-Segment)
lethal_muts = rm_completely(lethal_muts,"rows")
lethal_muts = rm_completely(lethal_muts,"cols")


#make lethal muts dataframe same as viable muts and then bind the two
lethal_muts$Fitness.Mean = rep('0',nrow(lethal_muts))
lethal_muts$Fitness.SD = rep(NA,nrow(lethal_muts))
colnames(lethal_muts)[2] = 'Amino.Acid.Change'

all_muts = rbind(viable_muts,lethal_muts)

#"Mutation" has both mutation class and position info, parse that out into two different columns
all_muts = mutate(all_muts,Position = gsub("[A-Z]","",Mutation),Mutation = gsub("[0-9]","",Mutation)) #Mutate is awesome! I would normally have used an sapply with gsub, much more complicated.
all_muts = mutate(all_muts,Position = gsub(" ","",Position),Mutation = gsub(" ","",Mutation)) #Remove some weird spaces


#Note whether TS or TV
ts = c("AG","GA","CU","UC","TC","CT") 

all_muts = mutate(all_muts,TS_TV = ifelse(Mutation %in% ts,"TS","TV"))
                  
#Note whether S or NS (synonymous or non-syn)
s = c("SYN", "noncoding", "Non-Coding")
all_muts = mutate(all_muts, S_NS = ifelse(grepl(paste(s,collapse = "|"),Amino.Acid.Change),"S","NS"))

#Parse out segment data from clone.ID
all_muts = mutate(all_muts, Segment = gsub("-[0-9].*$","",Clone.ID))
```

Now parse all other viruses, calculate coefficient of variation and combine all data together.
```{r PARSE_OTHER}
#initial parse----
colnames(hiv_in) = c("AA_Sub","Mutation","Position_in_Codon","Fitness.Mean")
colnames(hiv_ca) = c("AA_Sub","Mutation","Position_in_Codon","Fitness.Mean")
hiv_in = mutate(hiv_in, Gene = "Integrase", Study = "HIV", Dataset = "IN" )
hiv_ca = mutate(hiv_ca, Gene = "Capsid", Study = "HIV", Dataset = "CA")
hiv = rbind(hiv_in, hiv_ca) %>% mutate(Mutation = gsub(" to ","",Mutation),Fitness.Mean = Fitness.Mean/100, Fitness.SD = NA, Position = NA, TS_TV = NA, S_NS = NA, N = NA, Fitness.CV = NA)


colnames(phix174) = c("Mutation", "Gene", "AA_Sub", "Relative_Fitness_Effect", "Other", "SEM")
phix174 = mutate(phix174, N = 3, Study = "phix174_Domingo_Calap_09" )

colnames(qb) = c("Mutation", "Gene", "AA_Sub", "Relative_Fitness_Effect", "Other", "SEM")
qb = mutate(qb, N = 3, Study = "qb_Domingo_Calap_09" )

colnames(tev) = c("Mutant.ID","Gene","Location","Mutation","AA_Sub","Polarity_Change","Relative_Fitness_SEM")
tev = ddply(tev,.(Location,Mutation), mutate, Mutation = gsub(" \\? ",Location,Mutation)) %>% 
  mutate(tev, Fitness.Mean = gsub(" ±.*","",Relative_Fitness_SEM),
         SEM = gsub(".*± ","",Relative_Fitness_SEM) %>% gsub("\\*+","",.),
         Study = "TEV_Carrasco_07_sub",
         N = 5)

colnames(tev_full) = c("Mutant.ID","Gene","Location","Mutation","AA_Sub","Polarity_Change","Relative_Fitness_CI")
tev_full = ddply(tev_full,.(Location,Mutation), mutate, Mutation = gsub("\\?",Location,Mutation)) %>% 
  mutate(tev_full, Fitness.Mean = gsub(" .*","",Relative_Fitness_CI),
         CI95 = gsub("^.* \\(","\\(",Relative_Fitness_CI) %>% gsub("\\*+","",.) %>% gsub("^0$","NA",.),
         Study = "TEV_Carrasco_07_full",
         N = 5)

vsv_genes = grep("[A-Z].*",vsv$Gene)
vsv$Gene[vsv_genes[1]:vsv_genes[2]-1] = vsv$Gene[vsv_genes[1]]
vsv$Gene[vsv_genes[2]:vsv_genes[3]-1] = vsv$Gene[vsv_genes[2]]
vsv$Gene[vsv_genes[3]:vsv_genes[4]-1] = vsv$Gene[vsv_genes[3]]
vsv$Gene[vsv_genes[4]:vsv_genes[5]-1] = vsv$Gene[vsv_genes[4]]
vsv$Gene[vsv_genes[5]:vsv_genes[6]-1] = vsv$Gene[vsv_genes[5]]
vsv$Gene[vsv_genes[6]:length(vsv$Gene)] = vsv$Gene[vsv_genes[6]] #was getting really annoying to do this with an sapply...
colnames(vsv) = c("Gene","Mutation","AA_Sub","Random_or_Dataset","Relative_Fitness_SEM")
vsv = mutate(vsv,N = 5, Study = "vsv_Sanjuan_04",
       SEM = gsub(".*± ","",Relative_Fitness_SEM) %>% gsub("\\*+","",.) %>% gsub("^0$","NA",.), #? need to change is changing all 0s.
       Fitness.Mean = gsub(" ±.*","",Relative_Fitness_SEM)
       )


vsv_genes = grep("[A-Z].*",vsv_full$Gene)
vsv_full$Gene[vsv_genes[1]:vsv_genes[2]-1] = vsv_full$Gene[vsv_genes[1]]
vsv_full$Gene[vsv_genes[2]:vsv_genes[3]-1] = vsv_full$Gene[vsv_genes[2]]
vsv_full$Gene[vsv_genes[3]:vsv_genes[4]-1] = vsv_full$Gene[vsv_genes[3]]
vsv_full$Gene[vsv_genes[4]:vsv_genes[5]-1] = vsv_full$Gene[vsv_genes[4]]
vsv_full$Gene[vsv_genes[5]:vsv_genes[6]-1] = vsv_full$Gene[vsv_genes[5]]
vsv_full$Gene[vsv_genes[6]:length(vsv_full$Gene)] = vsv_full$Gene[vsv_genes[6]] #was getting really annoying to do this with an sapply...
colnames(vsv_full) = c("Gene","Mutation","AA_Sub","Random_or_Dataset","Relative_Fitness_SEM")
vsv_full = mutate(vsv_full,N = 5, Study = "VSV",
       SEM = gsub(".*± ","",Relative_Fitness_SEM) %>% gsub("\\*+","",.) %>% gsub("^0$","NA",.), 
       Fitness.Mean = gsub(" ±.*","",Relative_Fitness_SEM)
       )
vsv_full[dim(vsv_full)[1],c("SEM","Fitness.Mean")] = c(NA,0) #for some reason the data for this mutation was lacking in the new table provided by sanjuan, but from the old supplemental table in the paper, i can get the info.


colnames(f1_1) = c("Mutation","Gene","AA_Sub","Relative_Fitness_SEM")
colnames(f1_2) = colnames(f1_1)
colnames(f1_3) = colnames(f1_1)
f1 = rbind(f1_1,f1_2,f1_3)
f1 = mutate(f1, Relative_Fitness_Effect = gsub(" ±.*","",Relative_Fitness_SEM),
            SEM = gsub(".*± ","",Relative_Fitness_SEM) %>% gsub("\\*+","",.) %>% gsub("-1","NA",.),
            N = sapply(as.numeric(Relative_Fitness_Effect),function(f){
              if(f>0){
                N = 6
              } else{
                N = 3
              }
              N
            }),
            Study = "f1_Peris_10" )

#calc RF, SD, & CV----
rfe_to_rf = function(data_frame){
  mutate(data_frame,Fitness.Mean = as.numeric(Relative_Fitness_Effect)+1)
}
phix174 = rfe_to_rf(phix174)
qb = rfe_to_rf(qb)
f1 = rfe_to_rf(f1)


SEM_to_SD_CV = function(data_frame){
  mutate(data_frame, Fitness.SD = as.numeric(SEM)*sqrt(N), Fitness.CV = Fitness.SD/as.numeric(Fitness.Mean))
}

SD_to_CV = function(data_frame){
  mutate(data_frame, Fitness.CV = as.numeric(Fitness.SD)/as.numeric(Fitness.Mean))
}

phix174 = SEM_to_SD_CV(phix174)
qb = SEM_to_SD_CV(qb)
vsv = SEM_to_SD_CV(vsv)
vsv_full = SEM_to_SD_CV(vsv_full)
tev = SEM_to_SD_CV(tev)
f1 = SEM_to_SD_CV(f1)

all_muts = SD_to_CV(all_muts) %>% mutate(Study = "Lauring")

#combine studies----
common_subset = function(data_frame){
  subset(data_frame,select = c(Gene,Mutation,AA_Sub,N,Fitness.Mean,Fitness.SD,Fitness.CV,Study)) %>% mutate(Position = NA, TS_TV = NA, S_NS = NA, Dataset = NA) 
}

vsv = mutate(vsv, Dataset = ifelse(Random_or_Dataset %in% "Yes", "Random","Described"))

vsv_full = mutate(vsv_full, Dataset = ifelse(Random_or_Dataset %in% "Yes", "Random","Described"))

common_subset_vsv = function(data_frame){
  subset(data_frame,select = c(Gene,Mutation,AA_Sub,N,Fitness.Mean,Fitness.SD,Fitness.CV,Study,Dataset)) %>% mutate(Position = NA, TS_TV = NA, S_NS = NA) 
}

tev_full = mutate(tev_full, Fitness.SD = NA, Fitness.CV = NA )

all_studies = rbind(subset(all_muts %>%
                                    rename(c("Segment" = "Gene", "Amino.Acid.Change" = "AA_Sub")) %>%
                                    mutate(N = 3), select = -c(Clone.ID)),
                    common_subset(phix174),common_subset(qb),common_subset(tev),common_subset(tev_full),common_subset(f1), common_subset_vsv(vsv_full))

#note TS/TV & S/NS----
ts = c("AG","GA","CU","UC","CT","TC") 
s_m = c("SYN","None","None/none","Synonymous","-")
nc_m= c("noncoding", "Non-Coding","Intergenic")

hiv %<>% mutate(Mutation = gsub(" ","",Mutation)) %>% mutate(TS_TV = ifelse(Mutation %in% ts, "TS", "TV"), S_NS = "NS") #there were no synonymous changes

all_studies = mutate(all_studies,Position = gsub("[A-Z]","",Mutation),Mutation = gsub("[0-9]","",Mutation))

all_studies = mutate(all_studies,TS_TV = ifelse(Mutation %in% ts,"TS",ifelse(is.na(Mutation),"NA","TV")))


all_studies = mutate(all_studies, S_NS = ifelse(grepl(paste(s_m,collapse = "|"),AA_Sub) & Gene != "Intergenic","S",ifelse(grepl(paste(nc_m,collapse = "|"),AA_Sub) | Gene == "Intergenic",NA,"NS")))
all_studies[grep("Lys436Arg/Lys264Arg/None",all_studies$AA_Sub),"S_NS"] = "NS" #change the one multi-reading frame that had an S in one but not all reading frames, to NS
row.names(all_studies) = 1:dim(all_studies)[1]


#N for studies----
n_fnx = function(study, nst, stops = "stop|STOP|Stop", xDataset = "xxx"){
  if(xDataset != "xxx"){
    s = subset(all_studies, Study == study & Dataset != xDataset)
  } else{
    s = subset(all_studies, Study == study)
    }
  n = data.frame(row.names = study,
                 Nst = nst,
                 N_tot = dim(s)[1],
                 N_ns = sum(with(s, S_NS == "NS" & !is.na(S_NS))),
                 xBeneficial = sum(with(s, S_NS == "NS" & !is.na(S_NS) & Fitness.Mean < 1)),
                 xLethal = sum(with(s, S_NS == "NS" & !is.na(S_NS) & Fitness.Mean > 0)),
                 xStop = sum(with(s, S_NS == "NS" & !is.na(S_NS) & !grepl(stops,AA_Sub)))
                 )
}


n_phix174 = n_fnx("phix174_Domingo_Calap_09",nst = 24)
n_qb = n_fnx("qb_Domingo_Calap_09",nst = 32)
n_vsv = n_fnx("vsv_Sanjuan_04",nst = 32, xDataset = "Described")
n_tev_sub = n_fnx("TEV_Carrasco_07_sub",nst = 52)
n_tev_full = n_fnx("TEV_Carrasco_07_full",nst = 52)
n_f1 = n_fnx("f1_Peris_10",nst = 51)

n_comp = rbind(n_phix174,n_qb,n_vsv,n_f1,n_tev_full,n_tev_sub)


##lauring N
Lauring = subset(all_studies,Study == "Lauring")
n_comp = data.frame(Nst = rep(NA,4),
                 N_tot = c(dim(Lauring)[1],
                         sum(with(Lauring,Dataset == "Genomewide"),na.rm = T),
                         sum(with(Lauring,!grepl("HA|NA",Gene)),na.rm = T),
                         sum(with(Lauring,grepl("HA|NA", Gene)),na.rm = T)
                         ),
                 N_ns = c(sum(with(Lauring, S_NS == "NS"),na.rm = T),
                         sum(with(Lauring,S_NS == "NS" & Dataset == "Genomewide"),na.rm = T),
                         sum(with(Lauring,S_NS == "NS" & !grepl("HA|NA",Gene)),na.rm = T),
                         sum(with(Lauring,S_NS == "NS" & grepl("HA|NA", Gene)),na.rm = T)
                         ),
                 N_nsBf = c(sum(with(Lauring, S_NS == "NS" & Fitness.Mean>1),na.rm = T),
                         sum(with(Lauring,S_NS == "NS" & Dataset == "Genomewide" & Fitness.Mean>1),na.rm = T),
                         sum(with(Lauring,S_NS == "NS" & !grepl("HA|NA",Gene) & Fitness.Mean>1),na.rm = T),
                         sum(with(Lauring,S_NS == "NS" & grepl("HA|NA", Gene) & Fitness.Mean>1),na.rm = T)
                         ),
                 N_nsLeth = c(sum(with(Lauring, S_NS == "NS" & Fitness.Mean == 0),na.rm = T),
                         sum(with(Lauring,S_NS == "NS" & Dataset == "Genomewide" & Fitness.Mean == 0),na.rm = T),
                         sum(with(Lauring,S_NS == "NS" & !grepl("HA|NA",Gene) & Fitness.Mean == 0),na.rm = T),
                         sum(with(Lauring,S_NS == "NS" & grepl("HA|NA", Gene) & Fitness.Mean == 0),na.rm = T)
                         ),
                 N_nsStop = c(sum(with(Lauring, S_NS == "NS" & grepl("stop",AA_Sub)),na.rm = T),
                         sum(with(Lauring,S_NS == "NS" & Dataset == "Genomewide" & grepl("stop",AA_Sub)),na.rm = T),
                         sum(with(Lauring,S_NS == "NS" & !grepl("HA|NA",Gene) & grepl("stop",AA_Sub)),na.rm = T),
                         sum(with(Lauring,S_NS == "NS" & grepl("HA|NA", Gene) & grepl("stop",AA_Sub)),na.rm = T)
                         ),
                 row.names = c("L_all","L_gw","L_xHAxNA","L_HANA")
                 ) %>% 
  mutate(xBeneficial = N_ns-N_nsBf, xLethal = N_ns - N_nsLeth, xStop = N_ns - N_nsStop) %>% 
  subset(select = c(N_tot,Nst,N_ns,xBeneficial,xLethal,xStop)) %>%
  rbind(n_comp,.)


all_studies = rbind(all_studies,hiv %>% subset(select = -c(Position_in_Codon)))

all_studies$Study = factor(all_studies$Study, levels =c("Lauring","f1_Peris_10","TEV_Carrasco_07_full","TEV_Carrasco_07_sub","qb_Domingo_Calap_09","phix174_Domingo_Calap_09", "VSV","HIV") )


#add in bloom----
prep.bloom = function(df,dtset,study,gene){
  df %>% 
    subset(TS_TV != "MUT23" & TS_TV != "BOTH") %>% 
    rename(c(Rel_site_pref = "Fitness.Mean",X.SITE = "Position")) %>%
    mutate(Mutation = NA,
           AA_Sub = paste0(WT_AA,Position,AA_Mut),
           Fitness.SD = NA,
           Dataset = dtset,
           TS_TV = mapvalues(TS_TV,c("TRUE","FALSE","S"),c("TS","TV","S")), #it actually works with T, F as logical  only and without "S" but I think it's probably better to be specific.
           S_NS = mapvalues(S_NS,c(T,F),c("S","NS")),
           Gene = gene,
           Fitness.CV = NA,
           Study = study,
           N = NA
           ) %>%
    subset(select = c(Mutation,AA_Sub,Fitness.Mean,Fitness.SD,Dataset,Position,TS_TV,S_NS,Gene,Fitness.CV,Study,N))
}

np.as = prep.bloom(np.annot,"NP","Bloom","NP")
np2.as = prep.bloom(np2.annot,"NP2","Bloom","NP")
ha.as = prep.bloom(ha.annot,"HA","Bloom","HA")
env.as = prep.bloom(env.annot,"env","Bloom_env","env")

all_studies %<>% rbind(np.as,np2.as,ha.as,env.as)

# count(np2.as,.(TS_TV,S_NS)) #C check it's only getting non-both non-mut23,and all s are s.
#C also check M1K mutation in np2 can only be caused by two TV, this is annotated correctly as TV here.

#mk fitness means numerical now, may cause problems later on.
all_studies$Fitness.Mean = all_studies$Fitness.Mean %>% as.numeric

#end----
```

Notes on CV analysis:

TEV: I only have the sem for 24 of the full number of mutations, because for the full mutations, they only report a confidence interval from a bootstrap statistical method...

N replicates for each obtained from papers:
Relative Fitness (RF) or Relative Fitness Effect (which is relative fitness-1)(RFE) obtained from papers:
Do they report SEM or SD?:

Study|N replicates|RF or RFE(RF-1)|SEM or SD
-|-|-|-
Lauring | 3| RF | SD
phix174 | 3 | RFE | SEM
qb | 3 | RFE | SEM
vsv | 5* | RF | SEM
tev | 5 | RF | SEM
f1 | 3, but-6 for wt | RFE | SEM

*5 for mutants, but may be only 3 for the progenitor of the mutants they were comparing fitness to, thus N would be...3. Unclear from paper.
##AUC-All and Viable Detrimental Missense Mutations
Calculate AUC for each individual dataset, using all of the detrimental mutations and only the viable detrimental missense mutations. Below also calculates AUC for all mutations including beneficial ones, all viable mutations including beneficial ones, and each of these subsets with and without nonsense mutations, which are not discussed in paper.
```{r AUC.AllViable}
#stat fnxs----
sub  = function(studies, Set = "all", Stops = "INCLUDE", xGenes = c("NONE"), xDatasets = c("NONE"), only_Genes = c("NONE"),data_frame = all_studies){
  #Set sets and stops options
  fit_cut_above = "NONE"
  fit_cut_below = "NONE"
  xAA_Sub = "xxx"
    
  if(Stops != "INCLUDE"){
    xAA_Sub = Stops
  }
  if(Set == "viable" | Set == "viable_xSTOP"){
    data_frame %<>% subset(Fitness.Mean>0)
  }
  if(Set == "detrimental & viable" | Set == "detrimental & viable_xSTOP"){
    data_frame %<>% subset(Fitness.Mean>0 & Fitness.Mean<=1)
  }
  if(Set == "detrimental" | Set == "detrimental_xSTOP"){
    data_frame %<>% subset(Fitness.Mean<=1)
  }
  
  if(Set == "beneficial" | Set == "beneficial_xSTOP"){
    data_frame %<>% subset(Fitness.Mean>1)
  }
  
  if(only_Genes != c("NONE")){
    s = subset(data_frame,grepl(paste(studies,collapse = "|"),Study) & S_NS == "NS" & !is.na(S_NS) & !grepl(paste(xGenes,collapse = "|"),Gene) & !grepl(paste(xDatasets,collapse = "|"),Dataset) & grepl(paste(only_Genes,collapse = "|"),Gene) & !grepl(xAA_Sub,AA_Sub))
  } else{
    s = subset(data_frame,grepl(paste(studies,collapse = "|"),Study) & S_NS == "NS" & !is.na(S_NS) & !grepl(paste(xGenes,collapse = "|"),Gene) & !grepl(paste(xDatasets,collapse = "|"),Dataset) & !grepl(xAA_Sub,AA_Sub))
  }
  if(fit_cut_above != "NONE"){
    s = subset(s, Fitness.Mean > fit_cut_above)
  }
  if(fit_cut_below != "NONE"){
    s = subset(s, Fitness.Mean < fit_cut_below)
  }
  s
}

p_rank = function(data_frame){
  Percentile = rank(data_frame$Fitness.Mean, na.last="keep")/sum(!is.na(data_frame$Fitness.Mean))
  d = cbind(data_frame,Percentile)
  d
}

stat2auc = function(x,y,stat){
  pairs = x*y
  auc = (pairs - as.numeric(stat))/pairs
  auc
}

rank2auc = function(data_frame){
  TS_id = grepl("TS",data_frame$TS_TV)
  TV_id = grepl("TV",data_frame$TS_TV)
  stats = wilcox.test(data_frame$Percentile[TV_id], data_frame$Percentile[TS_id], alternative = "less", conf.int = F) %>% unlist() %>% as.matrix() %>% t() %>% as.data.frame() %>% mutate(N_TS = sum(TS_id), N_TV = sum(TV_id), auc = stat2auc(N_TS,N_TV,as.vector(statistic.W)))
  stats
}

mean2auc = function(data_frame){
  TS_id = grepl("TS",data_frame$TS_TV)
  TV_id = grepl("TV",data_frame$TS_TV)
  stats = wilcox.test(data_frame$Fitness.Mean[TV_id] %>% as.numeric, data_frame$Fitness.Mean[TS_id] %>% as.numeric, alternative = "less", conf.int = F) %>% unlist() %>% as.matrix() %>% t() %>% as.data.frame() %>% mutate(N_TS = sum(TS_id), N_TV = sum(TV_id), auc = stat2auc(N_TS,N_TV,as.vector(statistic.W)))
  stats
}

#pipe fnxs----
study2auc = function(studies, Set = "all", Stops = "INCLUDE", xGenes = c("NONE"), xDatasets = c("NONE"), only_Genes = c("NONE"),data_frame = all_studies, study_name = studies){
  sub(studies, Set, Stops, xGenes, xDatasets, only_Genes,data_frame) %>% mean2auc() %>% mutate(Study = study_name, set = Set)
}

bfsets = c("beneficial","beneficial_xSTOP")

sets4all = c("all","all_xSTOP","detrimental",'detrimental_xSTOP',"viable",'viable_xSTOP',"detrimental & viable",'detrimental & viable_xSTOP')

all_studs = c("Lauring","phix174_Domingo_Calap_09","qb_Domingo_Calap_09","TEV_Carrasco_07_full","f1_Peris_10","HIV","Bloom","Bloom_env")
other_studs = c("phix174_Domingo_Calap_09","qb_Domingo_Calap_09","TEV_Carrasco_07_full","f1_Peris_10","HIV","Bloom", "Bloom_env")

c_sets2auc = function(group = all_studs, group_name = "all_combined", xGenes = c("NONE"), xDatasets = c("NONE"), only_Genes = c("NONE"),data_frame = all_studies,sets = sets4all){
  l = lapply(sets, function(set){
    d = sapply(group,function(s){
      if(grepl("STOP",set)){
        if(grepl("TEV|Lauring",s)){
          study_stops = "stop"
        }
        else if(grepl("VSV",s)){
          study_stops = "Stop"
        } 
        else {
          study_stops = "STOP"
        }
      }
      else {
        study_stops = "INCLUDE"
      }
      sub(s, Set = set, Stops = study_stops, xGenes, xDatasets, only_Genes,data_frame) %>% 
        p_rank() %>%
        subset(select = c(TS_TV, Percentile))
    })
    d= data.frame(TS_TV = d["TS_TV",] %>% unlist %>% unname, Percentile = d["Percentile",] %>% unlist %>% unname)
    rank2auc(d) %>% mutate(Study = group_name, set = set) %>% subset(select = c(Study, set, N_TS, N_TV, p.value, auc))
  })
  do.call(rbind,l)
}

sets2auc = function(studies, xGenes = c("NONE"), xDatasets = c("NONE"), only_Genes = c("NONE"),data_frame = all_studies,
                    study_name = studies,#change accordingly if doing particular genes or datasets to name study (eg Lauring_gw)
                    sets = sets4all
                    ){
  l = lapply(sets,function(set){
    if(grepl("STOP",set)){
      if(grepl("TEV|Lauring",studies)){
        study_stops = "stop"
      }
      else if(grepl("VSV",studies)){
          study_stops = "Stop"
        } 
      else {
        study_stops = "STOP"
      }
    }
    else {
      study_stops = "INCLUDE"
    }
    study2auc(studies = studies,Set = set, Stops = study_stops, xGenes, xDatasets, only_Genes, data_frame, study_name)
  })  
  do.call(rbind,l)
}

e_sets2auc = function(studies = all_studs, xGenes = c("NONE"), xDatasets = c("NONE"), only_Genes = c("NONE"),data_frame = all_studies){
  l = lapply(studies, function(s){
    sets2auc(studies = s, xGenes, xDatasets, only_Genes, data_frame, study_name = s)
  })
  do.call(rbind,l)
}


#calc auc: each----
each = rbind(e_sets2auc(),
             sets2auc("VSV",xDatasets = "Described"),
             sets2auc("HIV",xGenes = c("Capsid"),study_name = "HIV_IN"),
             sets2auc("HIV",xGenes = c("Integrase"),study_name = "HIV_CA"),
             sets2auc("Lauring",xGenes = c("HA","NA"),study_name = "Lauring_xHAxNA"),
             sets2auc("Lauring",only_Genes = c("HA","NA"),study_name = "Lauring_HANA"),
             sets2auc("Lauring",xDatasets = "HA/NA", study_name = "Lauring_gw"),
             sets2auc("Bloom",only_Genes = c("NP"), xDatasets = "NP2", study_name = "Bloom_NP",sets = c(sets4all,bfsets)),
             sets2auc("Bloom",only_Genes = c("NP"), study_name = "Bloom_NP2", data_frame = all_studies %>% subset(Dataset == "NP2"),sets = c(sets4all,bfsets)),
             sets2auc("Bloom",only_Genes = c("HA"),study_name = "Bloom_HA",sets = c(sets4all,bfsets)),
             sets2auc("Bloom",only_Genes = c("NP","HA"), study_name = "Bloom_Flu",sets = c(sets4all,bfsets)),
             sets2auc("Bloom_env",only_Genes = 'env', study_name = "Bloom_env",sets = c(bfsets))
             ) %>% subset(Study != "Bloom" & Study != "Bloom_xSTOP") #for some reason automated auc fnx was combining NP, HA, and env into "Bloom" study, so have to take those out.



xcols_wilcox = function(data_frame){
  colnames(data_frame)[grep("null.value.location shift",colnames(data_frame))] = "x"
  data_frame = data_frame %>% subset(select = -c(method, data.name,x))
  data_frame
}

each = xcols_wilcox(each)

each$set = factor(each$set,
                  levels = c("all","all_xSTOP","viable","viable_xSTOP","detrimental","detrimental_xSTOP","detrimental & viable","detrimental & viable_xSTOP","beneficial","beneficial_xSTOP"))

each$Study = factor(each$Study,
                    levels = rev(c("Lauring","Lauring_gw","Lauring_HANA","Lauring_xHAxNA","Bloom_Flu","Bloom_HA","Bloom_NP2","Bloom_NP","f1_Peris_10","TEV_Carrasco_07_full","VSV","qb_Domingo_Calap_09","phix174_Domingo_Calap_09","HIV","HIV_IN","HIV_CA","Bloom_env")))

```


##AUC-Increasing Thresholds
Calculate AUC for TS-TV fitness differences for missense mutations equal to or greater than a fitness level threshold.  
Do for only detrimental or neutral mutations, with last threshold being .9.
```{r INC.THRESHOLDS2AUC}
#stat fnxs----

all.studies.det = all_studies %>% subset(Fitness.Mean<=1)

sub  = function(studies, Set = "all", Stops = "INCLUDE", xGenes = c("NONE"), xDatasets = c("NONE"), only_Genes = c("NONE"),data_frame = all_studies){
  #Set sets and stops options
  fit_cut_above = "NONE"
  fit_cut_below = "NONE"
  xAA_Sub = "xxx"
    
  if(Stops != "INCLUDE"){
    xAA_Sub = Stops
  }
  if(Set == "viable" | Set == "viable_xSTOP"){
    data_frame %<>% subset(Fitness.Mean>0)
  }
  if(Set == "detrimental & viable" | Set == "detrimental & viable_xSTOP"){
    data_frame %<>% subset(Fitness.Mean>0 & Fitness.Mean<=1)
  }
  if(Set == "detrimental" | Set == "detrimental_xSTOP"){
    data_frame %<>% subset(Fitness.Mean<=1)
  }
  
  if(Set == "beneficial" | Set == "beneficial_xSTOP"){
    data_frame %<>% subset(Fitness.Mean>1)
  }
  
  if(only_Genes != c("NONE")){
    s = subset(data_frame,grepl(paste(studies,collapse = "|"),Study) & S_NS == "NS" & !is.na(S_NS) & !grepl(paste(xGenes,collapse = "|"),Gene) & !grepl(paste(xDatasets,collapse = "|"),Dataset) & grepl(paste(only_Genes,collapse = "|"),Gene) & !grepl(xAA_Sub,AA_Sub))
  } else{
    s = subset(data_frame,grepl(paste(studies,collapse = "|"),Study) & S_NS == "NS" & !is.na(S_NS) & !grepl(paste(xGenes,collapse = "|"),Gene) & !grepl(paste(xDatasets,collapse = "|"),Dataset) & !grepl(xAA_Sub,AA_Sub))
  }
  if(fit_cut_above != "NONE"){
    s = subset(s, Fitness.Mean > fit_cut_above)
  }
  if(fit_cut_below != "NONE"){
    s = subset(s, Fitness.Mean < fit_cut_below)
  }
  s
} #subset based on a threshold

stat2auc = function(x,y,stat){
  pairs = x*y
  auc = (pairs - as.numeric(stat))/pairs
  auc
} #calc auc from mwu statistic

mean2auc = function(data_frame){
  TS_id = grepl("TS",data_frame$TS_TV)
  TV_id = grepl("TV",data_frame$TS_TV)
  stats = wilcox.test(data_frame$Fitness.Mean[TV_id] %>% as.numeric, data_frame$Fitness.Mean[TS_id] %>% as.numeric, alternative = "less", conf.int = F) %>% unlist() %>% as.matrix() %>% t() %>% as.data.frame() %>% mutate(N_TS = sum(TS_id), N_TV = sum(TV_id), auc = stat2auc(N_TS,N_TV,as.vector(statistic.W)))
  stats
} #calc mwu statistic

max_fit = function(studies="ALL", Set = "all", Stops = "INCLUDE", xGenes = c("NONE"), xDatasets = c("NONE"), only_Genes = c("NONE"),data_frame =  all_studies,manual = "none"){
  if(manual == "none"){
    m = ifelse(studies=="ALL",data_frame$Fitness.Mean %>% max(na.rm = T), sub(studies, Set, Stops, xGenes, xDatasets, only_Genes,data_frame) %>% .$Fitness.Mean %>% max(na.rm = T))
  }
  else{
    m = manual
  }
  m
  
} # to use if want to do beneficial mutations as well, and have max set by each study max.

set_max_thresh = max_fit(manual = .9) #if want to pick thresholds by the max in the study or of all studies, set manual  to "none".


pick_thresholds  = function(break_prop=.1,max_thresh = set_max_thresh){
  seq(0,max_thresh,by = break_prop)
}

sub_thresholds =function(studies, threshold = "NONE", Stops = "INCLUDE", xGenes = c("NONE"), xDatasets = c("NONE"), only_Genes = c("NONE"),data_frame =  all.studies.det){
  #Set stops options
  xAA_Sub = "xxx"
    
  if(Stops != "INCLUDE"){
    xAA_Sub = Stops
  }
  
  if(only_Genes != c("NONE")){
    s = subset(data_frame,grepl(paste(studies,collapse = "|"),Study) & S_NS == "NS" & !is.na(S_NS) & !grepl(paste(xGenes,collapse = "|"),Gene) & !grepl(paste(xDatasets,collapse = "|"),Dataset) & grepl(paste(only_Genes,collapse = "|"),Gene) & !grepl(xAA_Sub,AA_Sub))
  } else{
    s = subset(data_frame,grepl(paste(studies,collapse = "|"),Study) & S_NS == "NS" & !is.na(S_NS) & !grepl(paste(xGenes,collapse = "|"),Gene) & !grepl(paste(xDatasets,collapse = "|"),Dataset) & !grepl(xAA_Sub,AA_Sub))
  }
  
  if(threshold != "NONE"){
    s = subset(s, Fitness.Mean >= threshold)
  }
  s
}


mean2auc_thresh = function(data_frame){
  
  TS_id = grepl("TS",data_frame$TS_TV)
  TV_id = grepl("TV",data_frame$TS_TV)
  
  if(sum(TS_id) < 1 | sum(TV_id) < 1){
    stats = matrix(NA,ncol = 11, nrow = 1)
    colnames(stats) = c("statistic.W","p.value","null.value.location shift","alternative","method","data.name","N_TS","N_TV","auc","Study","threshold")
    stats %>% as.data.frame()
  } else{
    stats = wilcox.test(data_frame$Fitness.Mean[TV_id] %>% as.numeric, data_frame$Fitness.Mean[TS_id] %>% as.numeric, alternative = "less", conf.int = F) %>% unlist() %>% as.matrix() %>% t() %>% as.data.frame() %>% mutate(N_TS = sum(TS_id), N_TV = sum(TV_id), auc = stat2auc(N_TS,N_TV,as.vector(statistic.W)))
    stats
  }
}

#pipe fnxs----
study2auc = function(studies, set_break_percent = .1, setmax = set_max_thresh,
                     Stops = "INCLUDE", xGenes = c("NONE"), xDatasets = c("NONE"), only_Genes = c("NONE"),
                     data_frame = all.studies.det,
                     study_name = studies){
  
  l = lapply(pick_thresholds(set_break_percent,setmax),function(threshold){
      sub_thresholds(studies, threshold, Stops, xGenes, xDatasets, only_Genes,data_frame) %>%
      mean2auc_thresh() %>% 
      mutate(Study = ifelse(Stops != "INCLUDE",paste0(study_name,"_xSTOP"),study_name), threshold = threshold)
  })
  do.call(rbind,l)

}

studs4thresh = c("HIV","Lauring")

studs2auc = function(studs = studs4thresh, set_break_per = .1, set_max_th = set_max_thresh,
                     Stops = "INCLUDE", xGenes = c("NONE"), xDatasets = c("NONE"), only_Genes = c("NONE"),
                     data_frame = all.studies.det){
  l = lapply(studs,function(stud){
    study2auc(stud,set_break_per,set_max_th,Stops,xGenes,xDatasets,only_Genes,data_frame,study_name = ifelse(Stops != "INCLUDE",paste0(stud,"_xSTOP"),stud))
  })
  do.call(rbind,l)
}


all_studs = c("Lauring","phix174_Domingo_Calap_09","qb_Domingo_Calap_09","TEV_Carrasco_07_full","f1_Peris_10","HIV","Bloom","Bloom_env")
other_studs = c("phix174_Domingo_Calap_09","qb_Domingo_Calap_09","TEV_Carrasco_07_full","f1_Peris_10","HIV","Bloom", "Bloom_env")

#calc thresholds auc----
thresholds = rbind(studs2auc(),
                   studs2auc(Stops = "stop"),
                   study2auc("HIV",xGenes = c("Capsid"),study_name = "HIV_IN"),
                   study2auc("HIV",xGenes = c("Integrase"),study_name = "HIV_CA"), 
                   study2auc("Lauring",xGenes = c("HA","NA"),study_name = "Lauring_xHAxNA"), 
                   study2auc("Lauring",only_Genes = c("HA","NA"),study_name = "Lauring_HANA"),
                   study2auc("Lauring",xDatasets = "HA/NA", study_name = "Lauring_gw"),
                   study2auc("HIV",xGenes = c("Capsid"),study_name = "HIV_IN",Stops = "stop"),
                   study2auc("HIV",xGenes = c("Integrase"),study_name = "HIV_CA",Stops = "stop"), 
                   study2auc("Lauring",xGenes = c("HA","NA"),study_name = "Lauring_xHAxNA",Stops = "stop"), 
                   study2auc("Lauring",only_Genes = c("HA","NA"),study_name = "Lauring_HANA",Stops = "stop"),
                   study2auc("Lauring",xDatasets = "HA/NA", study_name = "Lauring_gw",Stops = "stop"),
                   study2auc("Bloom_env",only_Genes = "env",study_name = "HIV_ENV", setmax = 1),
                   study2auc("Bloom",only_Genes = "NP",xDatasets = "NP2",study_name = "NP_Aichi"),
                   study2auc("Bloom",only_Genes = "NP",study_name = "NP_PR8",data_frame = all.studies.det %>% subset(Dataset == "NP2")),
                   study2auc("Bloom",only_Genes = "HA",study_name = "HA_WSN33"),
                   study2auc("Bloom",only_Genes = c("NP","HA"),study_name = "Bloom_Flu"),
                   study2auc("Bloom",only_Genes = c("NP","HA"),study_name = "Bloom_H1N1",data_frame = all.studies.det %>% subset(Dataset == "NP2"|Dataset=="HA"))
                   ) %>% subset(!is.na(threshold) & Study != "Bloom" & Study != "Bloom_xSTOP",select = c(p.value, N_TS,N_TV,auc, Study, threshold)) %>% mutate(Study = gsub("xSTOP_xSTOP","xSTOP",Study)) #for some reason automated auc fnx was combining NP, HA, and env into "Bloom" study, so have to take those out.


bloom_flu_studs = c("Bloom_Flu","Bloom_H1N1","HA_WSN33","NP_Aichi","NP_PR8")
bloom_studs = c(bloom_flu_studs,"HIV_ENV")


#end----
```


##AUC-Decreasing Thresholds
Calculate AUC for Ts-Tv fitness differences for mutants equal to or less than a fitness level threshold. Only detrimental or nuetral missense mutations.  
```{r DEC.THRESHOLDS2AUC}
#stat fnxs----

dec.thresholds = seq(.1,1,.1) %>% rev

sub_dec_thresholds =function(studies, dec_threshold = "NONE", Stops = "INCLUDE", xGenes = c("NONE"), xDatasets = c("NONE"), only_Genes = c("NONE"),data_frame =  all.studies.det){
  #Set stops options
  xAA_Sub = "xxx"
    
  if(Stops != "INCLUDE"){
    xAA_Sub = Stops
  }
  
  if(only_Genes != c("NONE")){
    s = subset(data_frame,grepl(paste(studies,collapse = "|"),Study) & S_NS == "NS" & !is.na(S_NS) & !grepl(paste(xGenes,collapse = "|"),Gene) & !grepl(paste(xDatasets,collapse = "|"),Dataset) & grepl(paste(only_Genes,collapse = "|"),Gene) & !grepl(xAA_Sub,AA_Sub))
  } else{
    s = subset(data_frame,grepl(paste(studies,collapse = "|"),Study) & S_NS == "NS" & !is.na(S_NS) & !grepl(paste(xGenes,collapse = "|"),Gene) & !grepl(paste(xDatasets,collapse = "|"),Dataset) & !grepl(xAA_Sub,AA_Sub))
  }
  
  if(dec_threshold != "NONE"){
    s = subset(s, Fitness.Mean <= dec_threshold)
  }
  s
}


mean2auc_thresh = function(data_frame){
  
  TS_id = grepl("TS",data_frame$TS_TV)
  TV_id = grepl("TV",data_frame$TS_TV)
  
  if(sum(TS_id) < 1 | sum(TV_id) < 1){
    stats = matrix(NA,ncol = 11, nrow = 1)
    colnames(stats) = c("statistic.W","p.value","null.value.location shift","alternative","method","data.name","N_TS","N_TV","auc","Study","threshold")
    stats %>% as.data.frame()
  } else{
    stats = wilcox.test(data_frame$Fitness.Mean[TV_id] %>% as.numeric, data_frame$Fitness.Mean[TS_id] %>% as.numeric, alternative = "less", conf.int = F) %>% unlist() %>% as.matrix() %>% t() %>% as.data.frame() %>% mutate(N_TS = sum(TS_id), N_TV = sum(TV_id), auc = stat2auc(N_TS,N_TV,as.vector(statistic.W)))
    stats
  }
}

#pipe fnxs----
study2auc_decth = function(studies, #set_break_percent = .1,
                     Stops = "INCLUDE", xGenes = c("NONE"), xDatasets = c("NONE"), only_Genes = c("NONE"),
                     data_frame =  all.studies.det,
                     study_name = studies){
  # l = lapply(c(pick_thresholds(set_break_percent,setmax),setmax) %>% rev,function(threshold){} # use this instead if want to use whole distibution dec. thresholds and set max threshold by study.
  
  l = lapply(dec.thresholds,function(threshold){
      sub_dec_thresholds(studies, threshold, Stops, xGenes, xDatasets, only_Genes,data_frame) %>%
      mean2auc_thresh() %>% 
      mutate(Study = ifelse(Stops != "INCLUDE",paste0(study_name,"_xSTOP"),study_name), threshold = threshold)
  })
  do.call(rbind,l)

}

studs4thresh = c("HIV","Lauring")

studs2auc_decth = function(studs = studs4thresh, #set_break_per = .1,
                     Stops = "INCLUDE", xGenes = c("NONE"), xDatasets = c("NONE"), only_Genes = c("NONE"),
                     data_frame = all.studies.det){
  l = lapply(studs,function(stud){
    study2auc_decth(stud,Stops,xGenes,xDatasets,only_Genes,data_frame,study_name = ifelse(Stops != "INCLUDE",paste0(stud,"_xSTOP"),stud))
  })
  do.call(rbind,l)
}

#calc thresholds auc----
dec_thresholds = rbind(studs2auc_decth(),
                   studs2auc_decth(Stops = "stop"),
                   study2auc_decth("HIV",xGenes = c("Capsid"),study_name = "HIV_IN"),
                   study2auc_decth("HIV",xGenes = c("Integrase"),study_name = "HIV_CA"), 
                   study2auc_decth("Lauring",xGenes = c("HA","NA"),study_name = "Lauring_xHAxNA"), 
                   study2auc_decth("Lauring",only_Genes = c("HA","NA"),study_name = "Lauring_HANA"),
                   study2auc_decth("Lauring",xDatasets = "HA/NA", study_name = "Lauring_gw"),
                   study2auc_decth("HIV",xGenes = c("Capsid"),study_name = "HIV_IN",Stops = "stop"),
                   study2auc_decth("HIV",xGenes = c("Integrase"),study_name = "HIV_CA",Stops = "stop"), 
                   study2auc_decth("Lauring",xGenes = c("HA","NA"),study_name = "Lauring_xHAxNA",Stops = "stop"), 
                   study2auc_decth("Lauring",only_Genes = c("HA","NA"),study_name = "Lauring_HANA",Stops = "stop"),
                   study2auc_decth("Lauring",xDatasets = "HA/NA", study_name = "Lauring_gw",Stops = "stop"),
                   study2auc_decth("Bloom_env",only_Genes = "env",study_name = "HIV_ENV"),
                   study2auc_decth("Bloom",only_Genes = "NP",xDatasets = "NP2",study_name = "NP_Aichi"),
                   study2auc_decth("Bloom",only_Genes = "NP",study_name = "NP_PR8",data_frame = all.studies.det %>% subset(Dataset == "NP2")),
                   study2auc_decth("Bloom",only_Genes = "HA",study_name = "HA_WSN33"),
                   study2auc_decth("Bloom",only_Genes = c("NP","HA"),study_name = "Bloom_Flu"),
                   study2auc_decth("Bloom",only_Genes = c("NP","HA"),study_name = "Bloom_H1N1",data_frame = all.studies.det %>% subset(Dataset == "NP2"|Dataset=="HA"))
                   ) %>% subset(!is.na(threshold) & Study != "Bloom" & Study != "Bloom_xSTOP",select = c(p.value, N_TS,N_TV,auc, Study, threshold)) %>% mutate(Study = gsub("xSTOP_xSTOP","xSTOP",Study)) #for some reason automated auc fnx was combining NP, HA, and env into "Bloom" study, so have to take those out.

dec_thresholds %<>% subset(!is.nan(p.value %>% as.vector %>% as.numeric))


#end----
```


##Fisher Thresholds
Perform fisher analysis on detrimental or neutral missense mutations. Can use below code to do same thing for beneficial mutations, commented out below. Note that the bloom DMS studies have no mutations annotated as lethal, so cannot do first threshold of 0 for those.
```{r FISHER}
#stat fnxs----
set_windows  = function(width=.1,min=0,max=set_max_thresh, slide_by = 1/2){
  up_end = seq(min+width,max,slide_by*width)
  list(low_end = seq(min,max,slide_by*width)[1:length(up_end)],
       up_end = up_end
       )
} #This creates thresholds.  create sliding windows or bins by setting the size of the width, the min and max the bins will go to, and the amount to slide by as a fraction of the width. slide_by = 1 if don't want sliding windows, just bins. Output is list of lower ends and upper ends for each window.

fisher_mat_thresh.sets =function(studies,
                            set = "ALL",
                            windows = F,low_th = F,up_th =F, #set windows options
                            Stops = "INCLUDE", xGenes = c("NONE"), xDatasets = c("NONE"), only_Genes = c("NONE"),
                            data_frame = all_studies){
  #Set up fisher matrix  
  m = matrix(data = rep(NA,4),
                nrow = 2,
                ncol = 2, 
                byrow = T,
                dimnames = list(c("TV","TS"),c("below_threshold","above_threshold"))
  )
  
  #Find complete sub
  ##Set stops options
  xAA_Sub = "xxx"
  
  if(Stops != "INCLUDE"){
    xAA_Sub = Stops
  }
  
  if(only_Genes != c("NONE")){
    s = subset(data_frame,grepl(paste(studies,collapse = "|"),Study) & S_NS == "NS" & !is.na(S_NS) & !grepl(paste(xGenes,collapse = "|"),Gene) & !grepl(paste(xDatasets,collapse = "|"),Dataset) & grepl(paste(only_Genes,collapse = "|"),Gene) & !grepl(xAA_Sub,AA_Sub))
  } else{
    s = subset(data_frame,grepl(paste(studies,collapse = "|"),Study) & S_NS == "NS" & !is.na(S_NS) & !grepl(paste(xGenes,collapse = "|"),Gene) & !grepl(paste(xDatasets,collapse = "|"),Dataset) & !grepl(xAA_Sub,AA_Sub))
  }
  
  if(set != "ALL"){
    if(set == "DET"){
      s %<>% subset(Fitness.Mean<=1)
    } else if(set == "BF"){
      s %<>% subset(Fitness.Mean>=1)
    }
  }
  
  #find total n ts, tv from complete sub
  N_TV = s %>% .$TS_TV %in% "TV" %>% sum
  N_TS = s %>% .$TS_TV %in% "TS" %>% sum
  
  # #sub with window
  # if(windows){
  #   s = subset(s, Fitness.Mean >= low_th & Fitness.Mean <= up_th)
  # }

  #now fill in fisher matrix
  m["TV","below_threshold"] = s %>% subset(Fitness.Mean <= low_th) %>% .$TS_TV %in% "TV" %>% sum
  m["TS","below_threshold"] = s %>% subset(Fitness.Mean <= low_th) %>% .$TS_TV %in% "TS" %>% sum
  m["TV","above_threshold"] = N_TV - m["TV","below_threshold"]
  m["TS","above_threshold"] = N_TS - m["TS","below_threshold"]
  m
}#gets fisher matrix for below or EQUAL to threshold and strictly above threshold. Based on set: all, detrimental (DET) (0-1, inclusive), or beneficial (BF) (1-max,inclusive).

mat2stat_thresh = function(m,study,low_th,up_th){
  test = fisher.test(m, conf.int = T)
  data.frame(p.value = test$p.value,
             OR = test$estimate,
             OR_CI95_lower = test$conf.int[1],
             OR_CI95_upper = test$conf.int[2],
             total_n = sum(m),
             # total_TV = m["TV",] %>% sum,
             # total_TS = m["TS",] %>% sum,
             below_thresh_TV = m["TV","below_threshold"],
             above_thresh_TV = m["TV","above_threshold"],
             below_thresh_TS = m["TS","below_threshold"],
             above_thresh_TS = m["TS","above_threshold"],
             Study = study,
             threshold = low_th,
             row.names = NULL
  )
}

#pipe fnxs----
study2fisher_thresh.sets = function(studies,
                                    set = "ALL",
                                    width = .05,
                                    min = ifelse(set == "BF",
                                                 1,
                                                 0),
                                    max = ifelse(set == "DET",
                                                 1,
                                                 max_fit(studies)
                                                 ),
                                    slide_by = 1, #windows settings
                                    Stops = "INCLUDE", xGenes = c("NONE"), xDatasets = c("NONE"), only_Genes = c("NONE"),
                                    data_frame = all_studies,
                                    study_name = studies){
  if(studies == "Bloom_env" & set == "BF"){
    max = min(max_fit("Bloom_env"),10)
  } else if(studies == "Bloom" & set == "BF"){
    max = min(max_fit("Bloom",xGenes = "env"),10)
  }
  if(grepl("Bloom",studies) & set == "DET"){
    min = width  
  }
  if(grepl("Bloom",studies) & set == "BF"){
    min = 1+width
  }
  win = set_windows(width,min,max,slide_by)
  lapply(1:length(win$up_end),function(x){
    fisher_mat_thresh.sets(studies,
                          set,
                          windows = T,low_th = win$low_end[x],up_th = win$up_end[x], #set windows options
                          Stops, xGenes, xDatasets, only_Genes,
                          data_frame) %>%
      mat2stat_thresh(m = .,study = ifelse(Stops != "INCLUDE",paste0(study_name,"_xSTOP"),study_name),
                      low_th = win$low_end[x],up_th = win$up_end[x])
  }) %>% do.call(rbind,.)
}

studs2fisher_thresh.sets = function(studs = studs4fisher,
                                    set = "ALL",
                                    width = .05,
                                    min = ifelse(set == "BF",
                                                 1,
                                                 0),
                                    max = ifelse(set == "DET",
                                                 1,
                                                 'SetbyStudy'
                                                 ),
                                    slide_by = 1, #windows settings
                                    Stops = "INCLUDE", xGenes = c("NONE"), xDatasets = c("NONE"), only_Genes = c("NONE"),
                                    data_frame = all_studies,
                                    study_name = studies){
  l = lapply(studs,function(stud){
    study2fisher_thresh.sets(stud,
                             set,
                             width,min,max = ifelse(max == "SetbyStudy",max_fit(studies = stud),max),slide_by,
                             Stops,xGenes,xDatasets,only_Genes,
                             data_frame,
                             study_name = stud)
  })
  do.call(rbind,l)
}

studs4fisher = c("HIV","Lauring")


#calc fisher sets----
##det----
fisher.bloom.det = rbind(
                   study2fisher_thresh.sets("Bloom_env",only_Genes = "env",study_name = "HIV_ENV", set = "DET"), #bloom env
                   study2fisher_thresh.sets("Bloom",only_Genes = "NP",xDatasets = "NP2",study_name = "NP_Aichi", set = "DET"),  #bloom np aichi
                   study2fisher_thresh.sets("Bloom",only_Genes = "NP",study_name = "NP_PR8",data_frame = all_studies %>% subset(Dataset == "NP2"), set = "DET"),  #bloom np pr8
                   study2fisher_thresh.sets("Bloom",only_Genes = "HA",study_name = "HA_WSN33", set = "DET"),   #bloom ha wsn33
                   study2fisher_thresh.sets("Bloom",only_Genes = c("NP","HA"),study_name = "Bloom_Flu", set = "DET"), #bloom ha,np comb
                   study2fisher_thresh.sets("Bloom",only_Genes = c("NP","HA"),study_name = "Bloom_H1N1",data_frame = all_studies %>% subset(Dataset == "NP2"|Dataset=="HA"), set = "DET") #bloom h1n1 comb
                   ) 

fisher.other.det = rbind(
  studs2fisher_thresh.sets(width = .1, set = "DET"), #hiv and lauring
  studs2fisher_thresh.sets(Stops = "stop",width = .1, set = "DET"), #hiv and lauring xSTOPs
  study2fisher_thresh.sets("HIV",xGenes = c("Capsid"),study_name = "HIV_IN",width = .1, set = "DET"), #hiv in
  study2fisher_thresh.sets("HIV",xGenes = c("Integrase"),study_name = "HIV_CA",width = .1, set = "DET"), #hiv ca
  study2fisher_thresh.sets("Lauring",xGenes = c("HA","NA"),study_name = "Lauring_xHAxNA",width = .1, set = "DET"), #la xhaxna
  study2fisher_thresh.sets("Lauring",only_Genes = c("HA","NA"),study_name = "Lauring_HANA",width = .1, set = "DET"), #la xhaxna
  study2fisher_thresh.sets("Lauring",xDatasets = "HA/NA", study_name = "Lauring_gw",width = .1, set = "DET"),#la gw
  study2fisher_thresh.sets("HIV",xGenes = c("Capsid"),study_name = "HIV_IN",Stops = "stop",width = .1, set = "DET"),
  study2fisher_thresh.sets("HIV",xGenes = c("Integrase"),study_name = "HIV_CA",Stops = "stop",width = .1, set = "DET"), 
  study2fisher_thresh.sets("Lauring",xGenes = c("HA","NA"),study_name = "Lauring_xHAxNA",Stops = "stop",width = .1, set = "DET"), 
  study2fisher_thresh.sets("Lauring",only_Genes = c("HA","NA"),study_name = "Lauring_HANA",Stops = "stop",width = .1, set = "DET"),
  study2fisher_thresh.sets("Lauring",xDatasets = "HA/NA", study_name = "Lauring_gw",Stops = "stop",width = .1, set = "DET")
  #study2fisher_thresh.sets("phix174_Domingo_Calap_09",study_name = "phix174",Stops = "STOP",width = .1, set = "DET"),
  #study2fisher_thresh.sets("qb_Domingo_Calap_09",study_name = "qb",Stops = "STOP",width = .1, set = "DET"),
  #study2fisher_thresh.sets("TEV_Carrasco_07_full",study_name = "TEV",Stops = "stop",width = .1, set = "DET"),
  #study2fisher_thresh.sets("f1_Peris_10",study_name = "f1",Stops = "STOP",width = .1, set = "DET"),
  #study2fisher_thresh.sets("VSV",study_name = "VSV",Stops = "Stop",width = .1, set = "DET")
) #not calculating for other studies commented out bc too low N.

##beneficials----
# fisher.bloom.bf = rbind(
#   study2fisher_thresh.sets("Bloom_env",only_Genes = "env",study_name = "HIV_ENV", set = "BF"), #bloom env
#   study2fisher_thresh.sets("Bloom",only_Genes = "NP",xDatasets = "NP2",study_name = "NP_Aichi", set = "BF"),  #bloom np aichi
#   study2fisher_thresh.sets("Bloom",only_Genes = "NP",study_name = "NP_PR8",data_frame = all_studies %>% subset(Dataset == "NP2"), set = "BF"),  #bloom np pr8
#   study2fisher_thresh.sets("Bloom",only_Genes = "HA",study_name = "HA_WSN33",set = "BF"),   #bloom ha wsn33
#   study2fisher_thresh.sets("Bloom",only_Genes = c("NP","HA"),study_name = "Bloom_Flu",set = "BF"), #bloom ha,np comb
#   study2fisher_thresh.sets("Bloom",only_Genes = c("NP","HA"),study_name = "Bloom_H1N1",data_frame = all_studies %>% subset(Dataset == "NP2"|Dataset=="HA"), set = "BF") #bloom h1n1 comb
# ) 
# 
# ###other
# fisher.other.bf = rbind(
#   studs2fisher_thresh.sets(width = .1, set = "BF"), #hiv and lauring
#   studs2fisher_thresh.sets(Stops = "stop",width = .1, set = "BF"), #hiv and lauring xSTOPs
#   study2fisher_thresh.sets("HIV",xGenes = c("Capsid"),study_name = "HIV_IN",width = .1, set = "BF"), #hiv in
#   study2fisher_thresh.sets("HIV",xGenes = c("Integrase"),study_name = "HIV_CA",width = .1, set = "BF"), #hiv ca
#   study2fisher_thresh.sets("Lauring",xGenes = c("HA","NA"),study_name = "Lauring_xHAxNA",width = .1, set = "BF"), #la xhaxna
#   study2fisher_thresh.sets("Lauring",only_Genes = c("HA","NA"),study_name = "Lauring_HANA",width = .1, set = "BF"), #la xhaxna
#   study2fisher_thresh.sets("Lauring",xDatasets = "HA/NA", study_name = "Lauring_gw",width = .1, set = "BF"),#la gw
#   study2fisher_thresh.sets("HIV",xGenes = c("Capsid"),study_name = "HIV_IN",Stops = "stop",width = .1, set = "BF"),
#   study2fisher_thresh.sets("HIV",xGenes = c("Integrase"),study_name = "HIV_CA",Stops = "stop",width = .1, set = "BF"), 
#   study2fisher_thresh.sets("Lauring",xGenes = c("HA","NA"),study_name = "Lauring_xHAxNA",Stops = "stop",width = .1, set = "BF"), 
#   study2fisher_thresh.sets("Lauring",only_Genes = c("HA","NA"),study_name = "Lauring_HANA",Stops = "stop",width = .1, set = "BF"),
#   study2fisher_thresh.sets("Lauring",xDatasets = "HA/NA", study_name = "Lauring_gw",Stops = "stop",width = .1, set = "BF"),
#   # study2fisher_thresh.sets("phix174_Domingo_Calap_09",study_name = "phix174",Stops = "STOP",width = .1, set = "BF"), #get error message "wrong sign in by", it's simply because there are no bf for this study or the study max is less than the slide-by of .1.  Comment out.  Below others commented out for same reason. Yes. For each of the studies here commented out, there are above 1s, but they are very close to 1 and less than 1.1, so can't slide or really test the window of 1-1.1 vs. above.
#   # study2fisher_thresh.sets("qb_Domingo_Calap_09",study_name = "qb",Stops = "STOP",width = .1, set = "BF"),
#   study2fisher_thresh.sets("TEV_Carrasco_07_full",study_name = "TEV",Stops = "stop",width = .1, set = "BF"),
#   # study2fisher_thresh.sets("f1_Peris_10",study_name = "f1",Stops = "STOP",width = .1, set = "BF"),
#   study2fisher_thresh.sets("VSV",study_name = "VSV",Stops = "Stop",width = .1, set = "BF")
# )
# 
# fisher.other.bf.05 = rbind(
#   studs2fisher_thresh.sets(width = .05, set = "BF"), #hiv and lauring
#   studs2fisher_thresh.sets(Stops = "stop",width = .05, set = "BF"), #hiv and lauring xSTOPs
#   study2fisher_thresh.sets("HIV",xGenes = c("Capsid"),study_name = "HIV_IN",width = .05, set = "BF"), #hiv in
#   study2fisher_thresh.sets("HIV",xGenes = c("Integrase"),study_name = "HIV_CA",width = .05, set = "BF"), #hiv ca
#   study2fisher_thresh.sets("Lauring",xGenes = c("HA","NA"),study_name = "Lauring_xHAxNA",width = .05, set = "BF"), #la xhaxna
#   study2fisher_thresh.sets("Lauring",only_Genes = c("HA","NA"),study_name = "Lauring_HANA",width = .05, set = "BF"), #la xhaxna
#   study2fisher_thresh.sets("Lauring",xDatasets = "HA/NA", study_name = "Lauring_gw",width = .05, set = "BF"),#la gw
#   study2fisher_thresh.sets("HIV",xGenes = c("Capsid"),study_name = "HIV_IN",Stops = "stop",width = .05, set = "BF"),
#   study2fisher_thresh.sets("HIV",xGenes = c("Integrase"),study_name = "HIV_CA",Stops = "stop",width = .05, set = "BF"), 
#   study2fisher_thresh.sets("Lauring",xGenes = c("HA","NA"),study_name = "Lauring_xHAxNA",Stops = "stop",width = .05, set = "BF"), 
#   study2fisher_thresh.sets("Lauring",only_Genes = c("HA","NA"),study_name = "Lauring_HANA",Stops = "stop",width = .05, set = "BF"),
#   study2fisher_thresh.sets("Lauring",xDatasets = "HA/NA", study_name = "Lauring_gw",Stops = "stop",width = .05, set = "BF"),
#   study2fisher_thresh.sets("phix174_Domingo_Calap_09",study_name = "phix174",Stops = "STOP",width = .05, set = "BF"), #get error message "wrong sign in by", it's simply because there are no bf for this study or the study max is less than the slide-by of .1.  Comment out.  Below others commented out for same reason. Yes. For each of the studies here commented out, there are above 1s, but they are very close to 1 and less than 1.1, so can't slide or really test the window of 1-1.1 vs. above.
#   # study2fisher_thresh.sets("qb_Domingo_Calap_09",study_name = "qb",Stops = "STOP",width = .05, set = "BF"),
#   study2fisher_thresh.sets("TEV_Carrasco_07_full",study_name = "TEV",Stops = "stop",width = .05, set = "BF"),
#   study2fisher_thresh.sets("f1_Peris_10",study_name = "f1",Stops = "STOP",width = .05, set = "BF"),
#   study2fisher_thresh.sets("VSV",study_name = "VSV",Stops = "Stop",width = .05, set = "BF")
# )


 
# #C 17_01_23
#  I checked whether total numbers of TS and TV used for this analysis matches the total number of TS and tv in the .annot
# 
# c.n = function(stud, df){
#   list(fish.total = fisher_thresh %>% subset(Study == stud) %>% .$total_n %>% unique, #total ts and tv exclduing both in fisher analysis
#               annot.total = df %>% subset(TS_TV == "FALSE" | TS_TV == "TRUE") %>% count(.(TS_TV)) %>% .$freq %>% sum, #total ts and tv excluding both in annot
#        fish.indv = fisher_thresh %>% subset(Study == stud & threshold == .05, select = c(below_thresh_TV,above_thresh_TV, below_thresh_TS, above_thresh_TS)) %>% c(tv = .[1,1]+.[1,2], ts = .[1,3]+.[1,4]), #total ts and tv individually in analysis
#        annot.indv = df %>% subset(TS_TV == "FALSE" | TS_TV == "TRUE") %>% count(.(TS_TV)) #total ts and tv individually in annot
#        )
# }
# 
# c.n("HIV_ENV",env.annot)
# c.n("NP_Aichi",np.annot)
# c.n("NP_PR8",np2.annot)
# c.n("HA_WSN33",ha.annot)
# ##all are correct
# ##I then checked whether total numbers of TS and TV used for this analysis matches the total number of TS and tv in the .annot, according to the below or above 1.
# c.n = function(stud, df, det.bf = "det"){
#   if(det.bf == "det"){
#     df2 = fisher.bloom.det
#     df %<>% subset(Rel_site_pref <= 1)
#   } 
#   else if(det.bf == "bf"){
#     df2 = fisher.bloom.bf
#     df %<>% subset(Rel_site_pref >= 1)
#   }
#   
#   list(fish.total = df2 %>% subset(Study == stud) %>% .$total_n %>% unique, #total ts and tv exclduing both in fisher analysis
#               annot.total = df %>% subset(TS_TV == "FALSE" | TS_TV == "TRUE") %>% count(.(TS_TV)) %>% .$freq %>% sum, #total ts and tv excluding both in annot
#        fish.indv = df2 %>% subset(Study == stud & threshold == .05, select = c(below_thresh_TV,above_thresh_TV, below_thresh_TS, above_thresh_TS)) %>% c(tv = .[1,1]+.[1,2], ts = .[1,3]+.[1,4]), #total ts and tv individually in analysis
#        annot.indv = df %>% subset(TS_TV == "FALSE" | TS_TV == "TRUE") %>% count(.(TS_TV)) #total ts and tv individually in annot
#        )
# }
# 
# c.n("HIV_ENV",env.annot)
# c.n("NP_Aichi",np.annot)
# c.n("NP_PR8",np2.annot)
# c.n("HA_WSN33",ha.annot)
# ##all N are correct.


#end----
```


##CDFs
Calculated for all studies but in paper only Bloom DMS, Lauring, and HIV studies are shown.
```{r CDF}
#sub for cdf
sub_cdf = function(studies, Set = "all_xSTOP", Stops = "STOP", xGenes = c("NONE"), xDatasets = c("NONE"), only_Genes = c("NONE"),data_frame = all_studies,study_name = studies){ #defualts then is without stops
  study_stops = Stops
  if(grepl("STOP",Set)){
    if(grepl("TEV|Lauring",studies)){
      study_stops = "stop"
    }
    else if(grepl("VSV",studies)){
      study_stops = "Stop"
    }
  } 
  sub(studies, Set, study_stops, xGenes, xDatasets, only_Genes,data_frame) %>% mutate(Study = study_name)
}


cdf.df = rbind(sub_cdf("Lauring",study_name = "Lauring"),
      sub_cdf("Lauring",xDatasets = "HA/NA",study_name = "Lauring_gw"),
      sub_cdf("Lauring",only_Genes = c("HA","NA"),study_name = "Lauring_HANA"),
      sub_cdf("Lauring",xGenes = c("HA","NA"),study_name = "Lauring_xHAxNA"),
      sub_cdf("Lauring",xGenes = c("HA","NA"),study_name = "Internal_Viable",data_frame = all_studies %>% subset(Fitness.Mean > 0)),
      sub_cdf("HIV",only_Genes = "Integrase",study_name = "HIV_IN"),
      sub_cdf("HIV",only_Genes = "Capsid",study_name = "HIV_CA"),
      sub_cdf("phix174_Domingo_Calap_09",study_name = "phix174"),
      sub_cdf("qb_Domingo_Calap_09",study_name = "qb"),
      sub_cdf("TEV_Carrasco_07_full",study_name = "TEV"),
      sub_cdf("f1_Peris_10",study_name = "f1"),
      sub_cdf("VSV"),
      sub_cdf("Bloom",only_Genes = "HA",study_name = "HA_WSN33"),
      sub_cdf("Bloom",only_Genes = "NP",xDatasets = "NP2",study_name = "NP_Aichi"),
      sub_cdf("Bloom",only_Genes = "NP",data_frame = all_studies %>% subset(Dataset == "NP2"),study_name = "NP_PR8"),
      sub_cdf("Bloom_env",only_Genes = "env",study_name = "HIV_ENV")
)




other.studs = cdf.df$Study %>% unique %>% remove_elements(elements = c("HIV_ENV","NP_PR8","NP_Aichi","HA_WSN33"))
decent.n.studs = other.studs %>% remove_elements(elements = c("qb","phix174"))
bloom.studs = cdf.df$Study %>% unique %>% retain_elements(elements = c("HIV_ENV","NP_PR8","NP_Aichi","HA_WSN33"))

cdf.df$Study = factor(cdf.df$Study,levels = c("Lauring","Lauring_gw","Lauring_xHAxNA","Internal_Viable","Lauring_HANA","HA_WSN33","NP_PR8","NP_Aichi","HIV_IN","HIV_CA","HIV_ENV","TEV","f1","VSV","phix174","qb"))

#end----
```


##Ts-Tv Fitness Difference within Radical (R) or Conservative Categories (C)
```{r TSTV.RC}
#mk r/c dataframes---- 
#get the r/c distincitons first.
choose.rc = function(dat,rc = "both",type = "none"){
  if(rc == "both"){
    dat
  } else if(rc == "r"){
    if(type == "C_ncharge"){
      dat %>% subset(C_ncharge == F)
    } else if(type == "C_npolarity"){
      dat %>% subset(C_npolarity == F)
    } else if(type == "C_npv"){
      dat %>% subset(C_npv == F)
    }
  } else if(rc == "c"){
    if(type == "C_ncharge"){
      dat %>% subset(C_ncharge == T)
    } else if(type == "C_npolarity"){
      dat %>% subset(C_npolarity == T)
    } else if(type == "C_npv"){
      dat %>% subset(C_npv == T)
    }
  }
}

aasub.rc = function(dat){
  dat %>% mutate(AA_Sub = paste0(WT_AA,X.SITE,AA_Mut))
}

sub.rc = function(dat1,id = "AA_Sub",dataset,dat2 = all_studies){
  muts.desired = dat1[,id]
  dat2 %<>% subset(Dataset == dataset)
  dat2[dat2[,id] %in% muts.desired,]
}

choose2sub.rc = function(dat1,rc = "both",type = "none",id = "AA_Sub",dataset,dat2 = all_studies){
  choose.rc(dat1,rc,type) %>% aasub.rc %>% sub.rc(id = id,dataset = dataset,dat2 = dat2)
}

sets.rc = c(r = "r",c = "c")
sets.type = c(charge = "C_ncharge",polarity = "C_npolarity",pv = "C_npv")
datasets = c("NP","NP2","HA","env")
dfs = rbind(np.annot %>% mutate(dtset = "NP"),np2.annot %>% mutate(dtset = "NP2"),ha.annot %>% mutate(dtset = "HA"),env.annot %>% mutate(dtset = "env"))

bloom.rc = lapply(sets.rc,function(set){
  lapply(sets.type,function(ty){
    lapply(datasets,function(dataset){
      choose2sub.rc(dat1 = dfs %>% subset(dtset == dataset),rc = set,type = ty,dataset = dataset)
    }) %>% do.call(rbind,.)
  })
}) #now have 6 dataframes with all 4 bloom datasets/genes in a list list$r$type and list$c$type made from selecting only the mutations of interest from all studies, can put each one of these dataframes into the analysis and get $rc$type analysis output for all 4 datasets/genes.

#addtl fisher and plotting fnx----
study2fisher_thresh = function(studies,
                         width = .05,min = 0,max = max_fit(studies),slide_by = 1, #windows settings
                         Stops = "INCLUDE", xGenes = c("NONE"), xDatasets = c("NONE"), only_Genes = c("NONE"),
                         data_frame = all_studies,
                         study_name = studies){
  if(studies == "Bloom_env"){
    max = min(max_fit(studies),10)
  } else if(grepl("Bloom",studies)){
    max = min(max_fit(studies,xGenes = "env"),10)
  }
  
  if(grepl("Bloom",studies)){
    min = width  
  }
  
  win = set_windows(width,min,max,slide_by)
  
  lapply(1:length(win$up_end),function(x){
    fisher_mat_thresh(studies,
               windows = T,low_th = win$low_end[x],up_th = win$up_end[x], #set windows options
               Stops, xGenes, xDatasets, only_Genes,
               data_frame) %>%
      mat2stat_thresh(m = .,study = ifelse(Stops != "INCLUDE",paste0(study_name,"_xSTOP"),study_name),
               low_th = win$low_end[x],up_th = win$up_end[x])
  }) %>% do.call(rbind,.)
}

fisher_mat_thresh =function(studies,
                       windows = F,low_th = F,up_th =F, #set windows options
                       Stops = "INCLUDE", xGenes = c("NONE"), xDatasets = c("NONE"), only_Genes = c("NONE"),
                       data_frame = all_studies){
  #Set up fisher matrix  
  m = matrix(data = rep(NA,4),
                nrow = 2,
                ncol = 2, 
                byrow = T,
                dimnames = list(c("TV","TS"),c("below_threshold","above_threshold"))
  )
  
  #Find complete sub
  ##Set stops options
  xAA_Sub = "xxx"
  
  if(Stops != "INCLUDE"){
    xAA_Sub = Stops
  }
  
  if(only_Genes != c("NONE")){
    s = subset(data_frame,grepl(paste(studies,collapse = "|"),Study) & S_NS == "NS" & !is.na(S_NS) & !grepl(paste(xGenes,collapse = "|"),Gene) & !grepl(paste(xDatasets,collapse = "|"),Dataset) & grepl(paste(only_Genes,collapse = "|"),Gene) & !grepl(xAA_Sub,AA_Sub))
  } else{
    s = subset(data_frame,grepl(paste(studies,collapse = "|"),Study) & S_NS == "NS" & !is.na(S_NS) & !grepl(paste(xGenes,collapse = "|"),Gene) & !grepl(paste(xDatasets,collapse = "|"),Dataset) & !grepl(xAA_Sub,AA_Sub))
  }
  
  #find total n ts, tv from complete sub
  N_TV = s %>% .$TS_TV %in% "TV" %>% sum
  N_TS = s %>% .$TS_TV %in% "TS" %>% sum
  
  # #sub with window
  # if(windows){
  #   s = subset(s, Fitness.Mean >= low_th & Fitness.Mean <= up_th)
  # }

  #now fill in fisher matrix
  m["TV","below_threshold"] = s %>% subset(Fitness.Mean <= low_th) %>% .$TS_TV %in% "TV" %>% sum
  m["TS","below_threshold"] = s %>% subset(Fitness.Mean <= low_th) %>% .$TS_TV %in% "TS" %>% sum
  m["TV","above_threshold"] = N_TV - m["TV","below_threshold"]
  m["TS","above_threshold"] = N_TS - m["TS","below_threshold"]
  m
}#gets fisher matrix for below or EQUAL to threshold and strictly above threshold.

fisher_thresh_pl_fnx = function(dat,
                                study = F,
                                xSTOPS = F,
                                xmin = 0, xmax = ifelse(study!=F,
                                                        max(dat %>% subset(Study == study) %>% .$threshold %>% as.vector %>% as.numeric),
                                                        max(dat %>% .$threshold %>% as.vector %>% as.numeric))){
  if(study != F){
    dat %<>% subset(Study == study)
  }
  if(!xSTOPS){
    dat %<>% subset(!grepl("xSTOP",Study))
  }
  if(study %in% c(bloom.studs,bloom_flu_studs)){
    xlabel = "Relative Preference Threshold"
  } else{
    xlabel = "Relative Fitness Threshold"
  }
  pl = ggplot(data = dat, aes(x = threshold, y = OR, group = Study))+
    geom_point(stat = "identity", size = 4, aes(color = ifelse(p.value %>% as.vector %>% as.numeric <.05,"p<.05","p>=.05")))+
    geom_line()+
    # geom_text(aes(y =  auc %>% as.numeric+.05,
    #               label = paste0(N_TS,",",N_TV),
    #               color = ifelse(p.value %>% as.vector %>% as.numeric <.1,"p<.1","p>=.1")
    # )
    # )+
    scale_color_manual(values=c("p<.05" = "red","p>=.05" = "black"),name = "p-value")+
    # ylim(0,1)+
    xlab(xlabel)+
    xlim(xmin,xmax)+
    ylab("Odds Ratio")+
    geom_hline(aes(yintercept = 1))+
    geom_vline(aes(xintercept = 1))+
    theme_bw()

  if(study != F){
    pl+ggtitle(study)
  } else{
    pl+facet_wrap(~Study)
  }
} #when study = F this gives a warning message "In max(dat %>% subset(Study == study) %>% .$threshold %>% as.vector %>%:  no non-missing arguments to max; returning -Inf" for the xmax, but this doesn't matter for what we need.

cdf.pl.fnx = function(dat,
                      study = F){
  if(study != F){
    dat %<>% subset(Study == study)
  }
  if(study %in% c(bloom.studs,bloom_flu_studs)){
    xlabel = "Relative Preference"
  } else{
    xlabel = "Relative Fitness"
  }

  pl = ggplot(dat,aes(x = Fitness.Mean))+
    stat_ecdf(aes(color=TS_TV,group = TS_TV))+
    scale_color_manual(values = c(TS = "blue", TV = "red"))+
    geom_vline(aes(xintercept = 1))+
    ylab("Proportion <= x")+
    xlab(xlabel)+
    theme_bw()

  if(study != F){
    pl+ggtitle(study)
  } else{
    pl+facet_wrap(~Study,scales = "free_x")
  }
}

#analysis r/c----
r.fish = lapply(bloom.rc$r,function(type){
  #calc
  dat = list(
    ##all
    fisher_thresh = rbind(
      study2fisher_thresh(studies = "Bloom_env",only_Genes = "env",study_name = "HIV_ENV",data_frame = type), #bloom env
      study2fisher_thresh("Bloom",only_Genes = "NP",xDatasets = "NP2",study_name = "NP_Aichi",data_frame = type),  #bloom np aichi
      study2fisher_thresh("Bloom",only_Genes = "NP",study_name = "NP_PR8",data_frame = type %>% subset(Dataset == "NP2"),max = min(10,max_fit(studies,xGenes = "env"))),  #bloom np pr8
      study2fisher_thresh("Bloom",only_Genes = "HA",study_name = "HA_WSN33",data_frame = type)   #bloom ha wsn33
    ) %>%
      subset(threshold != 0), #copy pasted from above
    
    ##det
    fisher.bloom.det = rbind(
      study2fisher_thresh.sets("Bloom_env",only_Genes = "env",study_name = "HIV_ENV", set = "DET",data_frame = type), #bloom env
      study2fisher_thresh.sets("Bloom",only_Genes = "NP",xDatasets = "NP2",study_name = "NP_Aichi", set = "DET",data_frame = type),  #bloom np aichi
      study2fisher_thresh.sets("Bloom",only_Genes = "NP",study_name = "NP_PR8",data_frame = type %>% subset(Dataset == "NP2"), set = "DET"),  #bloom np pr8
      study2fisher_thresh.sets("Bloom",only_Genes = "HA",study_name = "HA_WSN33", set = "DET",data_frame = type)
    ),  #bloom ha wsn33
    
    ##bf
    fisher.bloom.bf = rbind(
      study2fisher_thresh.sets("Bloom_env",only_Genes = "env",study_name = "HIV_ENV", set = "BF",data_frame = type), #bloom env
      study2fisher_thresh.sets("Bloom",only_Genes = "NP",xDatasets = "NP2",study_name = "NP_Aichi", set = "BF",data_frame = type),  #bloom np aichi
      study2fisher_thresh.sets("Bloom",only_Genes = "NP",study_name = "NP_PR8",data_frame = type %>% subset(Dataset == "NP2"), set = "BF"),  #bloom np pr8
      study2fisher_thresh.sets("Bloom",only_Genes = "HA",study_name = "HA_WSN33",set = "BF",data_frame = type)
    )#bloom ha wsn33
  ) #data list
  #plot
  pl = list(all = list(env = fisher_thresh_pl_fnx(dat$fisher_thresh,"HIV_ENV"),
                  np.h3n2 = fisher_thresh_pl_fnx(dat$fisher_thresh,"NP_Aichi"),
                  np.h1n1 = fisher_thresh_pl_fnx(dat$fisher_thresh,"NP_PR8"),
                  ha = fisher_thresh_pl_fnx(dat$fisher_thresh,"HA_WSN33")
                  ),
       det = list(flu = fisher_thresh_pl_fnx(dat$fisher.bloom.det %>% 
                                                         subset(!grepl("Bloom",Study) & Study != "HIV_ENV") %>%
                                                         mutate(Study = mapvalues(Study,
                                                                                  c("NP_Aichi","NP_PR8","HA_WSN33"),
                                                                                  c("NP_H3N2","NP_H1N1","HA_H1N1")
                                                         )))+
               theme_bw()+
               expand_limits(y=0)+xlab("Relative Preference Threshold"),
             env = fisher_thresh_pl_fnx(dat$fisher.bloom.det,"HIV_ENV")+theme_bw()+expand_limits(y=0)
             ),
       bf = list(flu = fisher_thresh_pl_fnx(dat$fisher.bloom.bf %>%
                                                            subset(!grepl("Bloom",Study) & Study != "HIV_ENV") %>%
                                                            mutate(Study = mapvalues(Study,c("NP_Aichi","NP_PR8","HA_WSN33"),
                                                                                     c("NP_H3N2","NP_H1N1","HA_H1N1"))))+
                   theme_bw()+
                   expand_limits(y=0)+xlab("Relative Preference Threshold"),
                 env = fisher_thresh_pl_fnx(dat$fisher.bloom.bf,"HIV_ENV")
                 )
  )
  list(data = dat,pl = pl)
})


c.fish = lapply(bloom.rc$c,function(type){
  #calc
  dat = list(
  ##all
  fisher_thresh = rbind(
    study2fisher_thresh(studies = "Bloom_env",only_Genes = "env",study_name = "HIV_ENV",data_frame = type), #bloom env
    study2fisher_thresh("Bloom",only_Genes = "NP",xDatasets = "NP2",study_name = "NP_Aichi",data_frame = type),  #bloom np aichi
    study2fisher_thresh("Bloom",only_Genes = "NP",study_name = "NP_PR8",data_frame = type %>% subset(Dataset == "NP2"),max = min(10,max_fit(studies,xGenes = "env"))),  #bloom np pr8
    study2fisher_thresh("Bloom",only_Genes = "HA",study_name = "HA_WSN33",data_frame = type)   #bloom ha wsn33
  ) %>%
    subset(threshold != 0), #copy pasted from above
  
  ##det
  fisher.bloom.det = rbind(
    study2fisher_thresh.sets("Bloom_env",only_Genes = "env",study_name = "HIV_ENV", set = "DET",data_frame = type), #bloom env
    study2fisher_thresh.sets("Bloom",only_Genes = "NP",xDatasets = "NP2",study_name = "NP_Aichi", set = "DET",data_frame = type),  #bloom np aichi
    study2fisher_thresh.sets("Bloom",only_Genes = "NP",study_name = "NP_PR8",data_frame = type %>% subset(Dataset == "NP2"), set = "DET"),  #bloom np pr8
    study2fisher_thresh.sets("Bloom",only_Genes = "HA",study_name = "HA_WSN33", set = "DET",data_frame = type)
  ),  #bloom ha wsn33
  
  ##bf
  fisher.bloom.bf = rbind(
    study2fisher_thresh.sets("Bloom_env",only_Genes = "env",study_name = "HIV_ENV", set = "BF",data_frame = type), #bloom env
    study2fisher_thresh.sets("Bloom",only_Genes = "NP",xDatasets = "NP2",study_name = "NP_Aichi", set = "BF",data_frame = type),  #bloom np aichi
    study2fisher_thresh.sets("Bloom",only_Genes = "NP",study_name = "NP_PR8",data_frame = type %>% subset(Dataset == "NP2"), set = "BF"),  #bloom np pr8
    study2fisher_thresh.sets("Bloom",only_Genes = "HA",study_name = "HA_WSN33",set = "BF",data_frame = type)
  )   #bloom ha wsn33
  )#data list
  #plot
  pl = list(all = list(env = fisher_thresh_pl_fnx(dat$fisher_thresh,"HIV_ENV"),
                  np.h3n2 = fisher_thresh_pl_fnx(dat$fisher_thresh,"NP_Aichi"),
                  np.h1n1 = fisher_thresh_pl_fnx(dat$fisher_thresh,"NP_PR8"),
                  ha = fisher_thresh_pl_fnx(dat$fisher_thresh,"HA_WSN33")
                  ),
       det = list(flu = fisher_thresh_pl_fnx(dat$fisher.bloom.det %>% 
                                                         subset(!grepl("Bloom",Study) & Study != "HIV_ENV") %>%
                                                         mutate(Study = mapvalues(Study,
                                                                                  c("NP_Aichi","NP_PR8","HA_WSN33"),
                                                                                  c("NP_H3N2","NP_H1N1","HA_H1N1")
                                                         )))+
               theme_bw()+
               expand_limits(y=0)+xlab("Relative Preference Threshold"),
             env = fisher_thresh_pl_fnx(dat$fisher.bloom.det,"HIV_ENV")+theme_bw()+expand_limits(y=0)
             ),
       bf = list(flu = fisher_thresh_pl_fnx(dat$fisher.bloom.bf %>%
                                                            subset(!grepl("Bloom",Study) & Study != "HIV_ENV") %>%
                                                            mutate(Study = mapvalues(Study,c("NP_Aichi","NP_PR8","HA_WSN33"),
                                                                                     c("NP_H3N2","NP_H1N1","HA_H1N1"))))+
                   theme_bw()+
                   expand_limits(y=0)+xlab("Relative Preference Threshold"),
                 env = fisher_thresh_pl_fnx(dat$fisher.bloom.bf,"HIV_ENV")
                 )
  )
  list(data = dat, pl = pl)
})

r.cdf = lapply(bloom.rc$r,function(type){
  #calc
  all = rbind(
      sub_cdf("Bloom",only_Genes = "HA",study_name = "HA_WSN33",data_frame = type),
      sub_cdf("Bloom",only_Genes = "NP",xDatasets = "NP2",study_name = "NP_Aichi",data_frame = type),
      sub_cdf("Bloom",only_Genes = "NP",data_frame = type %>% subset(Dataset == "NP2"),study_name = "NP_PR8"),
      sub_cdf("Bloom_env",only_Genes = "env",study_name = "HIV_ENV",data_frame = type)) %>% mutate(Study = mapvalues(Study,c("NP_Aichi","NP_PR8","HA_WSN33"),c("NP_H3N2","NP_H1N1","HA_H1N1")))
  
  all$Study %<>% factor(levels = c("NP_H3N2","NP_H1N1","HA_H1N1","HIV_ENV"))
  
  bf = all %>% subset(Fitness.Mean>=1)

  det = all %>% subset(Fitness.Mean<=1)
  
  data = list(all = all,det = det,bf = bf)
  
  pl = list(all = list(flu =cdf.pl.fnx(dat = data$all %>% subset(Study != "HIV_ENV"))+xlab("Relative Preference"),
                  env = cdf.pl.fnx(dat = data$all,"HIV_ENV")),
       det = list(flu = cdf.pl.fnx(dat = data$det %>% subset(Study != "HIV_ENV"))+theme_bw()+xlab("Relative Preference"),
                  env = cdf.pl.fnx(dat = data$det,"HIV_ENV")+theme_bw()+xlab("Relative Preference")),
       
       bf = list(flu = cdf.pl.fnx(dat = data$bf %>% subset(Study != "HIV_ENV"))+xlab("Relative Preference")+theme_bw(),
                 env = cdf.pl.fnx(dat = data$bf,"HIV_ENV")+xlab("Relative Preference")+theme_bw())
       )
  
  list(data = data,pl = pl)
  
})
c.cdf = lapply(bloom.rc$c,function(type){
  #calc
  all = rbind(
      sub_cdf("Bloom",only_Genes = "HA",study_name = "HA_WSN33",data_frame = type),
      sub_cdf("Bloom",only_Genes = "NP",xDatasets = "NP2",study_name = "NP_Aichi",data_frame = type),
      sub_cdf("Bloom",only_Genes = "NP",data_frame = type %>% subset(Dataset == "NP2"),study_name = "NP_PR8"),
      sub_cdf("Bloom_env",only_Genes = "env",study_name = "HIV_ENV",data_frame = type)) %>% mutate(Study = mapvalues(Study,c("NP_Aichi","NP_PR8","HA_WSN33"),c("NP_H3N2","NP_H1N1","HA_H1N1")))
  
  all$Study %<>% factor(levels = c("NP_H3N2","NP_H1N1","HA_H1N1","HIV_ENV"))
  
  bf = all %>% subset(Fitness.Mean>=1)

  det = all %>% subset(Fitness.Mean<=1)
  
  data = list(all = all,det = det,bf = bf)
  
  pl = list(all = list(flu =cdf.pl.fnx(dat = data$all %>% subset(Study != "HIV_ENV"))+xlab("Relative Preference"),
                  env = cdf.pl.fnx(dat = data$all,"HIV_ENV")),
       det = list(flu = cdf.pl.fnx(dat = data$det %>% subset(Study != "HIV_ENV"))+theme_bw()+xlab("Relative Preference"),
                  env = cdf.pl.fnx(dat = data$det,"HIV_ENV")+theme_bw()+xlab("Relative Preference")),
       
       bf = list(flu = cdf.pl.fnx(dat = data$bf %>% subset(Study != "HIV_ENV"))+xlab("Relative Preference")+theme_bw(),
                 env = cdf.pl.fnx(dat = data$bf,"HIV_ENV")+xlab("Relative Preference")+theme_bw())
       )
  
  list(data = data,pl = pl)
  
})


#calling plots/tables:
#r/c.fish/cdf$type$all/det/bf$flu/env = plot (except: fish$type$all$env/np.h3n2/np.h1n1/ha)
# 
# #flu pl----
# ##fish----
# r.fish$charge$det$flu
# c.fish$charge$det$flu
# 
# r.fish$polarity$det$flu
# c.fish$polarity$det$flu
# 
# r.fish$pv$det$flu
# c.fish$pv$det$flu
# 
# ##cdf----
# r.cdf$charge$det$flu
# c.cdf$charge$det$flu
# 
# r.cdf$polarity$det$flu
# c.cdf$polarity$det$flu
# 
# r.cdf$pv$det$flu
# c.cdf$pv$det$flu
# 
# # #hiv----
# # ##fish----
# r.fish$charge$det$env
# c.fish$charge$det$env
# 
# r.fish$polarity$det$env
# c.fish$polarity$det$env
# 
# r.fish$pv$det$env
# c.fish$pv$det$env

#end----
```


##R-C Fitness Differences
Are radical amino acid changes more detrimental than conservative ones?

Calculates for only detrimental & netural (shown in paper), for all mutations (not shown in paper), for only beneficials mutations (not shown in paper), and for only substitutions accessible by either a Ts or a Tv and not by both (shown in paper), all substitutions accessible by a single point mutation (not shown in paper), substitutions accessible by only more than one mutation and not single mutations (not shown in paper), and all substitutions (not shown in paper). In addition to genes from Bloom DMS studies, calculates for flu genes as well as HIV IN and CA (not shown in paper).
```{r RC.DIFF}
#fisher.rc fnxs----
dfs %<>% mutate(dtset = mapvalues(dtset,c("NP","NP2","HA","env"),c("NP.H3N2","NP.H1N1","HA.H1N1","HIV.ENV")))

rc.max.fit = function(only.dtset="all",df = dfs){
  ifelse(only.dtset=="all",
         df$Rel_site_pref %>% max(), 
         df %>% subset(dtset == only.dtset) %>% .$Rel_site_pref %>% max()
         )
}


rc.mat2stat.thresh = function(m,study,low_th){
  test = fisher.test(m, conf.int = T)
  data.frame(p.value = test$p.value,
             OR = test$estimate,
             OR_CI95_lower = test$conf.int[1],
             OR_CI95_upper = test$conf.int[2],
             total_n = sum(m),
             # total_TV = m["TV",] %>% sum,
             # total_TS = m["TS",] %>% sum,
             below_thresh_R = m["R","below_threshold"],
             above_thresh_R = m["R","above_threshold"],
             below_thresh_C = m["C","below_threshold"],
             above_thresh_C = m["C","above_threshold"],
             Study = study,
             threshold = low_th,
             row.names = NULL
  )
}

rc.fisher.mat.thresh = function(set = "ALL",
                                     only.dtset = "all",
                                     muts = c("all"),
                                     class,
                                     low_th = F,
                                     df = dfs){
  #Set up fisher matrix  
  m = matrix(data = rep(NA,4),
             nrow = 2,
             ncol = 2, 
             byrow = T,
             dimnames = list(c("R","C"),c("below_threshold","above_threshold"))
  )
  
  #Find complete sub
  ##sub according to options
  ###ns only
  df %<>% subset(S_NS == F)
  ###muts
  if(muts != "all"){
    df %<>% subset(grepl(paste(muts,collapse = "|"),TS_TV))
  }
  ###only.dtset
  if(only.dtset != "all"){
    df %<>% subset(dtset == only.dtset)
  }
  
  #Find set sub
  if(set != "ALL"){
    if(set == "DET"){
      df %<>% subset(Rel_site_pref<=1)
    } else if(set == "BF"){
      df %<>% subset(Rel_site_pref>=1)
    }
  }
  
  #Find total n C, R from complete sub
  N.R = df %>% .[,class] %in% F %>% sum()
  N.C = df %>% .[,class] %in% T %>% sum()
  
  #Now fill in fisher matrix
  m["R","below_threshold"] = df %>% subset(Rel_site_pref <= low_th) %>% .[,class] %in% F %>% sum()
  m["C","below_threshold"] = df %>% subset(Rel_site_pref <= low_th) %>% .[,class] %in% T %>% sum()
  m["R","above_threshold"] = N.R - m["R","below_threshold"]
  m["C","above_threshold"] = N.C - m["C","below_threshold"]
  m
}

rc.study2fisher.thresh = function(set = "ALL",
                                  only.dtset = "all",
                                  muts = c("all"),
                                  class,
                                  df = dfs,
                                  width = .05,
                                  min = ifelse(set == "BF",
                                               1+width,
                                               width),
                                  max = ifelse(set == "DET",
                                               1,
                                               min(rc.max.fit(only.dtset = only.dtset),10)
                                  ),
                                  slide_by = 1, #windows settings
                                  study.name = only.dtset){
  
  win = set_windows(width,min,max,slide_by)
  lapply(1:length(win$up_end),function(x){
    rc.fisher.mat.thresh(set,
                         only.dtset,
                         muts,
                         class,
                         low_th = win$low_end[x],
                         df) %>%
      rc.mat2stat.thresh(m = .,
                         study = study.name,
                         low_th = win$low_end[x])
  }) %>% do.call(rbind,.) %>% mutate(Mutations = ifelse("BOTH" %in% muts, "MUT1",
                                                        ifelse(muts == "MUT23","Mut23",
                                                               ifelse(muts == "all", "All.Codon.Muts",
                                                               "TS.TV.Only"))),
                                     Set = mapvalues(set,c("ALL","DET","BF"),c("All Fitness Levels","Detrimentals","Beneficials"),warn_missing = F),
                                     Class = mapvalues(class,c("C_ncharge","C_npolarity","C_npv"),c("Charge","Polarity","Polarity & Volume"),warn_missing = F))
}


rc.flu2fisher.thresh = function(set = "ALL",
                                muts = c("all"),
                                class
                                ){
  lapply(dfs$dtset %>% unique %>% remove_elements("HIV.ENV"),function(dts){
    rc.study2fisher.thresh(set,dts,muts,class)
  }) %>% do.call(rbind,.)
}


rc.classes2fisher.thresh = function(set = "ALL",
                             muts = c("all"),
                             class = classes,
                             flu.hiv = "flu"){
  if(flu.hiv == "flu"){
    
    lapply(class,function(cl){
      rc.flu2fisher.thresh(set,muts,class = cl)
    })
  } 
  else{
    
    lapply(class,function(cl){
    rc.study2fisher.thresh(set,only.dtset = "HIV.ENV",muts,class = cl)
      
    })
  }
}


rc.sets2fisher.thresh = function(set = sets,
                                 muts = c("all"),
                                 flu.hiv = "flu"){
  lapply(sets,function(s){
    rc.classes2fisher.thresh(set = s,muts,flu.hiv = flu.hiv)
  })
}


rc.muts2fisher.thresh = function(muts = muts.s,flu.hiv = "flu"){
  lapply(muts,function(m){
    rc.sets2fisher.thresh(muts = m,flu.hiv = flu.hiv)
  })
}

classes = c(charge = "C_ncharge",polarity = "C_npolarity",pv = "C_npv")
sets = c(all = "ALL",det = "DET", bf = "BF")
muts.s = list(all = "all",tstv = c("TRUE","FALSE"),mut1 = c("TRUE","FALSE","BOTH"),mut23 = "MUT23")

#cdf fnxs----
rc.cdf.sub = function(set = "ALL",
                      only.dtset = "all",
                      muts = c("all"),
                      class,
                      df = dfs){

  #Find complete sub
  ##sub according to options
  ###ns only
  df %<>% subset(S_NS == F)
  ###muts
  if(muts != "all"){
    df %<>% subset(grepl(paste(muts,collapse = "|"),TS_TV))
  }
  ###only.dtset
  if(only.dtset != "all"){
    df %<>% subset(dtset == only.dtset)
  }
  
  #Find set sub
  if(set != "ALL"){
    if(set == "DET"){
      df %<>% subset(Rel_site_pref<=1)
    } else if(set == "BF"){
      df %<>% subset(Rel_site_pref>=1) #cut off plot at about 10 though using min(rc.max.fit(only.dtset = only.dtset),10)
    }
  }
  
  ###Mutate with muts,set,class
  x = if("all" == muts){"Mut123"}else if("Mut23" == muts){"Mut23"} else if(("TRUE" %in% muts) & ("FALSE" %in% muts) & !("BOTH" %in% muts)){"TS.TV.Only"} else("Mut1")
  
  df %<>% mutate(Mutations = x ,
                 Set = mapvalues(set,c("ALL","DET","BF"),c("All Fitness Levels","Detrimentals","Beneficials"),warn_missing = F),
                 Class = mapvalues(class,c("C_ncharge","C_npolarity","C_npv"),c("Charge","Polarity","Polarity & Volume"),warn_missing = F))
  
  colnames(df)[grepl(class,colnames(df))] = "R.C"
  
  df %<>% mutate(R.C = mapvalues(R.C,c("TRUE","FALSE"),c("C","R"))) %>% subset(select = c(X.SITE,WT_AA,WT_codon,SITE_ENTROPY,WT_site_pref,AA_Mut,Mut_site_pref,S_NS,TS_TV,Rel_site_pref,wt.charge,mut.charge,wt.polarity,mut.polarity,wt.pv,mut.pv,R.C,dtset,Mutations,Set,Class))
  
  #return
  df
}

rc.flu.cdf.sub = function(set = "ALL",
                              muts = c("all"),
                              class,
                              df = dfs){
    lapply(dfs$dtset %>% unique %>% remove_elements("HIV.ENV"),function(dts){
    rc.cdf.sub(set,dts,muts,class)
  }) %>% do.call(rbind,.)
}



rc.cdf.pl.fnx = function(da,flu.hiv = "flu"){
  
    pl = ggplot(data = da,aes(x = Rel_site_pref))+
    stat_ecdf(aes(color=R.C,group = R.C))+
    scale_color_manual(values = c(C = "blue", R = "red"))+
    geom_vline(aes(xintercept = 1))+
    ylab("Proportion <= x")+
    xlab("Relative Preference")+
    theme_bw()
    
    tl = paste("CDF: R & C",da$Mutations %>% unique,da$Set %>% unique,da$Class %>% unique,sep = ": ")
  if(flu.hiv == "flu"){
    pl+facet_wrap(~dtset,scales = "free_x")+ggtitle(paste(tl,"Flu",sep = ": "))
  } 
  else{
    pl+ggtitle(paste(tl,"HIV.ENV",sep = ": "))+xlim(0,ifelse(max(da$Rel_site_pref)>1,10,1))
  }
}

rc.cdf.mkall.pl.fnx = function(flu.hiv = "flu"){
  lapply(muts.s,function(m){
    lapply(sets,function(s){
      lapply(classes,function(c){
        if(flu.hiv == "flu"){
          rc.flu.cdf.sub(s,m,c) %>% rc.cdf.pl.fnx(flu.hiv = "flu")
        }
        else{
          rc.cdf.sub(s,"HIV.ENV",m,c) %>% rc.cdf.pl.fnx(flu.hiv = "HIV.ENV")
        }
      })
    })
  })
}

#analysis r/c fit dif----
rc.flu.dat = rc.muts2fisher.thresh()
rc.env.dat = rc.muts2fisher.thresh(flu.hiv = "HIV.ENV")

rc.cdf.flu.dat = lapply(muts.s,function(m){
  lapply(sets,function(s){
    lapply(classes,function(c){
      rc.flu.cdf.sub(s,m,c)
    })})})

rc.cdf.env.dat = lapply(muts.s,function(m){
  lapply(sets,function(s){
    lapply(classes,function(c){
      rc.cdf.sub(s,"HIV.ENV",m,c)
    })})})


#plot fnx----
rc.fisher.pl.fnx = function(da,flu.hiv = "flu"){
  pl = ggplot(data = da, aes(x = threshold, y = OR, group = Study))+
    geom_point(stat = "identity", size = 4, aes(color = ifelse(p.value %>% as.vector %>% as.numeric <.05,"p<.05","p>=.05")))+
    geom_line()+
    scale_color_manual(values=c("p<.05" = "red","p>=.05" = "black"),name = "p-value")+
    xlab("Relative Preference Threshold")+
    ylab("Odds Ratio")+
    geom_hline(aes(yintercept = 1))+
    geom_vline(aes(xintercept = 1))+
    theme_bw()+
    expand_limits(y=0)
  
  tl = paste("Fisher Tests: R vs. C",da$Mutations %>% unique,da$Set %>% unique,da$Class %>% unique,sep = ": ")

  if(flu.hiv == "flu"){
    pl+facet_wrap(~Study)+ggtitle(paste(tl,"Flu",sep = ": "))
  } 
  else{
    pl+ggtitle(paste(tl,"HIV.ENV",sep = ": "))
  }
}

rc.mkall.pl.fnx = function(dat,flu.hiv = "flu"){
  lapply(dat,function(m){
    lapply(m,function(s){
      lapply(s,function(c){
        rc.fisher.pl.fnx(c,flu.hiv)
      })
    })
  })
}


##plots----
rc.flu.pl = rc.mkall.pl.fnx(rc.flu.dat)
rc.env.pl = rc.mkall.pl.fnx(rc.env.dat,flu.hiv = "HIV.ENV")

rc.cdf.flu.pl = rc.cdf.mkall.pl.fnx()
rc.cdf.env.pl = rc.cdf.mkall.pl.fnx(flu.hiv = "HIV.ENV")

#calling plots/tables:
#rc.fish/cdf.flu/env.pl$all/ts.tv.only/mut1/mut23$all/det/bf = plot


#end----
```


##Check Bloom DMS Data Code
Check some further things in Bloom data and in all_studies, see comments.
```{r CHECK.BLOOM}
#C
#1) check true ns and true s muts----
# 
# ##check annotation
# ch.1 = function(dat){
#   test = dat %>% subset(WT_AA != AA_Mut,select = c(WT_AA,AA_Mut,wt.charge,mut.charge,wt.polarity,mut.polarity,wt.pv,mut.pv)) %>% rename(c(wt.charge = 'o.charge',wt.polarity = 'o.polarity',wt.pv = 'o.pv'))
# 
#   test = mutate(test,
#                 o.ch.charge = ifelse(WT_AA %in% charge$pos,
#                                      "pos",
#                                      ifelse(WT_AA %in% charge$neg,
#                                             "neg",
#                                             ifelse(WT_AA %in% charge$neut,
#                                                    "neut",
#                                                    NA))),
#                 n.ch.charge = ifelse(AA_Mut %in% charge$pos,
#                                      "pos",
#                                      ifelse(AA_Mut %in% charge$neg,
#                                             "neg",
#                                             ifelse(AA_Mut %in% charge$neut,
#                                                    "neut",
#                                                    NA))),
#                 o.ch.polarity = ifelse(WT_AA %in% polarity$polar,
#                                        "polar",
#                                        ifelse(WT_AA %in% polarity$nonpolar,
#                                               "nonpolar",
#                                               NA)),
#                 n.ch.polarity = ifelse(AA_Mut %in% polarity$polar,
#                                        "polar",
#                                        ifelse(AA_Mut %in% polarity$nonpolar,
#                                               "nonpolar",
#                                               NA)),
#                 o.ch.pv = ifelse(WT_AA %in% pv$special,
#                                  "special",
#                                  ifelse(WT_AA %in% pv$neut_small,
#                                         "neut_small",
#                                         ifelse(WT_AA %in% pv$polar_small,
#                                                "polar_small",
#                                                ifelse(WT_AA %in% pv$polar_large,
#                                                       "polar_large",
#                                                       ifelse(WT_AA %in% pv$nonpolar_small,
#                                                              "nonpolar_small",
#                                                              ifelse(WT_AA %in% pv$nonpolar_large,
#                                                                     "nonpolar_large",
#                                                                     NA)))))),
#                 n.ch.pv = ifelse(AA_Mut %in% pv$special,
#                                  "special",
#                                  ifelse(AA_Mut %in% pv$neut_small,
#                                         "neut_small",
#                                         ifelse(AA_Mut %in% pv$polar_small,
#                                                "polar_small",
#                                                ifelse(AA_Mut %in% pv$polar_large,
#                                                       "polar_large",
#                                                       ifelse(AA_Mut %in% pv$nonpolar_small,
#                                                              "nonpolar_small",
#                                                              ifelse(AA_Mut %in% pv$nonpolar_large,
#                                                                     "nonpolar_large",
#                                                                     NA))))))
#   ) %>% mutate(v.o.charge = o.charge == o.ch.charge,
#                v.n.charge = mut.charge == n.ch.charge,
#                v.o.polarity = o.polarity == o.ch.polarity,
#                v.n.polarity = mut.polarity == n.ch.polarity,
#                v.o.pv = o.pv == o.ch.pv,
#                v.n.pv = mut.pv == n.ch.pv)
#   list(v.o.charge = count(test,.(v.o.charge)),
#        v.o.polarity = count(test,.(v.o.polarity)),
#        v.o.pv = count(test,.(v.o.pv)),
#        v.n.charge = count(test,.(v.n.charge)),
#        v.n.polarity = count(test,.(v.n.polarity)),
#        v.n.pv = count(test,.(v.n.pv))) #trues means the annotated and the check versions are the same. true is good.
# } #v.o.class means verifcation whether WT aa was annotated with its correct class, true means correct.  v.n.class means verficiation whether mut aa was annotated with its correct class-should all be true
# 
# ch.1.s = function(dat){
#   test = dat %>% subset(WT_AA == AA_Mut,select = c(WT_AA,AA_Mut,wt.charge,mut.charge,wt.polarity,mut.polarity,wt.pv,mut.pv)) %>% rename(c(wt.charge = 'o.charge',wt.polarity = 'o.polarity',wt.pv = 'o.pv'))
# 
#   test = mutate(test,
#                 o.ch.charge = ifelse(WT_AA %in% charge$pos,
#                                      "pos",
#                                      ifelse(WT_AA %in% charge$neg,
#                                             "neg",
#                                             ifelse(WT_AA %in% charge$neut,
#                                                    "neut",
#                                                    NA))),
#                 n.ch.charge = ifelse(AA_Mut %in% charge$pos,
#                                      "pos",
#                                      ifelse(AA_Mut %in% charge$neg,
#                                             "neg",
#                                             ifelse(AA_Mut %in% charge$neut,
#                                                    "neut",
#                                                    NA))),
#                 o.ch.polarity = ifelse(WT_AA %in% polarity$polar,
#                                        "polar",
#                                        ifelse(WT_AA %in% polarity$nonpolar,
#                                               "nonpolar",
#                                               NA)),
#                 n.ch.polarity = ifelse(AA_Mut %in% polarity$polar,
#                                        "polar",
#                                        ifelse(AA_Mut %in% polarity$nonpolar,
#                                               "nonpolar",
#                                               NA)),
#                 o.ch.pv = ifelse(WT_AA %in% pv$special,
#                                  "special",
#                                  ifelse(WT_AA %in% pv$neut_small,
#                                         "neut_small",
#                                         ifelse(WT_AA %in% pv$polar_small,
#                                                "polar_small",
#                                                ifelse(WT_AA %in% pv$polar_large,
#                                                       "polar_large",
#                                                       ifelse(WT_AA %in% pv$nonpolar_small,
#                                                              "nonpolar_small",
#                                                              ifelse(WT_AA %in% pv$nonpolar_large,
#                                                                     "nonpolar_large",
#                                                                     NA)))))),
#                 n.ch.pv = ifelse(AA_Mut %in% pv$special,
#                                  "special",
#                                  ifelse(AA_Mut %in% pv$neut_small,
#                                         "neut_small",
#                                         ifelse(AA_Mut %in% pv$polar_small,
#                                                "polar_small",
#                                                ifelse(AA_Mut %in% pv$polar_large,
#                                                       "polar_large",
#                                                       ifelse(AA_Mut %in% pv$nonpolar_small,
#                                                              "nonpolar_small",
#                                                              ifelse(AA_Mut %in% pv$nonpolar_large,
#                                                                     "nonpolar_large",
#                                                                     NA))))))
#   ) %>% mutate(v.o.charge = o.charge == o.ch.charge,
#                v.n.charge = mut.charge == n.ch.charge,
#                v.o.polarity = o.polarity == o.ch.polarity,
#                v.n.polarity = mut.polarity == n.ch.polarity,
#                v.o.pv = o.pv == o.ch.pv,
#                v.n.pv = mut.pv == n.ch.pv)
#   list(v.o.charge = count(test,.(v.o.charge)),
#        v.o.polarity = count(test,.(v.o.polarity)),
#        v.o.pv = count(test,.(v.o.pv)),
#        v.n.charge = count(test,.(v.n.charge)),
#        v.n.polarity = count(test,.(v.n.polarity)),
#        v.n.pv = count(test,.(v.n.pv))) #trues means the annotated and the check versions are the same. true is good.
# } #v.o.class means verifcation whether WT aa was annotated with its correct class, true means correct.  v.n.class means verficiation whether mut aa was annotated with its correct class-should only be true.
# 
# ch.1(np.annot)
# ch.1(np2.annot)
# ch.1(ha.annot)
# ch.1(env.annot)
# 
# 
# ch.1.s(np.annot)
# ch.1.s(np2.annot)
# ch.1.s(ha.annot)
# ch.1.s(env.annot)
# 
# ##check whether class change annotation C_nclass was correct
# 
# ch.1 = function(dat){
#   test = dat %>% subset(WT_AA != AA_Mut,select = c(WT_AA,AA_Mut,C_ncharge,C_npolarity,C_npv))
# 
#   test = mutate(test,
#                 o.ch.charge = ifelse(WT_AA %in% charge$pos,
#                                      "pos",
#                                      ifelse(WT_AA %in% charge$neg,
#                                             "neg",
#                                             ifelse(WT_AA %in% charge$neut,
#                                                    "neut",
#                                                    NA))),
#                 n.ch.charge = ifelse(AA_Mut %in% charge$pos,
#                                      "pos",
#                                      ifelse(AA_Mut %in% charge$neg,
#                                             "neg",
#                                             ifelse(AA_Mut %in% charge$neut,
#                                                    "neut",
#                                                    NA))),
#                 o.ch.polarity = ifelse(WT_AA %in% polarity$polar,
#                                        "polar",
#                                        ifelse(WT_AA %in% polarity$nonpolar,
#                                               "nonpolar",
#                                               NA)),
#                 n.ch.polarity = ifelse(AA_Mut %in% polarity$polar,
#                                        "polar",
#                                        ifelse(AA_Mut %in% polarity$nonpolar,
#                                               "nonpolar",
#                                               NA)),
#                 o.ch.pv = ifelse(WT_AA %in% pv$special,
#                                  "special",
#                                  ifelse(WT_AA %in% pv$neut_small,
#                                         "neut_small",
#                                         ifelse(WT_AA %in% pv$polar_small,
#                                                "polar_small",
#                                                ifelse(WT_AA %in% pv$polar_large,
#                                                       "polar_large",
#                                                       ifelse(WT_AA %in% pv$nonpolar_small,
#                                                              "nonpolar_small",
#                                                              ifelse(WT_AA %in% pv$nonpolar_large,
#                                                                     "nonpolar_large",
#                                                                     NA)))))),
#                 n.ch.pv = ifelse(AA_Mut %in% pv$special,
#                                  "special",
#                                  ifelse(AA_Mut %in% pv$neut_small,
#                                         "neut_small",
#                                         ifelse(AA_Mut %in% pv$polar_small,
#                                                "polar_small",
#                                                ifelse(AA_Mut %in% pv$polar_large,
#                                                       "polar_large",
#                                                       ifelse(AA_Mut %in% pv$nonpolar_small,
#                                                              "nonpolar_small",
#                                                              ifelse(AA_Mut %in% pv$nonpolar_large,
#                                                                     "nonpolar_large",
#                                                                     NA))))))
#   ) %>% mutate(ch.c.charge = o.ch.charge == n.ch.charge,
#                ch.c.polarity = o.ch.polarity == n.ch.polarity,
#                ch.c.pv = o.ch.pv == n.ch.pv) %>%
#     mutate(v.c.charge = C_ncharge == ch.c.charge,
#            v.c.polarity = C_npolarity == ch.c.polarity,
#            v.c.pv = C_npv == ch.c.pv)
#   list(v.c.charge = count(test,.(v.c.charge)),
#        v.c.polarity = count(test,.(v.c.polarity)),
#        v.c.pv = count(test,.(v.c.pv))) #trues means the annotated and the check versions are the same. true is good.
# } #v.o.class means verifcation whether WT aa was annotated with its correct class, true means correct.  v.n.class means verficiation whether mut aa was annotated with its correct class
# ch.1.s = function(dat){
#   test = dat %>% subset(WT_AA == AA_Mut,select = c(WT_AA,AA_Mut,C_ncharge,C_npolarity,C_npv))
#   count(test,.(C_ncharge,C_npolarity,C_npv))
# } #should only see S for synonymous
# 
# ch.1(np.annot)
# ch.1(np2.annot)
# ch.1(ha.annot)
# ch.1(env.annot)
# 
# 
# ch.1.s(np.annot)
# ch.1.s(np2.annot)
# ch.1.s(ha.annot)
# ch.1.s(env.annot)
# 
# ###okay good all checks out, also same numbers as above so checks out.
# 
# 
# 
# #2) Only true NS mutations used for analysis?----
# ##a) Any S mutations in all studies annotated wrongly?
# 
# test = subset(all_studies, grepl("Bloom",Study))
# 
# test %<>% mutate(Position = gsub("[a-zA-Z]*","",AA_Sub),
#                                                                            Weird_Len = ifelse((gsub("[0-9]*","",AA_Sub) %>% nchar) >2,T,F),
#                                                                            WT_AA = gsub("([0-9]*)([a-z]*)[A-Z]$","",AA_Sub),
#                                                                            Mut_AA = gsub("^[A-Z]([0-9]*)[a-z]*","",AA_Sub),
#                                                                            ch.NS = WT_AA == Mut_AA)
# count(test,.(S_NS,ch.NS)) #should only see NS and FALSE, S and TRUE
# #Okay great, all bloom annotated in all_studies as NS are truly NS and all S are truly S.
# 
# ##note some of the AA_Subs due to position having letters position 142 has a-e. Okay. This is for Env and it has no effects on analysis.
# 
# ##c) Correct Number of S, NS in all_studies? Here should only see in "test" S-S, TS-NS, TV-NS and these should be the same numbers as in the annot TS_TV-S S_NS-TRUE, TS_TV-TRUE S_NS-FALSE, TS_TV-FALSE S-NS-FALSE, respectively, and there should be only two more combinations in the annots TS-TV-BOTH S-NS-FALSE, MUT23 FALSE.
# 
# test %>% subset(Dataset == "NP") %>% count(.(TS_TV,S_NS))
# count(np.annot,.(TS_TV,S_NS)) #yes same numbers
# 
# test %>% subset(Dataset == "NP2") %>% count(.(TS_TV,S_NS))
# count(np2.annot,.(TS_TV,S_NS)) #ns checks out perfectly.
# 
# test %>% subset(Dataset == "env") %>% count(.(TS_TV,S_NS))
# count(env.annot,.(TS_TV,S_NS)) #NS: checks out perfectly S: checks out perfectly
# 
# test %>% subset(Dataset == "HA") %>% count(.(TS_TV,S_NS))
# count(ha.annot,.(TS_TV,S_NS)) #NS: checks out perfectly. S: checks out perfectly
# 
# #end----
```
Everything checks out.


##Power Simulations
Simulate power of smaller studies (F1,qb,VSV,TEV) by drawing randomly their N fitness values from our study x1000 and then directly calculating AUC from that or creating a normal distribution around each fitness value using the respective study's average fitness standard deviation and drawing from that distribution to then calculate AUC (should come up with pretty much the same power each way).  Also create a null distribution of TS-TV bias by randomizing the TS-TV labels among the fitness values in our study x1000 and calculating AUC. Power for each study can be either the number of times the simulation detected an AUC higher than 5% of the highest AUC calculated in the null distribution divided by 1000, or simply the number of times the simulation detected a .05 p-value for AUC divided by 1000. I used the slightly easier way to code it for paper, which is the latter. I also performed an additional simulation for studies that had info on the measurement error-in addition to drawing mutations randomly, I created a normal distribution around the mean of each drawn mutation and drew a value from that distribution, to kind of simulate effect of measurement error on power, this is not shown in paper either.

I calculated power for each of the different AUC calculations from section "AUC-All and Viable Detrimental Missense Mutations", matching the N of each study for that calculation type. Did with and without stop mutations, only without stop mutations shown in paper.
```{r POWER.CALC}
#sim fnxs----
n_any = function(studies, Set = "all", Stops = "INCLUDE", xGenes = c("NONE"), xDatasets = c("NONE"), only_Genes = c("NONE")){
  n_ts = sub(studies, Set,Stops,xGenes,xDatasets,only_Genes)  %>% .$TS_TV %>% grepl("TS",.) %>% sum()
  n_tv = sub(studies, Set,Stops,xGenes,xDatasets,only_Genes)  %>% .$TS_TV %>% grepl("TV",.) %>% sum()
  c(n_ts,n_tv)
}

n_for_sims = function(study, Set){
  n_ts = subset(each,Study == study & set == Set, select = N_TS) %>% unlist %>% as.numeric
  n_tv = subset(each,Study == study & set == Set, select = N_TV) %>% unlist %>% as.numeric
  c(n_ts, n_tv)
}

avg_sd = function(studies, Set = "all", Stops = "INCLUDE", xGenes = c("NONE"), xDatasets = c("NONE"), only_Genes = c("NONE")){
  sub(studies, Set,Stops,xGenes,xDatasets,only_Genes)  %>% .$Fitness.SD %>% as.numeric %>% mean(na.rm = T)
  #don't need to worry about xdatasets except for obvi lauring (vsv is one with described dataset, but it doesn't have sd and don't know n of ts or tv (though could use total n--but it's total n is same as qbeta so whatever don't worry about it.))
} 

n_sim = function(data_frame,N_sims = 1000, n_ts, n_tv){
  nsim = sapply(1:N_sims,function(sim){
    set.seed(sim)
    #pick fitness values randomly
    ts = subset(data_frame, TS_TV == "TS", select = Fitness.Mean) %>%
      .[sample(nrow(.),n_ts),] %>%
      as.numeric
    tv = subset(data_frame, TS_TV == "TV", select = Fitness.Mean) %>%
      .[sample(nrow(.),n_tv),] %>%
      as.numeric
    #calculate stats from vectors
    stats_sim = wilcox.test(tv,ts,alternative = "less", conf.int = F)
    stats_sim[c("N_TS","N_TV", "auc")] = c(n_ts,n_tv, stat2auc(n_ts,n_tv,stats_sim$statistic))
    stats_sim
  })
  nsim %>% t() %>% as.data.frame()
}

n_sd_sim = function(data_frame,N_sims = 1000, n_ts, n_tv, sd){
  n_sd_sim = sapply(1:N_sims,function(sim){
    set.seed(sim)
    #pick fitness values randomly from N
    ts = subset(data_frame, TS_TV == "TS", select = Fitness.Mean) %>%
      .[sample(nrow(.),n_ts),] %>%
      as.numeric
    tv = subset(data_frame, TS_TV == "TV", select = Fitness.Mean) %>%
      .[sample(nrow(.),n_tv),] %>%
      as.numeric
    #randomly pick fitness values from SD around mean
    ts = rnorm(ts,ts,sd)
    tv = rnorm(tv,tv,sd)
    #calculate stats from vectors
    stats_sim = wilcox.test(tv,ts,alternative = "less", conf.int = F)
    stats_sim[c("N_TS","N_TV", "auc")] = c(n_ts,n_tv, stat2auc(n_ts,n_tv,stats_sim$statistic))
    stats_sim
  })
  n_sd_sim %>% t() %>% as.data.frame()
}

#pipe fnxs----
study2auc = function(study = "Lauring", Set = "all", study_stops = "INCLUDE", xGenes = c("NONE"), xDatasets = c("NONE"), only_Genes = c("NONE"),
                     sim_study = "xxx",
                     nsims = 1000, n_ts = "xxx", n_tv = "xxx",
                     n_sd_sim = "n_sim",
                     study_name = study
                     ){
  if(sim_study != "xxx"){
    n_ts = n_for_sims(sim_study,Set)[1]
    n_tv = n_for_sims(sim_study,Set)[2]
  }
  s = sub(study, 
          Set, 
          Stops = study_stops, 
          xGenes, 
          xDatasets,
          only_Genes)
  if(n_sd_sim == "n_sim"){
    n_sim(s,nsims,n_ts,n_tv) %>% 
      mutate(Study = study_name, set = Set, sim_study = sim_study)
  } else{
    n_sd_sim(s,nsims,n_ts,n_tv,avg_sd(sim_study,Set=Set)) %>% 
      mutate(Study = study_name, set = Set, sim_study = sim_study)
  }
}


sim_sets = function(study = "Lauring", xGenes = c("NONE"), xDatasets = c("NONE"), only_Genes = c("NONE"),
                     sim_study = "xxx",
                     nsims = 1000, n_ts = "xxx", n_tv = "xxx",
                     n_sd_sim = "n_sim",
                     study_name = study #change accordingly if doing particular genes or datasets to name study (eg Lauring_gw)
                     ){
  l = lapply(sets4all, function(set){
    if(grepl("STOP",set)){study_stops = "stop"} else {study_stops = "INCLUDE"}
    study2auc(study = "Lauring", Set = set, study_stops, xGenes, xDatasets, only_Genes,
                     sim_study,
                     nsims, n_ts, n_tv,
                     n_sd_sim,
                     study_name
                     )
})
  do.call(rbind,l)
}

sets4all = c("all","all_xSTOP","detrimental",'detrimental_xSTOP',"viable",'viable_xSTOP',"detrimental & viable",'detrimental & viable_xSTOP')


#null distribution----
vectors2auc = function(ts,tv){
  n_ts = length(ts)
  n_tv = length(tv)
  stats = wilcox.test(tv,ts,alternative = "less", conf.int = F)
  stats[c("N_TS","N_TV", "auc")] = c(n_ts,n_tv, stat2auc(n_ts,n_tv,stats$statistic))
  stats
}


null_sim = function(N_sims = 1000, studies = "Lauring", Set = "all", Stops = "INCLUDE", xGenes = c("NONE"), xDatasets = c("NONE"), only_Genes = c("NONE"), study_name = studies){
  sapply(1:N_sims,function(sim){
    set.seed(sim)
    #make ts tv vectors
    f = sub(studies, Set, Stops, xGenes, xDatasets, only_Genes)$Fitness.Mean %>% unlist %>% as.numeric
    ts = sample(1:length(f),n_any(studies, Set, Stops, xGenes, xDatasets, only_Genes)[1])
    tv = f[-c(ts)]
    ts = f[ts]
    #calc stats
    vectors2auc(ts,tv)
  }) %>% t %>% as.data.frame() %>% mutate(Study = study_name, set = Set)
}

null_sim_sets = function(study = "Lauring", xGenes = c("NONE"), xDatasets = c("NONE"), only_Genes = c("NONE"),
                     sim_study = "xxx",
                     nsims = 1000,
                     study_name = study #change accordingly if doing particular genes or datasets to name study (eg Lauring_gw)
                     ){
  l = lapply(sets4all, function(set){
    if(grepl("STOP",set)){study_stops = "stop"} else {study_stops = "INCLUDE"}
    null_sim(N_sims = nsims, studies = study, Set = set, Stops = study_stops, xGenes, xDatasets, only_Genes, study_name = study_name)
})
  do.call(rbind,l)
} 

null = null_sim_sets() %>% mutate(sim = "null", sim_study = "null")

#power fnxs----
##calc by number of times found p value of less than a particular value. (not calculated off of null distribution, should be about the same though).
power_calc = function(data_frame = n_sims, Set, Sim_study, p = .05, nsims = 1000){
  subset(data_frame, set == Set & sim_study == Sim_study & p.value < p) %>% dim(.) %>% .[1]/nsims
}

power_sets = function(data_frame = n_sims,Sim_study, p = .05, nsims = 1000){
  sapply(sets4all,function(set){
    power_calc(data_frame,Set = set, Sim_study, p, nsims)
  })
  
}

q1_n_sim_studs = c("Lauring_HANA","Lauring_xHAxNA","f1_Peris_10","phix174_Domingo_Calap_09","qb_Domingo_Calap_09","TEV_Carrasco_07_full","VSV")

# c("f1_Peris_10","TEV_Carrasco_07_full","qb_Domingo_Calap_09","VSV") #previous

q1_sd_sim_studs = c("f1_Peris_10","qb_Domingo_Calap_09")

power_studies_sets = function(data_frame = n_sims, p = .05, nsims = 1000, sim_studs = q1_n_sim_studs){
  sapply(sim_studs, function(s){
    power_sets(data_frame, Sim_study = s, p, nsims)
  })
}


#all sims----
n_sims = lapply(q1_n_sim_studs,function(s){
  sim_sets(sim_study = s)
}) %>% do.call(rbind,.) %>%  mutate(sim = "n_sim")

sd_sims = rbind(sim_sets(sim_study = "f1_Peris_10", n_sd_sim = "sd_sim"),
               sim_sets(sim_study = "qb_Domingo_Calap_09",n_sd_sim = "sd_sim")
               ) %>% mutate(sim = "sd_sim") 



##full q1 dataframe
q1 = rbind(null, n_sims, sd_sims)

#plot sims----
##get things in right order
q1$set = factor(q1$set, levels = c("all","all_xSTOP","detrimental","detrimental_xSTOP","viable","viable_xSTOP","detrimental & viable","detrimental & viable_xSTOP"))
q1$sim_study = factor(q1$sim_study, levels = c("null",q1_n_sim_studs))
q1$sim = factor(q1$sim, levels = c("null","n_sim","sd_sim"))

# c("null","f1_Peris_10","TEV_Carrasco_07_full","VSV","qb_Domingo_Calap_09")) # previous

##significant auc values for horizontal line
sig_auc_value = function(Set,p = .05){
  auc = q1 %>% subset(sim == "null" & set == Set) %>% .$auc %>% as.numeric
  auc[rank(auc)/sum(length(auc)) > 1-p] %>% min()
}
auc05 = sapply(sets4all, function(set){
  sig_auc_value(Set = set)
  })

actual_auc = sapply(sets4all,function(Set){
  subset(each, set == Set & Study == "Lauring", select = auc) %>% as.numeric
})

q1 = mutate(q1, sig_auc = ifelse(set == "all",auc05["all"],
                               ifelse(set == "all_xSTOP",auc05["all_xSTOP"],
                                      ifelse(set == "detrimental",auc05["detrimental"],
                                             ifelse(set == "detrimental_xSTOP",auc05["detrimental_xSTOP"],
                                                    ifelse(set == "viable",auc05["viable"],
                                                           ifelse(set == "viable_xSTOP",auc05["viable_xSTOP"],
                                                                  ifelse(set == "detrimental & viable",auc05["detrimental & viable"],
                                                                         ifelse(set == "detrimental & viable_xSTOP",auc05["detrimental & viable_xSTOP"],
                                                                                "x")))))))),
            actual_auc = ifelse(set == "all",actual_auc["all"],
                               ifelse(set == "all_xSTOP",actual_auc["all_xSTOP"],
                                      ifelse(set == "detrimental",actual_auc["detrimental"],
                                             ifelse(set == "detrimental_xSTOP",actual_auc["detrimental_xSTOP"],
                                                    ifelse(set == "viable",actual_auc["viable"],
                                                           ifelse(set == "viable_xSTOP",actual_auc["viable_xSTOP"],
                                                                  ifelse(set == "detrimental & viable",actual_auc["detrimental & viable"],
                                                                         ifelse(set == "detrimental & viable_xSTOP",actual_auc["detrimental & viable_xSTOP"],
                                                                                "x"))))))))
            )
  
#power table----
power_n_sim = power_studies_sets()
power_sd_sim = power_studies_sets(data_frame = sd_sims, sim_studs = q1_sd_sim_studs)

#end----
```

##Lauring-Bloom Correlations (Spearman)
```{r CORR_BLOOM_LA}
library(Hmisc)
#prep bloom----
#prep data for all mutants bloom-in all studies I don't have TS_TV both, though some of those boths I can correlate with the same AA mutation from our study. I do use synonymous mutations, see note below chunk.
prep.bloom.corr = function(df,dtset,study,gene){
  df %>% 
    subset(TS_TV != "MUT23") %>% 
    rename(c(Rel_site_pref = "Fitness.Mean",X.SITE = "Position")) %>%
    mutate(Mutation = NA,
           AA_Sub = paste0(WT_AA,Position,AA_Mut),
           Fitness.SD = NA,
           Dataset = dtset,
           TS_TV = mapvalues(TS_TV,c("TRUE","FALSE","S"),c("TS","TV","S")), #it actually works with T, F as logical  only and without "S" but I think it's probably better to be specific.
           S_NS = mapvalues(S_NS,c(T,F),c("S","NS")),
           Gene = gene,
           Fitness.CV = NA,
           Study = study,
           N = NA
    ) %>%
    subset(select = c(Mutation,AA_Sub,Fitness.Mean,Fitness.SD,Dataset,Position,TS_TV,S_NS,Gene,Fitness.CV,Study,N))

} #tstv == MUT23 are those aa muts that cannot be reached by single mutations, we don't have those in our analysis so exclude here.

np.corr = prep.bloom.corr(np.annot,"NP","Bloom","NP")
np2.corr = prep.bloom.corr(np2.annot,"NP2","Bloom","NP")
ha.corr = prep.bloom.corr(ha.annot,"HA","Bloom","HA")
env.corr = prep.bloom.corr(env.annot,"env","Bloom_env","env")

#find shared aa----
all_studies2 = all_studies %>% mutate(AA_Sub = ifelse(Study == "Lauring" & S_NS == "S",
                                       paste0(gsub("[0-9]+","",gsub(" \\(SYN)","",AA_Sub)),
                                              gsub("[A-Z]","",gsub(" \\(SYN)","",AA_Sub)),
                                              gsub("[0-9]+","",gsub(" \\(SYN)","",AA_Sub))),
                                       AA_Sub
                                       )
                       ) #make syn aa-sub notation in lauring the same as in bloom eg A1A

AAshared = function(bloom.df,la.df,gene){
  (bloom.df %>% subset(Study == "Bloom" & Gene == gene) %>% .$AA_Sub)[(bloom.df %>% subset(Study == "Bloom" & Gene == gene) %>% .$AA_Sub) %in% (la.df %>% subset(Study == "Lauring" & Gene == gene) %>% .$AA_Sub)]
}

info_AAshared = function(study,gene,aashared,data_frame){
  data_frame %>% subset(Study == study & Gene == gene & AA_Sub %in% aashared, select = c(AA_Sub,Fitness.Mean,if(study == "Lauring"){c(Gene,S_NS)})) %>% arrange(AA_Sub) %>% rename(c(Fitness.Mean = paste0("Fit.",study),AA_Sub = paste0("AA_Sub.",study)))
}


shared_bloom_la = cbind(info_AAshared("Lauring","HA",AAshared(ha.corr,all_studies2,"HA"),all_studies2),
                        info_AAshared("Bloom","HA",AAshared(ha.corr,all_studies2,"HA"),ha.corr)
) %>%
  rbind(cbind(
    info_AAshared("Lauring","NP",AAshared(np.corr,all_studies2,"NP"),all_studies2),
    info_AAshared("Bloom","NP",AAshared(np.corr,all_studies2,"NP"),np.corr)
  )) %>%
  mutate(Dataset = Gene) %>%
  rbind(cbind(
    info_AAshared("Lauring","NP",AAshared(np2.corr,all_studies2,"NP"),all_studies2),
    info_AAshared("Bloom","NP",AAshared(np2.corr,all_studies2,"NP"),np2.corr)
  ) %>% 
    mutate(Dataset = "NP2")) %>%
  rbind(np.corr %>% subset(AA_Sub == "D372V",select = c(AA_Sub,Fitness.Mean,S_NS)) %>% rename(AA_Sub = "AA_Sub.Bloom",Fitness.Mean = "Fit.Bloom") %>% mutate(Dataset = "NP", Gene = "NP") %>% 
          cbind(all_studies %>% subset(Study == "Lauring" & Gene == "NP" & AA_Sub == "E372V", select = c(AA_Sub,Fitness.Mean)) %>% rename(AA_Sub = "AA_Sub.Lauring",Fitness.Mean = "Fit.Lauring")
                )
  )

#corrs----
rcorr.df = function(df,cols = 1:ncol(df),rows = 1:nrow(df),type = "spearman"){
  df[rows,cols] %>% as.matrix %>% rcorr(type = type)
} #to useful fnx: performs correlations with rcorr given column names or indices and row names or indicies. Defaults are all columns, all rows.

cols = c("Fit.Lauring","Fit.Bloom")

rcorr.extract = function(corr,corr1,corr2){
  c(n = corr$n[corr1,corr2],r = corr$r[corr1,corr2],p = corr$P[corr1,corr2])
} #to useful fnx: extract relevant values from correlation (n, r, p) for variables that I want (corr1,corr2), variables can be names or numbers as positions in the correlation matrix. Not sure yet how it works with multiple variables...

rcorr2df = function(DF,corr_name,sns, Cols = cols,Rows = 1:nrow(DF),Type = "spearman",corr1 = cols[1], corr2 = cols[2]){
  data.frame(Corr = corr_name, S_NS = sns) %>% cbind(rcorr.df(DF,Cols,Rows,Type) %>% rcorr.extract(.,corr1,corr2) %>% as.matrix %>% t() %>% as.data.frame)
}

bloomxla_corr = rbind(rcorr2df(shared_bloom_la,"All","Both"),
      rcorr2df(shared_bloom_la %>% subset(S_NS == "NS"),"All","NS"),
      rcorr2df(shared_bloom_la %>% subset(Gene == "HA"|Dataset == "NP2"),"H1N1","Both"),
      rcorr2df(shared_bloom_la %>% subset((Gene == "HA"|Dataset == "NP2") & S_NS == "NS"),"H1N1","NS"),
      rcorr2df(shared_bloom_la %>% subset(Gene == "HA"),"HA_H1N1","Both"),
      rcorr2df(shared_bloom_la %>% subset(Gene == "HA" & S_NS == "NS"),"HA_H1N1","NS"),
      rcorr2df(shared_bloom_la %>% subset(Gene == "NP" & Dataset == "NP2"),"NP_H1N1","NS"),
      rcorr2df(shared_bloom_la %>% subset(Gene == "NP" & Dataset == "NP"),"NP_H3N2","NS")
      ) %>% mutate(prev_r = c(NA,NA,NA,NA,.66,NA,.60,.26),prev_p = c(NA,NA,NA,NA,.001,NA,0.0450,0.4031))

detach("package:Hmisc", unload = T)
#end----
```

Bloom Studies:
i) NP is Aichi/1968 NP, an H3N2
ii)NP2 is A/Puerto Rico/8/1934 (H1N1, PR8)
ii) HA is A/WSN/1933 (H1N1, WSN33)

N for Correlations:
HA: 31 mutations that are shared (Lauring study has 32 total HA mutations, but 1 is a non-coding mutation thus not mutated in bloom)
NP and NP2: 12, see note for NP below though.

Note on NP correlation: WT E372 in Lauring (H1N1, WSN33) and NP2 (H1N1, PR8) is actually WT D372 in NP (H3N2).  This corresponds to a 1 nucleotide change between these strains from GAG to GAT.  I checked this with the sequences from these strains I have as well. Included the D372V mutation from NP (Lauring has E372V) here to do correlations. Even though it's not from the same WT, it's to the same mutant.


Note on synonymous mutations: Using synonymous mutations in bloom that are normalized to themselves ("fitness" of 1) to do correlations effectively assumes that all nucleotide mutations that make up that synonymous mutation in the mutational scan are nuetral, while in our data that's not true. It still increases the fit, however, because the synonymous mutations in our data are close to neutral.  Anyway, we didn't have any synonymous mutations in NP only about 3 I think in HA. Paper is shown without using synonymous mutations.


##Ts vs. Tv Odds of Being Radical
```{r TsTv.Odds.Radical}

class.counts = function(df){
  lapply(sets.type,function(class){
    lapply(c(TS_TV = "TS_TV",S_NS = "S_NS"),function(mut){
      count(df,vars = c(class,mut))
    })
  })
}
datasets =  c(NP.H3N2 = "NP.H3N2", NP.H1N1= "NP.H1N1", HA.H1N1 = "HA.H1N1", HIV.ENV = "HIV.ENV")

counts.bloom = lapply(datasets,function(dataset){
  class.counts(df = dfs %>% subset(dtset == dataset))
})


mk_cont = function(data_frame1 = "null",gene,Class){
  m = matrix(data = rep(NA,4),
                nrow = 2,
                ncol = 2, 
                byrow = T,
                dimnames = list(c("TV","TS"),c("R","C"))
  )
  
  df1 = counts.bloom[[gene]][[Class]][["TS_TV"]]
  
  m["TV","R"] = df1[df1[,1] == F,] %>% subset(TS_TV == F) %>% .$freq
  m["TV","C"] = df1[df1[,1] == T,] %>% subset(TS_TV == F) %>% .$freq
  
  m["TS","R"] = df1[df1[,1] == F,] %>% subset(TS_TV == T) %>% .$freq
  m["TS","C"] = df1[df1[,1] == T,] %>% subset(TS_TV == T) %>% .$freq
  
  m
}

cont2stat = function(m,gene,Class){
  test = fisher.test(m, conf.int = T)
  data.frame(Gene = gene,
             Comparison = paste("TV","TS",sep = " vs. "),
             Class = Class,
             pvalue = test$p.value,
             OR = test$estimate,
             OR_CI95_lower = test$conf.int[1],
             OR_CI95_upper = test$conf.int[2],
             row.names = NULL
             )
}


data2stat = function(data_frame = "null",gene,Class){
  mk_cont(data_frame,gene,Class) %>% cont2stat(gene,Class)
}

class.genes2stat = function(data_frame = "null",genes,classes = Classes){
  lapply(genes,function(g){
    lapply(classes,function(c){
      data2stat(data_frame = data_frame,gene = g,Class = c)
    }) %>% do.call(rbind,.)
  }) %>% do.call(rbind,.)
}


Classes = c(charge = "charge", polarity = "polarity", pv = "pv")
g1s = datasets


##calc stats
tstv.rc = rbind(class.genes2stat(genes = g1s))

```

##Ts:Tv Mutation Rate and Substitution Rate Bias
Use our and bloom's mutation rate data to calculate Ts:Tv mutation rate bias and average the observed influenza Ts:Tv substitution ratio from Duchêne et al.

Duchêne S, Ho SY, Holmes EC. 2015. Declining transition/transversion ratios through time reveal limitations to the accuracy of nucleotide substitution models. BMC Evolutionary Biology 15:36.

```{r TSTV.BIAS}

#Load & Parse Data----
snc = read.csv("tstv.snc.pauly.csv", stringsAsFactors = F) #load mutation rate data from our lab

sub.rate = read.csv("tstv.sub.bias.duchene.csv", stringsAsFactors = F) #load substitution rate data from Duchene et. al. as cited in manuscript

snc %<>% rename(c("A.Puerto.Rico.8.1934.H1N1" = "pr8","A.Hong.Kong.4801.2014.H3N2" = "hk.h3n2")) %>% mutate(Mutation.class = gsub(" to ","",Mutation.class), pr8 = gsub(" x 10","e",pr8) %>% gsub(" ±.*","",.) %>% as.numeric,hk.h3n2 = gsub(" x 10","e",hk.h3n2) %>% gsub(" ±.*","",.) %>% as.numeric) %>% subset(Mutation.class != "Overallb" & Mutation.class != "",select = c(Mutation.class,pr8,hk.h3n2)) #parse

snc %<>% mutate(ts.tv = ifelse(Mutation.class %in% ts_muts,"Ts","Tv"))

avg.tsTOtv.mut.rate.lauring = ddply(snc,.(ts.tv),summarize, avg.pr8 = mean(pr8),avg.hk.h3n2 = mean(hk.h3n2)) %>% t() %>% as.data.frame %>% rename(c("V1" = "Ts","V2" = "Tv")) %>% subset(Ts != "Ts") %>% mutate(strain = c("pr8","hk.h3n2"), Ts = Ts %>% as.vector %>% as.numeric, Tv = Tv %>% as.vector %>% as.numeric) %>% mutate(tsTotv = Ts/(Tv*2)) #calc ts to tv ratio


avg.sub.rate = sub.rate %>% subset(Genus == "Influenzavirus", select = c(Genus,ti.tv)) %>% .$ti.tv %>% mean

sng = read.csv("bloom.mut.rate.est.csv",stringsAsFactors = F) #load mutation rate data from bloom 2014 paper. It is substitions per nucleotide per tissue culture generation
sng %<>% mutate(Mutation.Type = gsub(".*\\(","",Mutation.Type) %>% gsub(").*","",.), Rate = Rate %>% as.vector %>% as.numeric) %>% subset(grepl("tran",Mutation.Type),select = -c(X)) 

avg.tsTOtv.mut.rate.bloom = ddply(sng,~Mutation.Type,summarize,average = mean(Rate)) %>% t() %>% as.data.frame %>% rename(c("V1" = "Ts","V2" = "Tv")) %>% subset(Ts != "transition" & Tv != "transversion")  %>% mutate(Ts = Ts %>% as.vector %>% as.numeric, Tv = Tv %>% as.vector %>% as.numeric) %>% mutate(tsTotv = Ts/(Tv*2)) #calc ts to tv ratio for bloom data.

```

##Misc.
Some miscellaneous calculations for paper.
```{r MISC}
med.per.gene = rbind(all_studies %>% subset(grepl("Lauring",Study) & S_NS == "NS" & !grepl("Stop|stop",AA_Sub)),all_studies %>% subset(AA_Sub == "NS1 E208stop, NEP M50I")) %>% count(.(Gene)) %>% .$freq %>% median #median number of missense mutations per gene 
```

#Figures
##Setup
```{r SETUP}
# library(extrafont)
knitr::opts_chunk$set(echo = T, warnings = FALSE, message = F)
# detach("package:seqinr", unload = T)


library(knitr)
library(plyr)
library(reshape2)
library(magrittr)
library(ggplot2)
library(ggrepel)
library(Biostrings)
library(cowplot)
setwd(analysis.path)




```


##Main Figures & Tables
1) Fig. 1: AUC Thresholds - Influenza
3) Fig. 2: Fisher & CDF -  Influenza & HIV
4) Table 1: Spearman - Bloom vs. Lauring
5) Fig. 3: Fisher & CDF - TsTv - Bloom
6) Fig. 4: Fisher & CDF - C/R
7) Table 2: Fisher TsTv OR for C/R
8) Fig. 5 & 6: Fisher & CDF - TsTv Among C/R

###1) Fig. 1: AUC Thresholds - Influenza
```{r FIG1}
#names, mapping, factoring fnxs----
##added as I realized different names were needed, that's why confusing.  But whenever need to check if names are correct from original to new df, check the levels of the dataframes, if the levels match up corretly, than map fnxs worked fine.
prev.names = c("NP.H3N2","NP.H1N1","HA.H1N1","HIV.ENV")
fr = c("Bloom_env","Bloom_HA","Bloom_NP","Bloom_NP2","NP_Aichi","NP_PR8","HA_WSN33","HIV_ENV", "HA_H1N1","NP_H1N1","NP_H3N2",prev.names)
fr.bl = c("Bloom_env","Bloom_HA","Bloom_NP","Bloom_NP2","NP_Aichi","NP_PR8","HA_WSN33","HIV_ENV", "HA_H1N1","NP_H1N1","NP_H3N2",prev.names)

to = c("HIV ENV","HA (H1N1)","NP (H3N2)","NP (H1N1)","NP (H3N2)","NP (H1N1)","HA (H1N1)","HIV ENV","HA (H1N1)","NP (H1N1)","NP (H3N2)","NP (H3N2)","NP (H1N1)","HA (H1N1)","HIV ENV")
to.names.bl = to

fr = c(fr,"Lauring_xSTOP","Lauring_HANA_xSTOP","Lauring_xHAxNA_xSTOP","HIV_xSTOP","HIV_IN_xSTOP","HIV_CA_xSTOP","f1_Peris_10_xSTOP","phix174_Domingo_Calap_09_xSTOP","qb_Domingo_Calap_09_xSTOP","TEV_Carrasco_07_full_xSTOP","VSV_xSTOP","Lauring","Lauring_HANA","Lauring_xHAxNA","HIV","HIV_IN","HIV_CA","Bloom_env","Bloom_HA","Bloom_NP","Bloom_NP2","f1_Peris_10","phix174_Domingo_Calap_09","qb_Domingo_Calap_09","TEV_Carrasco_07_full","VSV","TEV_Carrasco_07_sub")

to = c(to,"Total","Surface","Internal","HIV Combined","HIV IN","HIV CA","F1","Phix174","QB","TEV","VSV","Total","Surface","Internal","HIV Combined","HIV IN","HIV CA", "HIV ENV","HA (H1N1)","NP (H3N2)","NP (H1N1)","F1","Phix174","QB","TEV","VSV","TEV")

fr = c(fr,fr %>% gsub("_Peris_10|_Domingo_Calap_09|_Domingo_Calap_09|_Carrasco_07_full","",.))

to = c(to,to)

fr = c(fr,fr %>% gsub("_xSTOP","",.) %>% gsub("_Peris_10|_Domingo_Calap_09|_Domingo_Calap_09|_Carrasco_07_full","",.))

to = c(to,to)

other.levels = c("Total","Internal","Surface","HIV Combined","HIV IN","HIV CA","TEV","F1","VSV","QB","Phix174")
rh.la.levels = c("Total","Internal","Surface","HIV Combined","HIV IN","HIV CA")
bl.levels = c("NP (H3N2)","NP (H1N1)","HA (H1N1)","HIV ENV")
bl.flu.levels = bl.levels %>% remove_elements(elements = "ENV",whole = F)

gr.xstop = function(col){
  grepl("xSTOP",col)|col %in% fr.bl
} #subsets in only xSTOPs, use with subset

map.xstop = function(col){
  mapvalues(col,fr,to)
} #outputs the correct study names, use with mutate

#data----
fig1.dat = thresholds %>% subset(gr.xstop(Study) & !grepl("Lauring_gw|Bloom_Flu|Bloom_H1N1",Study)) %>% mutate(Study = map.xstop(Study)) %>% mutate(Study = factor(Study,levels = c(other.levels,bl.levels)))
fig1.dat$p.value = fig1.dat$p.value %>% as.vector %>% as.numeric

fig1.table = fig1.dat %>% subset(grepl("Total|Surface|Internal",Study), select = c(Study,threshold,auc,p.value)) %>% rename(c(auc = "AUC",Study = "Dataset",threshold = "Rel. Fitness Threshold"))

#pl & stats fnxs----
##stats fnxs----
bonf.corr = function(alpha=.05,nhypo){
  alpha/nhypo
}
  
  
##general pl fnxs----
global.font.size = 14
pl.lab.font.size = global.font.size
sub.pl = function(dat,study=NULL,muts=NULL,set=NULL,class=NULL){
#Study sub
  if(length(study) == 1){
    dat %<>% subset(Study == study)
  } #subset by a single study
  else if(length(study)>1){
    dat %<>% subset(Study %in% study)

  } #subset by multiple studies, if study==NULL, will use all studies

  #Mutations sub
  if(muts %>% length == 1){
    dat %<>% subset(Mutations == muts)
  }
  else if(muts %>% length > 1){
    dat %<>% subset(Mutations %in% muts)
  }

  #Set sub
  if(set %>% length == 1){
    dat %<>% subset(Set == set)
  }
  else if(set %>% length > 1){
    dat %<>% subset(Set %in% set)
  }
  #Class sub
  if((class %>% length == 1)){
    dat %<>% subset(Class == class)
  }
  else if(class %>% length > 1){
    dat %<>% subset(Class %in% class)
  }
  dat
}

xlims = function(dat,set){
  list(xmax = ifelse(set == "Detrimentals",1,max(dat$threshold)),xmin = ifelse(set == "Beneficials",1,0))
}

ylims = function(dat,study){
  l = dat %>% subset(Study %in% study) %>%.$OR %>% range %>% as.list
  names(l) = c("ymin","ymax")
  l$ymin = 0
  l$ymax = (l$ymax)*1.05
  l
} #use this only for fisher OR, min. should always be 0, max is the only thing that changes


xlims.cdf = function(set){
  list(xmax = ifelse(set == "Detrimentals",1,10),xmin = ifelse(set == "Beneficials",1,0))
}

xlabel.pl = function(study,fish.cdf){
  if(all(study %in% c(to.names.bl)|study == "bl")){
    if(fish.cdf == "fish"){
      "Relative Site Preference Cutoff"
    } 
    else{
      "Relative Site Preference"
    }
  }
  else if(fish.cdf == "fish"){
    "Relative Fitness Cutoff"
  }
  else{
    "Relative Fitness"
  }
}

sig.pl = function(dat,sig.level=.05,corr.method = "holm",nhypo = NULL){
  if(corr.method != "NONE"){
    
    if(corr.method == "bonf"){
      sig.level %<>% bonf.corr(nhypo)
    }
    
    else{
      subsets = dat %>% names
      
      for(x in c("p.val","N_TS","N_TV","auc","thresh","OR","total")){subsets = remove_elements(subsets,x,whole = F)}
    }
    
    if(corr.method == "holm"){
      dat %<>% ddply(subsets,mutate,p.value = p.adjust(p = p.value,method = "holm")) #calculate adjusted p values according to subset.
    } #for some reason stupid ddply wouldn't work with inputting corr.method as fnx variable, ketp saying object "corr.method" not found. I think it thinks its supposed to refer to dataframe column and for some reason doesn't look in wider environment.
    
  }
  
  dat %<>% mutate(sig = ifelse(p.value<sig.level,paste0("p<",sig.level),paste0("p>=",sig.level)))
  dat$sig %<>% factor(levels = c(paste0("p>=",sig.level),paste0("p<",sig.level))) #so levels goes no sig and then sig, thus in alpha (transparency) put 0,1
  dat
}

faceting = function(pl,Facet,Scales){
  if(Facet == "Study"){
    pl = pl+facet_grid(~Study,scales = Scales)
  }
  if(Facet ==  "Class"){
    pl = pl+facet_grid(~Class,scales = Scales)
  }
  if(Facet == "Set"){
    pl = pl+facet_grid(~Set,scales = Scales)
  }
  if(Facet == "Mutations"){
    pl = pl+facet_grid(~Mutations,scales = Scales)
  }
  if(Facet == "NULL"){
    pl = pl
  }
  pl
}

x.ticks = function(pl){
  pl = pl+scale_x_continuous(breaks = seq(0,1,.2),labels = c("0.0","0.2","0.4","0.6","0.8","1.0"),limits = c(0,1),expand = c(0.05, 0.01))
  pl
}

text.fnx = function(pl,off){
  if(off){
    pl = pl+
      theme(axis.title = element_blank())
  }
  pl
}

facet.label = function(pl,off){
  if(off){
    pl = pl+theme(strip.background = element_blank(),strip.text = element_blank())
  }
  pl
}

axes = function(pl,y.off,x.off){
  if(y.off){
    pl = pl+theme(axis.line.y = element_blank(),
                  axis.ticks.y = element_blank(),
                  axis.text.y = element_blank()) #element_text(color = "white")
  }
  if(x.off){
    pl = pl+theme(axis.line.x = element_blank(),
                  axis.ticks.x = element_blank(),
                  axis.text.x = element_blank())
  }
  pl
}

##fig1 pl fnx----
fig1.fnx = function(dat = fig1.dat,
                    sig.level = .05,
                    study = NULL,
                    title = study,
                    Facet = "Study",
                    Scales = "fixed",
                    text.off = T,
                    y.off = F,
                    x.off = F,
                    facet.label.off = F,
                    corr.method = "NONE",
                    nhypo = 10,
                    point.size = 4,
                    font.size = global.font.size,
                    rel.axis.size = .65){
  
  dat %<>% ddply(.(Study),mutate,n.ts = c(paste0("TS:",N_TS)[1:2], rep(NA,length(N_TS)-2)), n.tv = c(paste0("TV:",N_TV)[1:2], rep(NA,length(N_TV)-2))) #annotate N of first 2 data points (so lethals and viables)
  
  dat %<>% sub.pl(study)
  
  dat %<>% sig.pl(sig.level,corr.method,nhypo)
  
  pl = ggplot(data = dat, aes(x = threshold, y = auc, group = Study))+
    geom_point(stat = "identity", size = point.size)+
    geom_line()+
    geom_point(stat = "identity",
               size = point.size,
               shape =21,
               fill = "white")+ #plot outlines of all points.
    geom_point(stat = "identity",
               size = point.size,
               mapping  = aes(alpha = sig))+
    scale_alpha_manual(values =c(0,1))+ #plot solid fill for all sig points.
    ylim(0,1)+
    xlab("Relative Fitness Threshold")+
    ylab("AUC")+
    geom_hline(aes(yintercept = .5))+
    ggtitle(title)
  
  pl %<>% faceting(Facet,Scales)
  
  pl = pl+
    facet_grid(Study~.)+
    theme_cowplot(font_size = font.size)+
    theme(legend.position = "none",
          axis.text = element_text(size = rel(rel.axis.size))
    )
  # #can also use below for specific legend ctrl:
  # +guides(alpha = F)
  # +guides(linetype = F)
  
  pl %<>% x.ticks()
  pl %<>% text.fnx(text.off)
  pl %<>% axes(y.off,x.off)
  pl %<>% facet.label(facet.label.off)
  
  #return
  pl
}

smbe.fig1.fnx = function(dat = fig1.dat,
                    sig.level = .05,
                    study = NULL,
                    title = study,
                    Facet = "Study",
                    Scales = "fixed",
                    text.off = T,
                    y.off = F,
                    x.off = F,
                    facet.label.off = F,
                    corr.method = "NONE",
                    nhypo = 10,
                    point.size = 4,
                    font.size = global.font.size,
                    rel.axis.size = .65,
                    outline.size = 0){
  
  dat %<>% ddply(.(Study),mutate,n.ts = c(paste0("TS:",N_TS)[1:2], rep(NA,length(N_TS)-2)), n.tv = c(paste0("TV:",N_TV)[1:2], rep(NA,length(N_TV)-2))) #annotate N of first 2 data points (so lethals and viables)
  
  dat %<>% sub.pl(study)
  
  dat %<>% sig.pl(sig.level,corr.method,nhypo)
  
  pl = ggplot(data = dat, aes(x = threshold, y = auc, group = Study))+
    geom_point(stat = "identity", size = point.size)+
    geom_line()+
    geom_point(stat = "identity",
               size = point.size,
               shape =21,
               fill = "white",
               color = "black",
               stroke = outline.size)+ #plot outlines of all points.
    geom_point(stat = "identity",
               size = point.size,
               mapping  = aes(alpha = sig))+
    scale_alpha_manual(values =c(0,1))+ #plot solid fill for all sig points.
    ylim(0,1)+
    xlab("Relative Fitness Threshold")+
    ylab("AUC")+
    geom_hline(aes(yintercept = .5))+
    ggtitle(title)
  
  pl %<>% faceting(Facet,Scales)
  
  pl = pl+
    facet_grid(Study~.)+
    theme_cowplot(font_size = font.size)+
    theme(legend.position = "none",
          axis.text = element_text(size = rel(rel.axis.size))
    )
  # #can also use below for specific legend ctrl:
  # +guides(alpha = F)
  # +guides(linetype = F)
  
  pl %<>% x.ticks()
  pl %<>% text.fnx(text.off)
  pl %<>% axes(y.off,x.off)
  pl %<>% facet.label(facet.label.off)
  
  #return
  pl
}

#fig 1 pl----
fig1.total = fig1.fnx(study = "Total",title = NULL)
fig1.int = fig1.fnx(study = "Internal",title = NULL,text.off = F)+ylab("AUC")+xlab(NULL)
fig1.surf = fig1.fnx(study = "Surface",title = NULL)

fig1 = plot_grid(fig1.total,fig1.int,fig1.surf,nrow = 3, ncol = 1,align = "v")

fig1 = ggdraw()+draw_plot(fig1,0,.03,1,.97)+draw_label("Relative Fitness Threshold",x = .5,y=.025,size = 18)


pres.fig1 = fig1.fnx(study = c("Total","Internal","Surface"),title = NULL,text.off = F)+facet_grid(~Study)+xlab("Relative Fitness Threshold")

#smbe
fig1.smbe = smbe.fig1.fnx(study = "Internal",title = "Internal Proteins",text.off = F,facet.label.off = T,font.size = 28,point.size = 9,rel.axis.size = .85,outline.size = 1)

ggsave("fig1.smbe.png",fig1.smbe,width = 9.13,height =6.5, units = "in")


#end----
```

Filled circles are significant. (could use stars here, but to be consistent with bloom studies below where stars look terrible).

As St. & Norris did a one-tailed test for AUC, did a one-tailed test here as well. Two tailed for rest of fisher studies is good.


###2) Fig. 2: Fisher & CDF -  Influenza & HIV
```{r FIG2}
#data----
fig2.fish.dat = fisher.other.det %>% subset(gr.xstop(Study) & !grepl("Lauring_gw",Study)) %>% mutate(Study = map.xstop(Study)) %>% mutate(Study = factor(Study,c(other.levels,bl.levels)))


fig2.cdf.dat = cdf.df %>% subset(!grepl("Lauring_gw",Study) & Study != "Internal_Viable") %>% mutate(Study = map.xstop(Study)) %>% mutate(Study = factor(Study,c(other.levels,bl.levels)))

fig2.cdf.dat = fig2.cdf.dat %>% subset(Study == "HIV IN"|Study == "HIV CA") %>% mutate(Study = "HIV Combined") %>% rbind(fig2.cdf.dat) %>% mutate(Study = factor(Study,c(other.levels,bl.levels)))

fig2.table = fig2.fish.dat %>% subset(select = c(Study,threshold,OR,p.value)) %>% rename(c(Study = "Dataset",threshold = "Rel. Fitness Threshold",OR = "Odds Ratio"))

#pl fnx----
fig2.fish.fnx = function(dat = fig2.fish.dat,
                         sig.level = .05,
                         study = NULL,
                         bl.other = "other",
                         set = "Detrimentals",
                         title = study,
                         Group = "Study",
                         Facet = "Study", #otherwise use "none"
                         Scales = "fixed", #can be "free","free_x","free_y",
                         text.off = T,
                         use.ylims = T,
                         ymax.study = study,
                         y.off = F,
                         x.off = F,
                         facet.label.off = T,
                         font.size = global.font.size,
                         corr.method = "holm",
                         nhypo = 10,
                         point.size = 2,
                         rel.axis.size = .65
){

  dat %<>% sub.pl(study)
  xlabel = xlabel.pl(bl.other,fish.cdf = "fish")
  dat %<>% sig.pl(sig.level,corr.method,nhypo)
  lims = xlims(dat,set)

  #plot
  pl = ggplot(data = dat, aes(x = threshold,
                              y = ifelse(OR == Inf,
                                         OR_CI95_lower,
                                         OR)
  )
  )+
    geom_point(stat = "identity",
               size = point.size,
               shape =21,
               fill = "white")+ #plot outlines of all points.
    geom_point(stat = "identity",
               size = point.size,
               # shape = 21,
               mapping  = aes(alpha = sig))+
    scale_alpha_manual(values = c(0,1))+ #plot solid fill for all sig points, see sig.level fnx, makes >=.05 first level in factor, so 0,1 for alpha.
    geom_line(aes(group = if(Group == "Class"){Class}else if(Group == "Mutations"){Mutations} else if(Group == "Set"){Set} else{Study}))+
    xlab(xlabel)+
    ylab("Odds Ratio")+
    geom_hline(aes(yintercept = 1))+
    ggtitle(title)
  # +xlim(lims$xmin,lims$xmax)

  pl %<>% faceting(Facet,Scales)
  
  pl = pl+
    expand_limits(y = 0)+
    theme_cowplot(font_size = font.size)+
    theme(legend.position = "none",
          axis.text = element_text(size = rel(rel.axis.size))
    )
  
  pl %<>% x.ticks()
  pl %<>% text.fnx(text.off)
  if(use.ylims){
    pl = pl+ylim(0,ylims(dat,ymax.study)$ymax)
  }
  pl %<>% axes(y.off,x.off)
  pl %<>% facet.label(facet.label.off)
  
  #return
  pl
}

smbe.fig2.fish.fnx = function(dat = fig2.fish.dat,
                         sig.level = .05,
                         study = NULL,
                         bl.other = "other",
                         set = "Detrimentals",
                         title = study,
                         Group = "Study",
                         Facet = "Study", #otherwise use "none"
                         Scales = "fixed", #can be "free","free_x","free_y",
                         text.off = T,
                         use.ylims = T,
                         ymax.study = study,
                         y.off = F,
                         x.off = F,
                         facet.label.off = T,
                         font.size = global.font.size,
                         corr.method = "holm",
                         nhypo = 10,
                         point.size = 2,
                         rel.axis.size = .65,
                         outline.size = 1
){

  dat %<>% sub.pl(study)
  xlabel = xlabel.pl(bl.other,fish.cdf = "fish")
  dat %<>% sig.pl(sig.level,corr.method,nhypo)
  lims = xlims(dat,set)

  #plot
  pl = ggplot(data = dat, aes(x = threshold,
                              y = ifelse(OR == Inf,
                                         OR_CI95_lower,
                                         OR)
  )
  )+
    geom_point(stat = "identity",
               size = point.size,
               shape =21,
               fill = "white",
               color = "black",
               stroke = outline.size)+ #plot outlines of all points.
    geom_point(stat = "identity",
               size = point.size,
               # shape = 21,
               mapping  = aes(alpha = sig))+
    scale_alpha_manual(values = c(0,1))+ #plot solid fill for all sig points, see sig.level fnx, makes >=.05 first level in factor, so 0,1 for alpha.
    geom_line(aes(group = if(Group == "Class"){Class}else if(Group == "Mutations"){Mutations} else if(Group == "Set"){Set} else{Study}))+
    xlab(xlabel)+
    ylab("Odds Ratio")+
    geom_hline(aes(yintercept = 1))+
    ggtitle(title)
  # +xlim(lims$xmin,lims$xmax)

  pl %<>% faceting(Facet,Scales)
  
  pl = pl+
    expand_limits(y = 0)+
    theme_cowplot(font_size = font.size)+
    theme(legend.position = "none",
          axis.text = element_text(size = rel(rel.axis.size))
    )
  
  pl %<>% x.ticks()
  pl %<>% text.fnx(text.off)
  if(use.ylims){
    pl = pl+ylim(0,ylims(dat,ymax.study)$ymax)
  }
  pl %<>% axes(y.off,x.off)
  pl %<>% facet.label(facet.label.off)
  
  #return
  pl
}


fig2.cdf.fnx = function(dat = fig2.cdf.dat,
                        study = NULL,
                        bl.other = "other",
                        set = "Detrimentals",
                        title = study,
                        Facet = "Study",
                        Scales = "fixed", #can be "free","free_x","free_y"
                        text.off = T,
                        y.off = F,
                        x.off = F,
                        facet.label.off = T,
                        font.size = global.font.size,
                        rel.axis.size = .65
){
  
  dat %<>% sub.pl(study)
  lims = xlims.cdf(set = "Detrimentals")
  dat %<>% subset(Fitness.Mean<=lims$xmax & Fitness.Mean>=lims$xmin)
  xlabel = xlabel.pl(bl.other,fish.cdf = "cdf")
  
  #plot
  pl = ggplot(data = dat, aes(x = Fitness.Mean))+
    stat_ecdf(#size = 1,
              aes(linetype=TS_TV
                  )
    )+
  scale_linetype_manual(values = c("TS" = "solid","TV" = "dotted"))+ #used longdash before
    ylab("Percentile")+
    xlab(xlabel)+
    ggtitle(title)
  
  pl %<>% faceting(Facet,Scales)

  pl = pl+
    theme_cowplot(font_size = font.size)+
    theme(legend.position = "none",
          axis.text = element_text(size = rel(rel.axis.size))
    )
  
  pl %<>% x.ticks()
  pl %<>% text.fnx(text.off)
  pl %<>% axes(y.off,x.off)
  pl %<>% facet.label(facet.label.off)
  
  #return
  pl
}



pres.fig2.cdf.fnx = function(dat = fig2.cdf.dat,
                        study = NULL,
                        bl.other = "other",
                        set = "Detrimentals",
                        title = study,
                        Facet = "Study",
                        Scales = "fixed", #can be "free","free_x","free_y"
                        text.off = T,
                        y.off = F,
                        x.off = F,
                        facet.label.off = T,
                        font.size = global.font.size
){
  
  dat %<>% sub.pl(study)
  lims = xlims.cdf(set = "Detrimentals")
  dat %<>% subset(Fitness.Mean<=lims$xmax & Fitness.Mean>=lims$xmin)
  xlabel = xlabel.pl(bl.other,fish.cdf = "cdf")
  
  #plot
  pl = ggplot(data = dat, aes(x = Fitness.Mean))+
    stat_ecdf(size = 1.5,
              aes(color=TS_TV
                  )
    )+
  scale_color_manual(values = c("TS" = "blue","TV" = "red"))+ #used longdash before
    ylab("Percentile")+
    xlab(xlabel)+
    ggtitle(title)
  
  pl %<>% faceting(Facet,Scales)

  pl = pl+
    theme_cowplot(font_size = font.size)+
    theme(legend.position = "none",
          axis.text = element_text(size = rel(.65))
    )
  
  pl %<>% x.ticks()
  pl %<>% text.fnx(text.off)
  pl %<>% axes(y.off,x.off)
  pl %<>% facet.label(facet.label.off)
  
  #return
  pl
}


smbe.fig2.cdf.fnx = function(dat = fig2.cdf.dat,
                        study = NULL,
                        bl.other = "other",
                        set = "Detrimentals",
                        title = study,
                        Facet = "Study",
                        Scales = "fixed", #can be "free","free_x","free_y"
                        text.off = T,
                        y.off = F,
                        x.off = F,
                        facet.label.off = T,
                        font.size = global.font.size,
                        rel.axis.size = .65,
                        line.size = 1
){
  
  dat %<>% sub.pl(study)
  lims = xlims.cdf(set = "Detrimentals")
  dat %<>% subset(Fitness.Mean<=lims$xmax & Fitness.Mean>=lims$xmin)
  xlabel = xlabel.pl(bl.other,fish.cdf = "cdf")
  
  #plot
  pl = ggplot(data = dat, aes(x = Fitness.Mean))+
    stat_ecdf(size = line.size,
              aes(linetype=TS_TV,
                  color = TS_TV
                  )
    )+
  scale_linetype_manual(values = c("TS" = "solid","TV" = "solid"))+ #used longdash before
  scale_color_manual(values = c("TS" = "blue","TV" = "red"))+
    ylab("Percentile")+
    xlab(xlabel)+
    ggtitle(title)
  
  pl %<>% faceting(Facet,Scales)

  pl = pl+
    theme_cowplot(font_size = font.size)+
    theme(legend.position = "none",
          axis.text = element_text(size = rel(rel.axis.size))
    )
  
  pl %<>% x.ticks()
  pl %<>% text.fnx(text.off)
  pl %<>% axes(y.off,x.off)
  pl %<>% facet.label(facet.label.off)
  
  #return
  pl
}

#Faceting with facet titles cdf only, all axes titles-seems to be the best.----
fig2.cdf.flu = fig2.cdf.fnx(study = c("Total","Internal","Surface"),title = NULL,text.off = F,facet.label.off = F)
fig2.fish.flu = fig2.fish.fnx(study = c("Total","Internal","Surface"),title = NULL,text.off = F)

fig2.cdf.hiv = fig2.cdf.fnx(study = c("HIV Combined","HIV IN","HIV CA"),title = NULL,text.off = F,facet.label.off = F)
fig2.fish.hiv = fig2.fish.fnx(study = c("HIV Combined","HIV IN","HIV CA"),title = NULL,text.off = F)

fig2 = plot_grid(fig2.cdf.flu,fig2.fish.flu,fig2.cdf.hiv,fig2.fish.hiv,rel_heights = c(.6,.4,.6,.4),nrow = 4,align = "v",labels = c("A","","B"), label_size = pl.lab.font.size)

#pres
pres.fig2.cdf.int = pres.fig2.cdf.fnx(study = "Internal",facet.label.off = F,text.off = F,title = NULL)
pres.fig2.fish.int = fig2.fish.fnx(study = "Internal",title = NULL, text.off = F, facet.label.off = F)

#smbe
fig2.cdf.smbe = smbe.fig2.cdf.fnx(study = "Internal",title = "Internal Proteins",text.off = F,facet.label.off = T,font.size = 28,rel.axis.size = .85,line.size = 2)
ggsave("fig2.cdf.smbe.png",fig2.cdf.smbe,width = 9.13,height =6.5, units = "in")

fig2.fish.smbe = smbe.fig2.fish.fnx(study = "Internal",title = "Internal Proteins",text.off = F,facet.label.off = T,point.size = 9,font.size = 28,rel.axis.size = .85)
ggsave("fig2.fish.smbe.png",fig2.fish.smbe,width = 9.13,height =6.5, units = "in")


#end----
```

###4) Table 1: Spearman - Bloom vs. Lauring
```{r TABLE1}

table1 = bloomxla_corr %>% subset(Corr != "H1N1" & S_NS != "Both", select = c(Corr, n, r, p)) %>% mutate(Corr = map.xstop(Corr)) %>% rename(c(Corr = "Gene", n = "Number of Shared Mutations", r = "rho")) %>% mutate(rho = rho %>% round(2), p = p %>% signif(digits = 2)) %>% rename(c(p = "P value"))

# #asterisks instead of p values:
# table1 = bloomxla_corr %>% subset(Corr != "H1N1" & S_NS != "Both", select = c(Corr, n, r, p)) %>% mutate(Corr = map.xstop(Corr)) %>% rename(c(Corr = "Gene", n = "Number of Shared Mutations", r = "rho")) %>% mutate(sig = ifelse(p<.01,"*",""), rho = rho %>% round(2)) %>% mutate(rho = paste0(rho,sig)) %>% subset(select = - c(sig,p))

```
For HA, we have 3 synonymous mutations that are obviously the WT in the bloom data, so I excluded them from the correlation analysis, because the relative site preference will obviously be 1 no matter what because of how we calculate relative site preference (although our synonymous mutations might not be 1), so I don't think such points are informative, and likely would inflate correlation. See section above "Lauring-Bloom Correlations (Spearman)" for more info.

Rearranged table in excel.

###5) Fig. 3: Fisher & CDF - TsTv - Bloom
```{r FIG3}
#data----
fig3.fish.dat = fisher.bloom.det %>% subset(Study != "Bloom_Flu" & Study != "Bloom_H1N1") %>% mutate(Study = map.xstop(Study))

fig3.cdf.dat = fig2.cdf.dat %>% subset(Study %in% bl.levels)

# #check study names correct (have to do before changing levels)
# levels(fisher.bloom.det$Study)
# levels(fig3.fish.dat$Study)

fig3.fish.dat$Study = factor(fig3.fish.dat$Study,levels = bl.levels)
fig3.cdf.dat$Study = factor(fig3.cdf.dat$Study,levels = bl.levels)

fig3.table = fig3.fish.dat %>% subset(select = c(Study,threshold,OR,OR_CI95_lower,p.value)) %>% rename(c(Study = "Dataset",threshold = "Rel. Site Pref. Cutoff",OR = "Odds Ratio"))

# if("HIV.ENV" %in% study){
  #   pl = pl + ylim(0,
  #        max(dat %>% subset(OR==Inf) %>% .$OR_CI95_lower,
  #              dat %>% subset(OR<Inf) %>% .$OR) %>% round_any(.75,f = ceiling)
  #        )
  # }
  # 

#pl----
fig3.cdf.int = fig2.cdf.fnx(dat = fig3.cdf.dat,bl.other = "bl",study = bl.levels[1:2],title = NULL,text.off = F,facet.label.off = F)
fig3.fish.int = fig2.fish.fnx(dat = fig3.fish.dat,bl.other = "bl",study = bl.levels[1:2],title = NULL,use.ylims = F,text.off = F,point.size = 2)+scale_y_continuous(expand = c(.07,0)) #change from default .05, otherwise highest points get a little cut off, .06 still too.

fig3.cdf.surf = fig2.cdf.fnx(dat = fig3.cdf.dat,bl.other = "bl",study = bl.levels[3:4], title = NULL,text.off = F,facet.label.off = F)
fig3.fish.surf = fig2.fish.fnx(dat = fig3.fish.dat,bl.other = "bl",study = bl.levels[3:4],title = NULL,use.ylims = F,text.off = F,point.size = 2)+scale_y_continuous(expand = c(.07,0))

fig3 = plot_grid(fig3.cdf.int,fig3.fish.int,fig3.cdf.surf,fig3.fish.surf,rel_heights = c(.6,.4,.6,.4),nrow = 4,align = "v",labels = c("A","","B"), label_size = pl.lab.font.size)

#pres
pres.fig3.fish.flu = fig2.fish.fnx(dat = fig3.fish.dat,bl.other = "bl",study = bl.levels[1:3],title = NULL,use.ylims = F,text.off = F,point.size = 3.5,facet.label.off = F,corr.method = "NONE")


#smbe
fig3.int.smbe = smbe.fig2.fish.fnx(dat = fig3.fish.dat,bl.other = "bl",study = bl.levels[1:2],title = NULL,use.ylims = F,text.off = F,point.size = 9,font.size = 32, rel.axis.size = .85, facet.label.off = F)

fig3.surf.smbe = smbe.fig2.fish.fnx(dat = fig3.fish.dat,bl.other = "bl",study = bl.levels[3:4],title = NULL,use.ylims = F,text.off = F,point.size = 9,font.size = 32, rel.axis.size = .85, facet.label.off = F)

fig3.smbe = plot_grid(fig3.int.smbe,fig3.surf.smbe,nrow = 2,align = "hv")

ggsave("fig3.smbe.png",fig3.smbe,width = 19.33,height = 12.37, units = "in")


#end----
```

Notes:

1) For HIV.ENV, the first cutoff has no TS below it (50 TV below it), so the OR is Inf, causing two problems-The OR is at the top of the graph looks like it's cut off but isn't, that's just how Inf is plotted in ggplot. Second,there should be three stars on it as significant, but for this it can't be plotted above it, would have to be next to it.

One approach to that problem is that if it is Inf, take the lower CI (or -Inf, take the upper CI), this is what I did here for Inf. (Only affects HIV.ENV, first piont).

2)  Lots of significant values, unless graph is quite wide, the three vs two etc. stars are obscured. Could use 1 star for simply significant, less than .05, or a + for less than .01 and some other symbol(?) (remember should use postscript base fonts) for less than .001 (or make only two cut offs). But honestly, statistically speaking the only reason p-values below the alpha should be shown is concerns over correcting for multiple testing.

3) The OR for HIV.ENV at least for the first 2 points are quite high as compared to the flu OR, so if you plot fixed axes according to the HIV OR, then can't see the differences much of the flu ones, especially when make the full plot grid and there isn't much vertical space.  So I put all flu on one axes going up to about 2 and the hiv on a separate axes going up to about 5.  Not sure what to do, I guess the goal is to show a good sense of the magnitude of differences. I think that the CDF does the best job of this visually, but the OR gives easier to read off numbers, and when where it _appears_ as though not much difference in CDFeg, at the ends of the distribution, the OR can still be impressive, eg for NP.H1N1, you are almost 2X more likely to be below .75 relative pref if you are a TV as compared to a TS, even though the CDF doesn't look too impressive there.

###6) Fig. 4: Fisher & CDF - C/R
```{r FIG4}
#dat----
fig4.fish.dat = rbind(do.call(rbind,rc.env.dat) %>% do.call(rbind,.) %>% do.call(rbind,.),
      do.call(rbind,rc.flu.dat) %>% do.call(rbind,.) %>% do.call(rbind,.)
      ) %>% mutate(Study = map.xstop(Study),
                   Set = mapvalues(Set,"All Fitness Levels","All.Fitness.Levels"), 
                   Mutations = mapvalues(Mutations,c("All.Codon.Muts","MUT1"),c("Mut123","Mut1")))

fig4.cdf.dat = rbind(do.call(rbind,rc.cdf.env.dat) %>% do.call(rbind,.) %>% do.call(rbind,.),
                     do.call(rbind,rc.cdf.flu.dat) %>% do.call(rbind,.) %>% do.call(rbind,.)) %>% rename(c(dtset = "Study")) %>% mutate(Study = map.xstop(Study),Set = mapvalues(Set,"All Fitness Levels","All.Fitness.Levels"))


fig4.fish.dat$Study = factor(fig4.fish.dat$Study,levels = bl.levels)


fig4.cdf.dat$Study = factor(fig4.cdf.dat$Study,levels = bl.levels)

fig4.table = fig4.fish.dat %>% subset(Mutations == "TS.TV.Only" & Set == "Detrimentals", select = c(Study,Class,threshold,OR,p.value)) %>% rename(c(Study = "Dataset",threshold = "Rel. Site Pref. Cutoff",OR = "Odds Ratio"))

#pl fnx----
fig4.fish.fnx = function(dat = fig4.fish.dat,
                         sig.level = .05,
                         study = NULL,
                         bl.other = "bl",
                         muts = NULL,
                         set = NULL,
                         class = NULL,
                         title = class,
                         Group = "Class",
                         Color = "Class",
                         Facet = "Study",
                         Scales = "fixed", #can be "free","free_x","free_y"
                         text.off = T,
                         use.ylims = T,
                         ymax.study = study,
                         y.off = F,
                         x.off = F,
                         facet.label.off = T,
                         font.size = global.font.size,
                         point.size = 2,
                         corr.method = "holm",
                         nhypo = 19,
                         rel.axis.size = .65
){
  
  dat %<>% sub.pl(study,muts,set,class)
  xlabel = xlabel.pl(bl.other,fish.cdf = "fish")
  dat %<>% sig.pl(sig.level,corr.method,nhypo)
  lims = xlims(dat,set)
  
  #plot
  pl = ggplot(data = dat, aes(x = threshold,
                              y = ifelse(OR == Inf,
                                         OR_CI95_lower,
                                         OR),color = if(Color == "Class"){Class}else if(Color == "Mutations"){Mutations} else if(Color == "Set"){Set} else{Study}
  )
  )+
    scale_color_manual(values = c("black","red","blue","green"),name = Color)+
    geom_point(stat = "identity",
               size = point.size,
               shape =21,
               fill = "white")+ #plot outlines of all points.
    geom_point(stat = "identity",
               size = point.size,
               # shape = 21,
               mapping  = aes(alpha = sig))+
    scale_alpha_manual(values =c(0,1))+  #plot solid fill for all sig points, see sig.level fnx, makes >=.05 first level in factor, so 0,1 for alpha.
    geom_line(aes(group = if(Group == "Class"){Class}else if(Group == "Mutations"){Mutations} else if(Group == "Set"){Set} else{Study}))+
    xlab(xlabel)+
    ylab("Odds Ratio")+
    geom_hline(aes(yintercept = 1))+
    ggtitle(title)+
    xlim(lims$xmin,lims$xmax)
  
  pl %<>% faceting(Facet,Scales)
  
  pl = pl+
    expand_limits(y = 0)+
    theme_cowplot(font_size = font.size)+
    theme(legend.position = "none",
          axis.text = element_text(size = rel(rel.axis.size))
    )
  pl %<>% x.ticks()
  pl %<>% text.fnx(text.off)
  if(use.ylims){
    pl = pl+ylim(0,ylims(dat,ymax.study)$ymax)
  }
  pl %<>% axes(y.off,x.off)
  pl %<>% facet.label(facet.label.off)
  #return
  pl
}

smbe.fig4.fish.fnx = function(dat = fig4.fish.dat,
                         sig.level = .05,
                         study = NULL,
                         bl.other = "bl",
                         muts = NULL,
                         set = NULL,
                         class = NULL,
                         title = class,
                         Group = "Class",
                         Color = "Class",
                         Facet = "Study",
                         Scales = "fixed", #can be "free","free_x","free_y"
                         text.off = T,
                         use.ylims = T,
                         ymax.study = study,
                         y.off = F,
                         x.off = F,
                         facet.label.off = T,
                         font.size = global.font.size,
                         point.size = 2,
                         corr.method = "holm",
                         nhypo = 19,
                         rel.axis.size = .65,
                         outline.size = 1
){
  
  dat %<>% sub.pl(study,muts,set,class)
  xlabel = xlabel.pl(bl.other,fish.cdf = "fish")
  dat %<>% sig.pl(sig.level,corr.method,nhypo)
  lims = xlims(dat,set)
  
  #plot
  pl = ggplot(data = dat, aes(x = threshold,
                              y = ifelse(OR == Inf,
                                         OR_CI95_lower,
                                         OR),color = if(Color == "Class"){Class}else if(Color == "Mutations"){Mutations} else if(Color == "Set"){Set} else{Study}
  )
  )+
    scale_color_manual(values = c("black","red","blue","green"),name = Color)+
    geom_point(stat = "identity",
               size = point.size,
               shape =21,
               fill = "white",
               stroke = outline.size)+ #plot outlines of all points.
    geom_point(stat = "identity",
               size = point.size,
               stroke = outline.size,
               # shape = 21,
               mapping  = aes(alpha = sig))+
    scale_alpha_manual(values =c(0,1))+  #plot solid fill for all sig points, see sig.level fnx, makes >=.05 first level in factor, so 0,1 for alpha.
    geom_line(aes(group = if(Group == "Class"){Class}else if(Group == "Mutations"){Mutations} else if(Group == "Set"){Set} else{Study}))+
    xlab(xlabel)+
    ylab("Odds Ratio")+
    geom_hline(aes(yintercept = 1))+
    ggtitle(title)+
    xlim(lims$xmin,lims$xmax)
  
  pl %<>% faceting(Facet,Scales)
  
  pl = pl+
    expand_limits(y = 0)+
    theme_cowplot(font_size = font.size)+
    theme(legend.position = "none",
          axis.text = element_text(size = rel(rel.axis.size))
    )
  pl %<>% x.ticks()
  pl %<>% text.fnx(text.off)
  if(use.ylims){
    pl = pl+ylim(0,ylims(dat,ymax.study)$ymax)
  }
  pl %<>% axes(y.off,x.off)
  pl %<>% facet.label(facet.label.off)
  #return
  pl
}


fig4.cdf.fnx = function(dat = fig4.cdf.dat,
                        study = NULL,
                        bl.other = "bl",
                        muts = NULL,
                        set = NULL,
                        class = NULL,
                        title = class,
                        Color = "Class",
                        Facet = "Study",
                        Scales = "fixed", #can be "free","free_x","free_y"
                        text.off = T,
                        y.off = F,
                        x.off = F,
                        facet.label.off = T,
                        font.size = global.font.size
){
  
  dat %<>% sub.pl(study,muts,set,class)
  lims = xlims.cdf(set = "Detrimentals")
  dat %<>% subset(Rel_site_pref<=lims$xmax & Rel_site_pref>=lims$xmin)
  xlabel = xlabel.pl(bl.other,fish.cdf = "cdf")
  
  #plot
  pl = ggplot(data = dat, aes(x = Rel_site_pref))+
    stat_ecdf(aes(linetype=R.C,
                  color = if(Color == "Class"){Class}else if(Color == "Mutations"){Mutations} else if(Color == "Set"){Set} else{Study}
                  )
    )+
    scale_color_manual(values = c("black","red","blue","green"),name = Color)+ #takes into account when there are 4 values, green is 4th, only when color = Mutations will have 4 colors.
  scale_linetype_manual(values = c("C" = "solid","R" = "dotted"))+
    ylab("Percentile")+
    xlab(xlabel)+
    ggtitle(title)
  
  pl %<>% faceting(Facet,Scales)
  
  pl = pl+
    theme_cowplot(font_size = font.size)+
    theme(legend.position = "none",
          axis.text = element_text(size = rel(.65))
    )
  
  pl %<>% x.ticks()
  pl %<>% text.fnx(text.off)
  pl %<>% axes(y.off,x.off)
  pl %<>% facet.label(facet.label.off)
  
  #return
  pl
}

#pl----
##Mut123 
fig4.cdf.int.m123 = fig4.cdf.fnx(muts = "Mut123",set = "Detrimentals",study = bl.levels[1:2],title = NULL,text.off = F,facet.label.off = F)
fig4.fish.int.m123 = fig4.fish.fnx(muts = "Mut123",set = "Detrimentals",study = bl.levels[1:2],title = NULL,text.off = F)


fig4.cdf.surf.m123 = fig4.cdf.fnx(muts = "Mut123",set = "Detrimentals",study = bl.levels[3:4],title = NULL,text.off = F,facet.label.off = F)
fig4.fish.surf.m123 = fig4.fish.fnx(muts = "Mut123",set = "Detrimentals",study = bl.levels[3:4],title = NULL,text.off = F)


fig4.m123 = plot_grid(fig4.cdf.int.m123,fig4.fish.int.m123,fig4.cdf.surf.m123,fig4.fish.surf.m123,rel_heights = c(.6,.4,.6,.4),nrow = 4,align = "v",labels = c("A","","B"), label_size = pl.lab.font.size)


#Mut1 mutations (incl boths)
fig4.cdf.int.m1 = fig4.cdf.fnx(muts = "Mut1",set = "Detrimentals",study = bl.levels[1:2],title = NULL,text.off = F,facet.label.off = F)
fig4.fish.int.m1 = fig4.fish.fnx(muts = "Mut1",set = "Detrimentals",study = bl.levels[1:2],title = NULL,text.off = F)


fig4.cdf.surf.m1 = fig4.cdf.fnx(muts = "Mut1",set = "Detrimentals",study = bl.levels[3:4],title = NULL,text.off = F,facet.label.off = F)
fig4.fish.surf.m1 = fig4.fish.fnx(muts = "Mut1",set = "Detrimentals",study = bl.levels[3:4],title = NULL,text.off = F)


fig4.m1 = plot_grid(fig4.cdf.int.m1,fig4.fish.int.m1,fig4.cdf.surf.m1,fig4.fish.surf.m1,rel_heights = c(.6,.4,.6,.4),nrow = 4,align = "v",labels = c("A","","B"), label_size = pl.lab.font.size)

#TS.TV.Only mutations
fig4.cdf.int.tstv = fig4.cdf.fnx(muts = "TS.TV.Only",set = "Detrimentals",study = bl.levels[1:2],title = NULL,text.off = F,facet.label.off = F)
fig4.fish.int.tstv = fig4.fish.fnx(muts = "TS.TV.Only",set = "Detrimentals",study = bl.levels[1:2],title = NULL,text.off = F)


fig4.cdf.surf.tstv = fig4.cdf.fnx(muts = "TS.TV.Only",set = "Detrimentals",study = bl.levels[3:4],title = NULL,text.off = F,facet.label.off = F)
fig4.fish.surf.tstv = fig4.fish.fnx(muts = "TS.TV.Only",set = "Detrimentals",study = bl.levels[3:4],title = NULL,text.off = F)


fig4.tstv = plot_grid(fig4.cdf.int.tstv,fig4.fish.int.tstv,fig4.cdf.surf.tstv,fig4.fish.surf.tstv,rel_heights = c(.6,.4,.6,.4),nrow = 4,align = "v",labels = c("A","","B"), label_size = pl.lab.font.size)

#SMBE
fig4.smbe = smbe.fig4.fish.fnx(muts = "TS.TV.Only",set = "Detrimentals",study = bl.levels[4],title = NULL,text.off = F, font.size = 28,rel.axis.size = .85, point.size = 7,facet.label.off = F)

ggsave("fig4.smbe.png",fig4.smbe,width = 8.62,height = 7, units = "in")

# horizontal
# fig4.fish.det = fig4.fish.fnx(muts = "Mut123",set = "Detrimentals")#charge = black, polarity = red, polarity & volume = blue
# # fig4.fish.bf = fig4.fish.fnx(muts = "Mut123",set = "Beneficials")
# 
# fig4.cdf.det = fig4.cdf.fnx(muts = "Mut123",set = "Detrimentals")#charge = black, polarity = red, polarity & volume = blue
# # fig4.cdf.bf = fig4.cdf.fnx(muts = "Mut123",set = "Beneficials")+xlim(1,8)
# 
# fig4 = plot_grid(fig4.cdf.det,fig4.fish.det,nrow = 2, ncol = 1,align = "v")


# #Option 2-one class individually, charge as example
# fig4.fish.det.2 = fig4.fish.fnx(muts = "Mut123",set = "Detrimentals",class = "Charge")
# # fig4.fish.bf.2 = fig4.fish.fnx(muts = "Mut123",set = "Beneficials",class = "Charge")
# 
# fig4.cdf.det2 = fig4.cdf.fnx(muts = "Mut123",set = "Detrimentals",class = "Charge")
# # fig4.cdf.bf.2 = fig4.cdf.fnx(muts = "Mut123",set = "Beneficials",class = "Charge")
# 
# fig4.2 = plot_grid(fig4.fish.det.2,fig4.cdf.det2,labels = c("A","B"),nrow = 2, ncol = 1,align = "v")

# other mutations options doesn't change much.
# fig4.fish.fnx(muts = "TS.TV.Only",set = "Detrimentals",title = "TS.TV.Only")
# fig4.fish.fnx(muts = "TS.TV.Only",set = "Beneficials",title = "TS.TV.Only")
# 
# fig4.fish.fnx(muts = "MUT1",set = "Detrimentals",title = "MUT1")
# fig4.fish.fnx(muts = "MUT1",set = "Beneficials",title = "MUT1")
# 
# fig4.fish.fnx(muts = "Mut23",set = "Detrimentals",title = "MUT23")
# fig4.fish.fnx(muts = "Mut23",set = "Beneficials",title = "MUT23")

#end----
```
For option 1, with color:
In A and B, Black = Charge, Red = Polarity, Blue = Polarity & Volume.
In A, filled circles are significant.
In B, dashed lines = radical, solid = conservative

For option 2, black and white, each class on its own:
In A, filled circles are significant.
In B, dashed lines = radical, solid lines = conservative

For cdfs, obviously option 2 is better, though option 1 is better in terms of not having so many figures.  For beneficials, if show, option 2 is nescessary.

Figure 4, shown in paper is fitness difference between C and R of mutations used for Ts vs Tv fitness differences, that is, amino acid substitutions accessible by either a single Ts or a single Tv and not both.  Figure is saved as "fig4.tstv".  Other figures here are also made, C vs. R fitness for all amino acid substitutions accessible by by a single point mutation ("fig4.m1") and all amino acid substitutions ("fig4.m123").


###7) Table 2: Fisher TsTv OR for C/R
```{r TABLE2}
#data----
table2.dat = tstv.rc
row.names(table2.dat) = NULL
table2.dat %<>% rename(c(pvalue = "p.value")) %>% mutate(Study = map.xstop(Gene),Class = mapvalues(Class,c("charge","polarity","pv"),c("Charge","Polarity","Polarity & Volume")))

#table----
table2 = table2.dat %>% subset(Study %in% bl.levels,select = c(Gene,Class,OR,p.value)) %>% arrange(Class) %>% mutate(sig = ifelse(p.value<.05,"*","")) %>% subset(select = c(Gene,Class,OR,sig)) %>% mutate(OR = round(OR,2) %>% paste0(sig)) %>% subset(select = c(Gene,Class,OR))



#end----
```

No multiple testing correction for these, since just want to know if each of the classes are significant in each of the genes, not if any of them are significant.


###8) Fig. 5 & 6: Fisher & CDF - TsTv Among C/R
```{r FIG56}
#data----
##fish----
fig5.r.fish = rbind(
  rbind(r.fish$charge$data$fisher.bloom.det %>% mutate(Set = "Detrimentals",Class = "Charge"),
        r.fish$polarity$data$fisher.bloom.det %>% mutate(Set = "Detrimentals",Class = "Polarity"),
        r.fish$pv$data$fisher.bloom.det %>% mutate(Set = "Detrimentals",Class = "Polarity & Volume")
  ),
  rbind(r.fish$charge$data$fisher.bloom.bf %>% mutate(Set = "Beneficials",Class = "Charge"),
        r.fish$polarity$data$fisher.bloom.bf %>% mutate(Set = "Beneficials",Class = "Polarity"),
        r.fish$pv$data$fisher.bloom.bf %>% mutate(Set = "Beneficials",Class = "Polarity & Volume")
  )
) %>% mutate(Study = map.xstop(Study))


fig6.c.fish = rbind(rbind(c.fish$charge$data$fisher.bloom.det %>% mutate(Set = "Detrimentals",Class = "Charge"),
                c.fish$polarity$data$fisher.bloom.det %>% mutate(Set = "Detrimentals",Class = "Polarity"),
                c.fish$pv$data$fisher.bloom.det %>% mutate(Set = "Detrimentals",Class = "Polarity & Volume")
                ),
          rbind(c.fish$charge$data$fisher.bloom.bf %>% mutate(Set = "Beneficials",Class = "Charge"),
                c.fish$polarity$data$fisher.bloom.bf %>% mutate(Set = "Beneficials",Class = "Polarity"),
                c.fish$pv$data$fisher.bloom.bf %>% mutate(Set = "Beneficials",Class = "Polarity & Volume")
                )
) %>% mutate(Study = map.xstop(Study))

fig5.r.fish$Study = factor(fig5.r.fish$Study, levels = bl.levels)
fig6.c.fish$Study = factor(fig6.c.fish$Study, levels = bl.levels)
##cdf----
fig5.r.cdf = rbind(
  rbind(r.cdf$charge$data$det %>% mutate(Set = "Detrimentals",Class = "Charge"),
        r.cdf$polarity$data$det %>% mutate(Set = "Detrimentals",Class = "Polarity"),
        r.cdf$pv$data$det %>% mutate(Set = "Detrimentals",Class = "Polarity & Volume")
  ),
  rbind(r.cdf$charge$data$bf %>% mutate(Set = "Beneficials",Class = "Charge"),
        r.cdf$polarity$data$bf %>% mutate(Set = "Beneficials",Class = "Polarity"),
        r.cdf$pv$data$bf %>% mutate(Set = "Beneficials",Class = "Polarity & Volume")
  )
) %>% mutate(Study = map.xstop(Study))

fig6.c.cdf = rbind(
  rbind(c.cdf$charge$data$det %>% mutate(Set = "Detrimentals",Class = "Charge"),
        c.cdf$polarity$data$det %>% mutate(Set = "Detrimentals",Class = "Polarity"),
        c.cdf$pv$data$det %>% mutate(Set = "Detrimentals",Class = "Polarity & Volume")
  ),
  rbind(c.cdf$charge$data$bf %>% mutate(Set = "Beneficials",Class = "Charge"),
        c.cdf$polarity$data$bf %>% mutate(Set = "Beneficials",Class = "Polarity"),
        c.cdf$pv$data$bf %>% mutate(Set = "Beneficials",Class = "Polarity & Volume")
  )
) %>% mutate(Study = map.xstop(Study))

fig5.table = fig5.r.fish %>% subset(Set == "Detrimentals", select = c(Study,Class,threshold,OR,OR_CI95_lower,p.value)) %>% rename(c(Study = "Dataset",threshold = "Rel. Site Pref. Cutoff",OR = "Odds Ratio"))

fig6.table = fig6.c.fish %>% subset(Set == "Detrimentals", select = c(Study,Class,threshold,OR,OR_CI95_lower,p.value)) %>% rename(c(Study = "Dataset",threshold = "Rel. Site Pref. Cutoff",OR = "Odds Ratio"))

#pl fnx----
fig5.cdf.fnx = function(
  dat,
  study = NULL,
  bl.other = "bl",
  muts = NULL,
  set = NULL,
  class = NULL,
  title = class,
  Color = "Class",
  Facet = "Study",
  Scales = "fixed", #can be "free","free_x","free_y"
  text.off = T,
  y.off = F,
  x.off = F,
  facet.label.off = T,
  font.size = global.font.size
){
  
  dat %<>% sub.pl(study,muts,set,class)
  lims = xlims.cdf(set)
  dat %<>% subset(Fitness.Mean>=lims$xmin & Fitness.Mean<=lims$xmax)
  xlabel = xlabel.pl(bl.other,fish.cdf = "cdf")
  
  #plot
  pl = ggplot(data = dat, aes(x = Fitness.Mean))+
    stat_ecdf(
              aes(linetype=TS_TV,
                  color = if(Color == "Class"){Class}else if(Color == "Mutations"){Mutations} else if(Color == "Set"){Set} else{Study}
                  )
    )+
    scale_color_manual(values = c("black","red","blue","green"),name = Color)+ #takes into account when there are 4 values, green is 4th, only when color = Mutations will have 4 colors.
  scale_linetype_manual(values = c("TS" = "solid","TV" = "dotted"))+
    ylab("Percentile")+
    xlab(xlabel)+
    ggtitle(title)
  
  pl %<>% faceting(Facet,Scales)
  
  
  pl = pl+
    theme_cowplot(font_size = font.size)+
    theme(legend.position = "none",
          axis.text = element_text(size = rel(.65))
    )
  
  pl %<>% x.ticks()
  pl %<>% text.fnx(text.off)
  pl %<>% axes(y.off,x.off)
  pl %<>% facet.label(facet.label.off)
  #return
  pl
}


#Option 1-all classes together----
#fig6
fig5.r.cdf.int = fig5.cdf.fnx(dat = fig5.r.cdf,set = "Detrimentals",study = bl.levels[1:2],title = NULL,text.off = F,facet.label.off = F)
fig5.r.fish.int = fig4.fish.fnx(dat = fig5.r.fish,set = "Detrimentals",study = bl.levels[1:2],title = NULL,text.off = F,use.ylims = F)

fig5.r.cdf.surf = fig5.cdf.fnx(dat = fig5.r.cdf,set = "Detrimentals",study = bl.levels[3:4],title = NULL,text.off = F,facet.label.off = F)
fig5.r.fish.surf = fig4.fish.fnx(dat = fig5.r.fish,set = "Detrimentals",study = bl.levels[3:4],title = NULL,text.off = F,use.ylims = F)

fig5.r = plot_grid(fig5.r.cdf.int,fig5.r.fish.int,fig5.r.cdf.surf,fig5.r.fish.surf,rel_heights = c(.6,.4,.6,.4),nrow = 4,align = "v",labels = c("A","","B"), label_size = pl.lab.font.size)

#fig
fig6.c.cdf.int = fig5.cdf.fnx(dat = fig6.c.cdf,set = "Detrimentals",study = bl.levels[1:2],title = NULL,text.off = F,facet.label.off = F)
fig6.c.fish.int = fig4.fish.fnx(dat = fig6.c.fish,set = "Detrimentals",study = bl.levels[1:2],title = NULL,text.off = F,use.ylims = F)

fig6.c.cdf.surf = fig5.cdf.fnx(dat = fig6.c.cdf,set = "Detrimentals",study = bl.levels[3:4],title = NULL,text.off = F,facet.label.off = F)
fig6.c.fish.surf = fig4.fish.fnx(dat = fig6.c.fish,set = "Detrimentals",study = bl.levels[3:4],title = NULL,text.off = F,use.ylims = F)

fig6.c = plot_grid(fig6.c.cdf.int,fig6.c.fish.int,fig6.c.cdf.surf,fig6.c.fish.surf,rel_heights = c(.6,.4,.6,.4),nrow = 4,align = "v",labels = c("A","","B"), label_size = pl.lab.font.size)

#pres
pres.fig5.fish.np = fig4.fish.fnx(dat = fig5.r.fish,set = "Detrimentals",class  = "Charge",study = "NP (H1N1)",title = NULL,use.ylims = F,text.off = F,facet.label.off = F,corr.method = "NONE",point.size = 3.5)+ylim(0,3)
pres.fig3.fish.np = fig2.fish.fnx(dat = fig3.fish.dat,bl.other = "bl",study = "NP (H1N1)",title = NULL,use.ylims = F,text.off = F,point.size = 3.5,facet.label.off = F,corr.method = "NONE")+ylim(0,3)

pres.fig3.fish.env = fig2.fish.fnx(dat = fig3.fish.dat,bl.other = "bl",study = "HIV ENV",title = NULL,use.ylims = F,text.off = F,point.size = 3.5,facet.label.off = F,corr.method = "NONE")+ylim(0,6)
pres.fig5.fish.env = fig4.fish.fnx(dat = fig5.r.fish,set = "Detrimentals",class  = "Polarity",study = "HIV ENV",title = NULL,use.ylims = F,text.off = F,facet.label.off = F,corr.method = "NONE",point.size = 3.5)+ylim(0,6)

#SMBE
fig5.smbe = smbe.fig4.fish.fnx(dat = fig6.c.fish,set = "Detrimentals",study = bl.levels[4],title = NULL,text.off = F, font.size = 28,rel.axis.size = .85, point.size = 7,facet.label.off = F,use.ylims = F)

ggsave("fig5.smbe.png",fig5.smbe,width = 8.62,height = 7, units = "in")


#prev. option 1----
# fig6.r.fish.det = fig4.fish.fnx(dat = fig6.r.fish,set = "Detrimentals")
# #charge = black, polarity = red, polarity & volume = blue
# # fig6.r.fish.bf = fig4.fish.fnx(dat = fig6.r.fish,set = "Beneficials")
# fig6.r.cdf.det = fig6.cdf.fnx(dat = fig6.r.cdf,set = "Detrimentals")#charge = black, polarity = red, polarity & volume = blue
# # fig4.r.cdf.bf = fig6.cdf.fnx(dat = fig6.r.cdf,set = "Beneficials")+xlim(1,10) #same limit as last threshold in fish
# 
# fig6.c.fish.det = fig4.fish.fnx(dat = fig6.c.fish,set = "Detrimentals")#charge = black, polarity = red, polarity & volume = blue
# # fig6.c.fish.bf = fig4.fish.fnx(dat = fig6.c.fish,set = "Beneficials")
# fig6.c.cdf.det = fig6.cdf.fnx(dat = fig6.c.cdf,set = "Detrimentals")#charge = black, polarity = red, polarity & volume = blue
# # fig4.c.cdf.bf = fig6.cdf.fnx(dat = fig6.c.cdf,set = "Beneficials")+xlim(1,10) #same limit as last threshold in fish
# 
# fig6.r = plot_grid(fig6.r.fish.det,fig6.r.cdf.det,labels = c("A","B"),nrow = 2, ncol = 1,align = "v")
# fig6.c = plot_grid(fig6.c.fish.det,fig6.c.cdf.det,labels = c("A","B"),nrow = 2, ncol = 1,align = "v")
#Option 2-one class individually, Polarity & Volume as example----
# fig6.r.fish.det.2 = fig4.fish.fnx(dat = fig6.r.fish,set = "Detrimentals",class = "Polarity & Volume")
# fig6.r.cdf.det.2 = fig6.cdf.fnx(dat = fig6.r.cdf,set = "Detrimentals",class = "Polarity & Volume")
# 
# fig4.fish.fnx(dat = fig6.c.fish,set = "Detrimentals",class = "Charge", study = "NP.H1N1")
# 
# 
# fig6.c.fish.det.2 = fig4.fish.fnx(dat = fig6.c.fish,set = "Detrimentals",class = "Polarity & Volume")
# fig6.c.cdf.det.2 = fig6.cdf.fnx(dat = fig6.c.cdf,set = "Detrimentals",class = "Polarity & Volume")
# 
# fig6.r.2 = plot_grid(fig6.r.fish.det.2,fig6.r.cdf.det.2,labels = c("A","B"),nrow = 2, ncol = 1,align = "v")
# fig6.c.2 = plot_grid(fig6.c.fish.det.2,fig6.c.cdf.det.2,labels = c("A","B"),nrow = 2, ncol = 1,align = "v")

#end----
```



###Save PDF
```{r SAVE}

#fig 1
ggsave("fig1.pdf",fig1,width = 16.9,height = 24, units = "cm")

#fig 2
ggsave("fig2.pdf",fig2,width = 16.9,height = 24,units = "cm")

#table 1
write.csv(table1, file = "TABLE1.csv", row.names = F)

#fig 3
ggsave("fig3.pdf",fig3,width = 16.9,height = 24,units = "cm")

#fig 4
# ggsave("fig4.m123.pdf",fig4.m123,width = 16.9,height = 24,units = "cm")
# ggsave("fig4.m1.pdf",fig4.m1,width = 16.9,height = 24,units = "cm")
ggsave("fig4.pdf",fig4.tstv,width = 16.9,height = 24,units = "cm")

#table 2
table2 %>% write.csv(file = "TABLE2.csv",row.names = F)

#fig 5 & 6
ggsave("fig5.pdf",fig5.r,width = 16.9,height = 24, units = "cm")
ggsave("fig6.pdf",fig6.c,width = 16.9,height = 24, units = "cm")


```


##Supplemental Figures & Tables
1) Supplemental Table 1: N (& Power, removed)
2) Supplemental Table 1.removed: AUC Viable Mutations-Other Viruses (removed)
3) Supplemental Figure 1.removed: Coefficient of Variance-Other Viruses (removed)
4) Supplemental Figure 1: AUC Decreasing Thresholds
5) Supplemental Table 2: Amino Acid Categories
###1) Supplemental Table 1: N (& Power, removed in final manuscript)
```{r SUPP.TABLE1}
n = each %>% subset(grepl("STOP",set) & (set == "detrimental_xSTOP"|set == "detrimental & viable_xSTOP") & Study != "Bloom_Flu" & Study != "Lauring_gw", select = c(Study, set, N_TS,N_TV)) 

n$Study = n$Study %>% as.vector
n$set = n$set %>% as.vector

n.both = ddply(n,.(Study,set), summarize, TS = N_TS, TV = N_TV, Total = N_TS+N_TV) %>% mutate(set = mapvalues(set, c("detrimental_xSTOP", "detrimental & viable_xSTOP"), c("All","Viable")),Study = map.xstop(Study))

power = as.data.frame(power_n_sim) %>% 
  mutate(set = rownames(power_n_sim)) %>%
  subset(set == "detrimental_xSTOP" | set == "detrimental & viable_xSTOP") %>%
  melt(.(set)) %>%
  mutate(set = mapvalues(set, c("detrimental_xSTOP", "detrimental & viable_xSTOP"), c("All","Viable"))) %>%
  rename(c(variable = "Study", value = "Power")) %>%
  mutate(Power = Power * 100,Study = map.xstop(Study))

power = subset(n.both,! Study %in% power$Study, select = c(Study, set)) %>% mutate(Power = NA) %>% rbind(power)

table1.removed = merge(n.both,power) %>% arrange(set,-Total) %>% mutate(set = mapvalues(set,"All","All Fitness Levels")) %>% rename(c(Study = "Dataset"))

supp.table1 = table1.removed %>% subset(set == "All Fitness Levels", select = c(Dataset,TS,TV,Total))

```

###2) Supplemental Table 1.removed: AUC Viable Mutations-Other Viruses (removed)
```{r SUPP.TABLE1.removed}
supp.table1.removed = each %>%
  subset(set == "detrimental & viable_xSTOP",select = -c(alternative,statistic.W,N_TS,N_TV,set)) %>% 
  mutate(Study = map.xstop(Study),p.value = round(p.value %>% as.vector %>% as.numeric,3),auc = round(auc %>% as.vector %>% as.numeric,3)) %>% subset(!(Study %in% c(to.names.bl,"Bloom_Flu","Lauring_gw"))) %>% rename(c(Study = "Dataset",auc = "AUC",p.value = "p-value")) %>% subset(select = c("Dataset","AUC","p-value")) %>% arrange(Dataset)

row.names(supp.table1.removed) = NULL

supp.table1.removed = merge(table1.removed %>% subset(set == "Viable"),supp.table1.removed) %>% arrange(-Total) %>% subset(select = c("Dataset","Total","AUC","p-value")) %>% rename(c("p-value" = "P value", Total = "N"))   #adds N for viable, arranges in decreasing order of N.

```

###3) Supplemental Figure 1.removed: Coefficient of Variance-Other Viruses
```{r SUPP.FIG1.removed}
supp.fig1.dat.removed = all_studies %>% subset(!grepl("Bloom",Study) & !is.na(Fitness.SD) & S_NS == "NS" & (is.na(Dataset)|Dataset != "Described") & Fitness.Mean<=1 & !grepl("stop|STOP",AA_Sub) & !(grepl("Stop",AA_Sub) & Study == "VSV"), select = c(S_NS, Fitness.CV,Study,Dataset)) %>%
  mutate(Study = map.xstop(Study)) %>%
  mutate(Study = factor(Study,levels = other.levels))

supp.fig1.removed = ggplot(data = supp.fig1.dat.removed)+
  geom_boxplot(aes(x = Study,y = Fitness.CV %>% as.vector %>% as.numeric))+
  ylab("Coefficient of Variance for Each Mutation")

```
CV for ours and other viral studies, detrimental missense mutations only that we use for analysis. TEV only reported SEM for 24 of the total mutations. 


###4) Supplemental Figure 1: AUC Decreasing Thresholds
```{r SUPP.FIG1}

#data----
supp.fig1.dat = dec_thresholds %>% subset(gr.xstop(Study) & !grepl("Lauring_gw|Bloom_Flu|Bloom_H1N1",Study)) %>% mutate(Study = map.xstop(Study)) %>% mutate(Study = factor(Study,levels = c(other.levels,bl.levels)))

supp.fig1.dat$p.value = supp.fig1.dat$p.value %>% as.vector %>% as.numeric

supp.fig1.table = supp.fig1.dat %>% subset(grepl("Total|Surface|Internal|HIV Combined|HIV IN|HIV CA",Study), select = c(Study,threshold,auc,p.value)) %>% rename(c(auc = "AUC",Study = "Dataset",threshold = "Rel. Fitness Threshold"))

#pl----
supp.fig1 = fig1.fnx(dat = supp.fig1.dat,study = c("Total","Internal","Surface","HIV Combined","HIV IN","HIV CA"),title = NULL,text.off = F,point.size = 2)+facet_wrap(~Study,nrow = 2,ncol = 3)


```

###5) Supplemental Table 2: Amino Acid Categories
```{r SUPP.TABLE2}
supp.table2.dat = aa %>% mutate(charge = mapvalues(charge,c("neut","pos","neg"),c("neutral","positive","negative")),
                                pol_vol = gsub("neut","neutral",pol_vol) %>% gsub("_"," & ",.)) %>% arrange(pol_vol,charge,polarity) %>% subset(select = c(AA,code1,pol_vol,charge,polarity)) %>% rename(c(AA = "Amino Acids",pol_vol = "Polarity.Size",charge = "Charge", polarity = "Polarity"))

#OPTION 2

supp.table2 = rbind(c(Classification = "CHARGE",""),
      ddply(supp.table2.dat,"Charge",summarise,"Amino Acids" = paste(code1,collapse = ",")) %>% rename(c("Charge" = "Classification")),
c(Classification = "POLARITY",""),
ddply(supp.table2.dat,"Polarity",summarise,"Amino Acids" = paste(code1,collapse = ",")) %>% rename(c("Polarity" = "Classification")),
c(Classification = "POLARITY & SIZE",""),
ddply(supp.table2.dat,"Polarity.Size",summarise,"Amino Acids" = paste(code1,collapse = ",")) %>% rename(c("Polarity.Size" = "Classification")))



```
Boxes added in excel.


###Save PDF Supp.
```{r SAVE}

# write.csv(table1.removed, file = "TABLE1.removed.csv", row.names = F)
write.csv(supp.table1, file = "SUPP.TABLE1.csv", row.names = F)

# write.csv(supp.table1.removed, file = "SUPP.TABLE1.removed.csv", row.names = F)

# ggsave("supp.fig1.removed.pdf",supp.fig1.removed,width = 16.9,height = 24,units = "cm")

ggsave("supp.fig1.pdf",supp.fig1,width = 16.9,height = 24,units = "cm")

write.csv(supp.table2, file = "SUPP.TABLE2.csv", row.names = F)


```




##Github Tables
Tables giving AUC or OR and raw, un-adjusted p-values for all figures with AUC or OR.
```{r GITHUB.TABLES}

write.csv(fig1.table, file = "fig1.table.csv", row.names = F)
write.csv(fig2.table, file = "fig2.table.csv", row.names = F)
write.csv(fig3.table, file = "fig3.table.csv", row.names = F)
write.csv(fig4.table, file = "fig4.table.csv", row.names = F)
write.csv(fig5.table, file = "fig5.table.csv", row.names = F)
write.csv(fig6.table, file = "fig6.table.csv", row.names = F)
write.csv(supp.fig1.table, file = "supp.fig1.table.csv", row.names = F)


```

